<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Summary • rbiom</title><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/Noto_Sans-0.4.9/font.css" rel="stylesheet"><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet"><meta property="og:title" content="Summary"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="articles/rbiom.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="articles/aes.html">Mapping Metadata to Aesthetics</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Summary</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/paper.md" class="external-link"><code>paper.md</code></a></small>
    </div>


<div id="summary" class="section level1">

<p>Microbes live all around us, on us, and even inside our bodies. Their influence on health and disease is profound, and only beginning to be fully understood. Studying microbial populations is becoming easier with modern DNA sequencing technology; examining trends across thousands of samples today is common. The bottleneck is longer collecting data, but rather analyzing and interpreting the results.</p>
<p><code>rbiom</code> is an R package for working with abundance datasets, such as OTU or ASV counts from 16S amplicon sequencing. It enables importing/exporting all BIOM formats, subsetting, rarefying, manipulation of metadata/taxonomy/phylogeny, computation of alpha and beta diversity metrics, and summarizing counts per taxonomic rank. Computationally intensive tasks (including UniFrac [@unifrac]) have been implemented with multithreaded C++ to greatly reduce calculation time.</p>
<p>Visualization is a key component of <code>rbiom</code>. Rarefaction curves, taxa abundances, alpha diversity, and beta diversity can all be plotted in a variety of graphical formats, including correlation, heatmap, ordination, stacked bar, and box plots. In <code>rbiom</code>, box plots can be any combination of box, bar, violin, dot, strip, and/or range layers. Each plot includes provenance and modification history as attributes, as well as the <code>ggplot2</code> [@ggplot2] call used to render it to encourage downstream user customization.</p>
<p>Correlations between sample metadata and microbiome structure can be identified by mapping one or more metadata variables of interest to a plot’s axes, facets, and/or aesthetics. These mappings can optionally define color/shape/pattern assignments, category ordering, or subsetting parameters. When metadata is associated with a axis or aesthetic, <code>rbiom</code> will automatically run the appropriate statistical test, correct for multiple comparisons, and display significant differences on the plot, captioning it with a brief methodology.</p>
<p>Currently, <code>rbiom</code> can perform four types of significance testing. On correlation plots with a numeric metadata variable on the x-axis (e.g., Age, BMI), linear regression will be computed with R’s <code>lm</code> linear model function. For plots with two categories (e.g. Male vs Female), a Mann-Whitney test [@Mann1947] is run with R’s <code>wilcox.test</code>. When three or more categories are compared, the Kruskal-Wallis rank sum test [@Kruskal1952] is used instead via R’s <code>kruskal.test</code> function. P-values for ordinations are derived using the <code>adonis2</code> function from the <code>vegan</code> R package [@vegan], which randomly re-categorizes samples 1,000 times to estimate the significance of the observed clustering. P-values are corrected for multiple comparisons using the method described by @Benjamini1995 via R’s <code>p.adjust</code> function to control for the false discovery rate.</p>
<p><code>QIIME2</code> [@qiime2], <code>mothur</code> [@mothur], and <code>Phyloseq</code> [@phyloseq] offer overlapping functionality with rbiom, but with important distinctions. The first two are designed for command-line interaction, making them difficult to integrate into R projects. <code>Phyloseq</code> has been a staple of R bioinformatics for a decade, but is frustratingly slow for studies with thousands of samples.</p>
<p>This package is designed for users of all experience levels. Novice R users will appreciate that a couple commands will produce publication-ready figures. Advanced R users can use <code>rbiom</code> to complement their existing pipelines with faster and more flexible functions. <code>rbiom</code> is cross-platform compatible and available from CRAN and conda-forge. The latest development version is on GitHub.</p>
</div>
<div class="section level1">
<h1 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h1>
</div>


  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

