<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • rbiom</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"><meta property="og:image" content="https://cmmr.github.io/rbiom/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/cmmr/rbiom/blob/master/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="3.0.1" id="rbiom-301">rbiom 3.0.1<a class="anchor" aria-label="anchor" href="#rbiom-301"></a></h2>
<ul><li>Re-exports <code><a href="https://cmmr.github.io/ecodive/reference/metrics.html" class="external-link">ecodive::list_metrics</a></code>, <code><a href="https://cmmr.github.io/ecodive/reference/metrics.html" class="external-link">ecodive::match_metric</a></code>, and <code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggplot2::ggsave</a></code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="3.0.0" id="rbiom-300">rbiom 3.0.0<a class="anchor" aria-label="anchor" href="#rbiom-300"></a></h2>
<div class="section level3">
<h3 id="announcements-3-0-0">Announcements<a class="anchor" aria-label="anchor" href="#announcements-3-0-0"></a></h3>
<p>The code for <code>rbiom</code>’s very fast implementations of diversity metrics, rarefaction, and tree importing has been spun off into its own dependency-free package, <a href="https://github.com/cmmr/ecodive" class="external-link"><code>ecodive</code></a>, to encourage adoption by other R packages. This also means <code>rbiom</code> will be easier to install going forward as it will no longer need to be compiled.</p>
</div>
<div class="section level3">
<h3 id="breaking-changes-3-0-0">Breaking Changes<a class="anchor" aria-label="anchor" href="#breaking-changes-3-0-0"></a></h3>
<ul><li>
<code>weighted</code> and <code>normalized</code> arguments are no longer used by any functions.</li>
<li>Diversity metric names will now be all lowercase in returned matrices, tables, etc.</li>
</ul><p>The following functions have been deprecated since <code>rbiom</code> v2.0 and are now removed completely. They are generally functions which have been renamed for improved organization.</p>
<p><code>alpha.div</code>, <code>as.percent</code>, <code>beta.div</code>, <code>comments</code>, <code>comments&lt;-</code>, <code>counts</code>, <code>counts&lt;-</code>, <code>depth</code>, <code>depths_barplot</code>, <code>has.phylogeny</code>, <code>has.sequences</code>, <code>id</code>, <code>id&lt;-</code>, <code>info</code>, <code>is.rarefied</code>, <code>metadata</code>, <code>metadata&lt;-</code>, <code>nsamples</code>, <code>ntaxa</code>, <code>phylogeny</code>, <code>phylogeny&lt;-</code>, <code>rarefy</code>, <code>read.biom</code>, <code>read_biom</code>, <code>read.fasta</code>, <code>read.tree</code>, <code>repair</code>, <code>sample.names</code>, <code>sample.names&lt;-</code>, <code>sample_subset</code>, <code>sample.sums</code>, <code>sequences</code>, <code>sequences&lt;-</code>, <code>subtree</code>, <code>taxa.means</code>, <code>taxa.names</code>, <code>taxa.names&lt;-</code>, <code>taxa.ranks</code>, <code>taxa.ranks&lt;-</code>, <code>taxa.rollup</code>, <code>taxa.sums</code>, <code>taxonomy</code>, <code>taxonomy&lt;-</code>, <code>tips</code>, <code>top.taxa</code>, <code>top_taxa</code>, <code>unifrac</code>, <code>write.biom</code>, <code>write.fasta</code>, <code>write.tree</code>, <code>write.xlsx</code></p>
</div>
<div class="section level3">
<h3 id="additions-3-0-0">Additions<a class="anchor" aria-label="anchor" href="#additions-3-0-0"></a></h3>
<ul><li>
<code><a href="../reference/adiv_table.html">adiv_vector()</a></code> returns a simple named vector of alpha diversity values.</li>
<li>Replaced <code>rhdf5</code> dependency with <code>h5lite</code>.</li>
<li>Updated list of alpha diversity metrics: <code>c("ace", "berger", "brillouin", "chao1", "faith", "fisher", "simpson", "inv_simpson", "margalef", "mcintosh", "menhinick", "observed", "shannon", "squares")</code>
</li>
<li>Updated list of beta diversity metrics: <code>c("aitchison", "bhattacharyya", "bray", "canberra", "chebyshev", "chord", "clark", "sorensen", "divergence", "euclidean", "generalized_unifrac", "gower", "hamming", "hellinger", "horn", "jaccard", "jensen", "jsd", "lorentzian", "manhattan", "matusita", "minkowski", "morisita", "motyka", "normalized_unifrac", "ochiai", "psym_chisq", "soergel", "squared_chisq", "squared_chord", "squared_euclidean", "topsoe", "unweighted_unifrac", "variance_adjusted_unifrac", "wave_hedges", "weighted_unifrac")</code>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.2.1" id="rbiom-221">rbiom 2.2.1<a class="anchor" aria-label="anchor" href="#rbiom-221"></a></h2><p class="text-muted">CRAN release: 2025-06-27</p>
<ul><li>Compatibility fixes for ggplot2 4.0.0 <a href="https://github.com/cmmr/rbiom/issues/30" class="external-link">#30</a></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.2.0" id="rbiom-220">rbiom 2.2.0<a class="anchor" aria-label="anchor" href="#rbiom-220"></a></h2><p class="text-muted">CRAN release: 2025-04-04</p>
<ul><li>Satisfy new R CMD check: no setting attributes on built-in functions. <a href="https://github.com/cmmr/rbiom/issues/28" class="external-link">#28</a></li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.2" id="rbiom-212">rbiom 2.1.2<a class="anchor" aria-label="anchor" href="#rbiom-212"></a></h2><p class="text-muted">CRAN release: 2025-03-18</p>
<ul><li>Fix for <code>write_biom(format='json')</code> where <code>{...}</code> should be <code>[...]</code>.</li>
<li>Restore <code>rline</code> functionality in <code><a href="../reference/rare_corrplot.html">rare_corrplot()</a></code>.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.1.1" id="rbiom-211">rbiom 2.1.1<a class="anchor" aria-label="anchor" href="#rbiom-211"></a></h2><p class="text-muted">CRAN release: 2025-03-10</p>
<ul><li>
<code>unifrac()</code> no longer normalizes weighted values, for back-compatibility.</li>
<li>Option <code>underscores</code> allows keeping underscores in tree IDs.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="2.0.13" id="rbiom-2013">rbiom 2.0.13<a class="anchor" aria-label="anchor" href="#rbiom-2013"></a></h2><p class="text-muted">CRAN release: 2025-01-24</p>
<ul><li>Major release with significant new features.</li>
<li>Plotting added.</li>
<li>Statistics added.</li>
<li>Caching added.</li>
<li>Clearer naming scheme for functions. Previous names still work but are deprecated.</li>
<li>Generation of human-readable ggplot2 code for modifying plots outside of rbiom.</li>
<li>Provenance tracking for BIOM objects and derivatives.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.3" id="rbiom-103">rbiom 1.0.3<a class="anchor" aria-label="anchor" href="#rbiom-103"></a></h2><p class="text-muted">CRAN release: 2021-11-05</p>
<ul><li>Corrects for breaking changes in ‘rhdf5’ package.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.2" id="rbiom-102">rbiom 1.0.2<a class="anchor" aria-label="anchor" href="#rbiom-102"></a></h2><p class="text-muted">CRAN release: 2020-05-29</p>
<ul><li>Improved compatibility with Debian and Solaris.</li>
<li>The ‘rhdf5’ package is now an optional dependency.</li>
<li>The select() and subset() functions now subset sequences too.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.0" id="rbiom-100">rbiom 1.0.0<a class="anchor" aria-label="anchor" href="#rbiom-100"></a></h2><p class="text-muted">CRAN release: 2020-05-26</p>
<ul><li>Initial Release</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

