<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Test beta diversity for associations with metadata. — bdiv_stats • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Test beta diversity for associations with metadata. — bdiv_stats"><meta name="description" content="A convenience wrapper for bdiv_table() + stats_table()."><meta property="og:description" content="A convenience wrapper for bdiv_table() + stats_table()."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.0.9001</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Test beta diversity for associations with metadata.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/R/stats_table.r" class="external-link"><code>R/stats_table.r</code></a></small>
      <div class="d-none name"><code>bdiv_stats.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A convenience wrapper for <code><a href="bdiv_table.html">bdiv_table()</a></code> + <code><a href="stats_table.html">stats_table()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bdiv_stats</span><span class="op">(</span></span>
<span>  <span class="va">biom</span>,</span>
<span>  regr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stat.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bdiv <span class="op">=</span> <span class="st">"Bray-Curtis"</span>,</span>
<span>  weighted <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  tree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  within <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  between <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  split.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  transform <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  test <span class="op">=</span> <span class="st">"emmeans"</span>,</span>
<span>  fit <span class="op">=</span> <span class="st">"gam"</span>,</span>
<span>  at <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  level <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  alt <span class="op">=</span> <span class="st">"!="</span>,</span>
<span>  mu <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  p.adj <span class="op">=</span> <span class="st">"fdr"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-biom">biom<a class="anchor" aria-label="anchor" href="#arg-biom"></a></dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt id="arg-regr">regr<a class="anchor" aria-label="anchor" href="#arg-regr"></a></dt>
<dd><p>Dataset field with the x-axis (independent; predictive)
values. Must be numeric. Default: <code>NULL</code></p></dd>


<dt id="arg-stat-by">stat.by<a class="anchor" aria-label="anchor" href="#arg-stat-by"></a></dt>
<dd><p>Dataset field with the statistical groups. Must be
categorical. Default: <code>NULL</code></p></dd>


<dt id="arg-bdiv">bdiv<a class="anchor" aria-label="anchor" href="#arg-bdiv"></a></dt>
<dd><p>Beta diversity distance algorithm(s) to use. Options are:
<code>"Bray-Curtis"</code>, <code>"Manhattan"</code>, <code>"Euclidean"</code>,
<code>"Jaccard"</code>, and <code>"UniFrac"</code>. For <code>"UniFrac"</code>, a
phylogenetic tree must be present in <code>biom</code> or explicitly
provided via <code>tree=</code>. Multiple/abbreviated values allowed.
Default: <code>"Bray-Curtis"</code></p></dd>


<dt id="arg-weighted">weighted<a class="anchor" aria-label="anchor" href="#arg-weighted"></a></dt>
<dd><p>Take relative abundances into account. When
<code>weighted=FALSE</code>, only presence/absence is considered.
Multiple values allowed. Default: <code>TRUE</code></p></dd>


<dt id="arg-tree">tree<a class="anchor" aria-label="anchor" href="#arg-tree"></a></dt>
<dd><p>A <code>phylo</code> object representing the phylogenetic
relationships of the taxa in <code>biom</code>. Only required when
computing UniFrac distances. Default: <code>biom$tree</code></p></dd>


<dt id="arg-within-between">within, between<a class="anchor" aria-label="anchor" href="#arg-within-between"></a></dt>
<dd><p>Dataset field(s) for intra- or inter- sample
comparisons. Alternatively, dataset field names given elsewhere can
be prefixed with <code>'=='</code> or <code>'!='</code> to assign them to <code>within</code> or
<code>between</code>, respectively. Default: <code>NULL</code></p></dd>


<dt id="arg-split-by">split.by<a class="anchor" aria-label="anchor" href="#arg-split-by"></a></dt>
<dd><p>Dataset field(s) that the data should be split by prior to
any calculations. Must be categorical. Default: <code>NULL</code></p></dd>


<dt id="arg-transform">transform<a class="anchor" aria-label="anchor" href="#arg-transform"></a></dt>
<dd><p>Transformation to apply. Options are:
<code>c("none", "rank", "log", "log1p", "sqrt", "percent")</code>. <code>"rank"</code> is
useful for correcting for non-normally distributions before applying
regression statistics. Default: <code>"none"</code></p></dd>


<dt id="arg-test">test<a class="anchor" aria-label="anchor" href="#arg-test"></a></dt>
<dd><p>Method for computing p-values: <code>'wilcox'</code>, <code>'kruskal'</code>,
<code>'emmeans'</code>, or <code>'emtrends'</code>. Default: <code>'emmeans'</code></p></dd>


<dt id="arg-fit">fit<a class="anchor" aria-label="anchor" href="#arg-fit"></a></dt>
<dd><p>How to fit the trendline. <code>'lm'</code>, <code>'log'</code>, or <code>'gam'</code>.
Default: <code>'gam'</code></p></dd>


<dt id="arg-at">at<a class="anchor" aria-label="anchor" href="#arg-at"></a></dt>
<dd><p>Position(s) along the x-axis where the means or slopes should be
evaluated. Default: <code>NULL</code>, which samples 100 evenly spaced positions
and selects the position where the p-value is most significant.</p></dd>


<dt id="arg-level">level<a class="anchor" aria-label="anchor" href="#arg-level"></a></dt>
<dd><p>The confidence level for calculating a confidence interval.
Default: <code>0.95</code></p></dd>


<dt id="arg-alt">alt<a class="anchor" aria-label="anchor" href="#arg-alt"></a></dt>
<dd><p>Alternative hypothesis direction. Options are <code>'!='</code>
(two-sided; not equal to <code>mu</code>), <code>'&lt;'</code> (less than <code>mu</code>), or <code>'&gt;'</code>
(greater than <code>mu</code>). Default: <code>'!='</code></p></dd>


<dt id="arg-mu">mu<a class="anchor" aria-label="anchor" href="#arg-mu"></a></dt>
<dd><p>Reference value to test against. Default: <code>0</code></p></dd>


<dt id="arg-p-adj">p.adj<a class="anchor" aria-label="anchor" href="#arg-p-adj"></a></dt>
<dd><p>Method to use for multiple comparisons adjustment of
p-values. Run <code>p.adjust.methods</code> for a list of available
options. Default: <code>"fdr"</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A tibble data.frame with fields from the table below. This tibble
object provides the <code>$code</code> operator to print the R code used to generate
the statistics.</p><table class="table table"><tr><td><strong>Field</strong></td><td><strong>Description</strong></td></tr><tr><td>.mean</td><td>Estimated marginal mean. See <code><a href="https://rvlenth.github.io/emmeans/reference/emmeans.html" class="external-link">emmeans::emmeans()</a></code>.</td></tr><tr><td>.mean.diff</td><td>Difference in means.</td></tr><tr><td>.slope</td><td>Trendline slope. See <code><a href="https://rvlenth.github.io/emmeans/reference/emtrends.html" class="external-link">emmeans::emtrends()</a></code>.</td></tr><tr><td>.slope.diff</td><td>Difference in slopes.</td></tr><tr><td>.h1</td><td>Alternate hypothesis.</td></tr><tr><td>.p.val</td><td>Probability that null hypothesis is correct.</td></tr><tr><td>.adj.p</td><td><code>.p.val</code> after adjusting for multiple comparisons.</td></tr><tr><td>.effect.size</td><td>Effect size. See <code><a href="https://rvlenth.github.io/emmeans/reference/eff_size.html" class="external-link">emmeans::eff_size()</a></code>.</td></tr><tr><td>.lower</td><td>Confidence interval lower bound.</td></tr><tr><td>.upper</td><td>Confidence interval upper bound.</td></tr><tr><td>.se</td><td>Standard error.</td></tr><tr><td>.n</td><td>Number of samples.</td></tr><tr><td>.df</td><td>Degrees of freedom.</td></tr><tr><td>.stat</td><td>Wilcoxon or Kruskal-Wallis rank sum statistic.</td></tr><tr><td>.t.ratio</td><td><code>.mean</code> / <code>.se</code></td></tr><tr><td>.r.sqr</td><td>Percent of variation explained by the model.</td></tr><tr><td>.adj.r</td><td><code>.r.sqr</code>, taking degrees of freedom into account.</td></tr><tr><td>.aic</td><td>Akaike Information Criterion (predictive models).</td></tr><tr><td>.bic</td><td>Bayesian Information Criterion (descriptive models).</td></tr><tr><td>.loglik</td><td>Log-likelihood goodness-of-fit score.</td></tr><tr><td>.fit.p</td><td>P-value for observing this fit by chance.</td></tr></table></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other beta_diversity:
<code><a href="bdiv_boxplot.html">bdiv_boxplot</a>()</code>,
<code><a href="bdiv_clusters.html">bdiv_clusters</a>()</code>,
<code><a href="bdiv_corrplot.html">bdiv_corrplot</a>()</code>,
<code><a href="bdiv_heatmap.html">bdiv_heatmap</a>()</code>,
<code><a href="bdiv_ord_plot.html">bdiv_ord_plot</a>()</code>,
<code><a href="bdiv_ord_table.html">bdiv_ord_table</a>()</code>,
<code><a href="bdiv_table.html">bdiv_table</a>()</code>,
<code><a href="distmat_stats.html">distmat_stats</a>()</code></p>
<p>Other stats_tables:
<code><a href="adiv_stats.html">adiv_stats</a>()</code>,
<code><a href="distmat_stats.html">distmat_stats</a>()</code>,
<code><a href="stats_table.html">stats_table</a>()</code>,
<code><a href="taxa_stats.html">taxa_stats</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span></span></span>
<span class="r-in"><span>      </span></span>
<span class="r-in"><span>    <span class="fu">bdiv_stats</span><span class="op">(</span><span class="va">biom</span>, stat.by <span class="op">=</span> <span class="st">"Sex"</span>, bdiv <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bray"</span>, <span class="st">"unifrac"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">7</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Model:    gam(.distance ~ Sex, method = "REML")</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .bdiv       Sex                   .mean.diff .h1    .p.val .adj.p .effect.size</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> Bray-Curtis Female - Male            0.059<span style="text-decoration: underline;">9</span>  != 0  0.004<span style="text-decoration: underline;">02</span> 0.021<span style="text-decoration: underline;">9</span>      0.261  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> Bray-Curtis Male - Female vs Male   -<span style="color: #BB0000;">0.048</span><span style="color: #BB0000; text-decoration: underline;">9</span>  != 0  0.007<span style="text-decoration: underline;">30</span> 0.021<span style="text-decoration: underline;">9</span>     -<span style="color: #BB0000;">0.235</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> UniFrac     Female - Female vs M…   -<span style="color: #BB0000;">0.025</span><span style="color: #BB0000; text-decoration: underline;">9</span>  != 0  0.086<span style="text-decoration: underline;">6</span>  0.173      -<span style="color: #BB0000;">0.109</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> UniFrac     Male - Female vs Male   -<span style="color: #BB0000;">0.024</span><span style="color: #BB0000; text-decoration: underline;">9</span>  != 0  0.187   0.281      -<span style="color: #BB0000;">0.115</span>  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> Bray-Curtis Female - Female vs M…    0.011<span style="text-decoration: underline;">0</span>  != 0  0.430   0.515       0.050<span style="text-decoration: underline;">3</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> UniFrac     Female - Male           -<span style="color: #BB0000;">0.001</span><span style="color: #BB0000; text-decoration: underline;">05</span> != 0  0.963   0.963      -<span style="color: #BB0000;">0.004</span><span style="color: #BB0000; text-decoration: underline;">19</span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="subset.html">subset</a></span><span class="op">(</span><span class="va">biom</span>, <span class="va">`Body Site`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Saliva'</span>, <span class="st">'Stool'</span>, <span class="st">'Buccal mucosa'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu">bdiv_stats</span><span class="op">(</span><span class="va">biom</span>, stat.by <span class="op">=</span> <span class="st">"Body Site"</span>, split.by <span class="op">=</span> <span class="st">"==Sex"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Model:    gam(.distance ~ `Body Site`, method = "REML")</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 30 × 6</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    Sex    `Body Site`                         .mean.diff .h1     .p.val   .adj.p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Female Buccal mucosa - Buccal mucosa vs S…     -<span style="color: #BB0000;">0.791</span> != 0  1.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-35</span> 2.91<span style="color: #949494;">e</span><span style="color: #BB0000;">-34</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> Female Buccal mucosa - Saliva vs Stool         -<span style="color: #BB0000;">0.789</span> != 0  1.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-35</span> 2.91<span style="color: #949494;">e</span><span style="color: #BB0000;">-34</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Female Buccal mucosa - Buccal mucosa vs S…     -<span style="color: #BB0000;">0.596</span> != 0  5.30<span style="color: #949494;">e</span><span style="color: #BB0000;">-23</span> 5.30<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Male   Saliva - Saliva vs Stool                -<span style="color: #BB0000;">0.528</span> != 0  4.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span> 3.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Male   Saliva - Buccal mucosa vs Stool         -<span style="color: #BB0000;">0.526</span> != 0  5.30<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span> 3.18<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Female Buccal mucosa vs Saliva - Buccal m…     -<span style="color: #BB0000;">0.195</span> != 0  2.44<span style="color: #949494;">e</span><span style="color: #BB0000;">-20</span> 1.22<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Female Buccal mucosa vs Saliva - Saliva v…     -<span style="color: #BB0000;">0.193</span> != 0  3.80<span style="color: #949494;">e</span><span style="color: #BB0000;">-20</span> 1.63<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> Female Saliva - Buccal mucosa vs Stool         -<span style="color: #BB0000;">0.505</span> != 0  5.18<span style="color: #949494;">e</span><span style="color: #BB0000;">-20</span> 1.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Female Saliva - Saliva vs Stool                -<span style="color: #BB0000;">0.503</span> != 0  5.95<span style="color: #949494;">e</span><span style="color: #BB0000;">-20</span> 1.98<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Male   Buccal mucosa - Saliva vs Stool         -<span style="color: #BB0000;">0.531</span> != 0  9.76<span style="color: #949494;">e</span><span style="color: #BB0000;">-19</span> 2.93<span style="color: #949494;">e</span><span style="color: #BB0000;">-18</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 20 more rows</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

