<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index • rbiom</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"><meta property="og:image" content="https://cmmr.github.io/rbiom/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="input-output">Input / Output<a class="anchor" aria-label="anchor" href="#input-output"></a></h2>

      <div class="section-desc"><p>Get data into and out of R objects.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="as_rbiom.html">as_rbiom()</a></code>

        </dt>
        <dd>Convert a variety of data types to an rbiom object.</dd>

        <dt>

          <code><a href="read_biom.html">read_biom()</a></code>

        </dt>
        <dd>Parse counts, metadata, taxonomy, and phylogeny from a BIOM file.</dd>

        <dt>

          <code><a href="read_fasta.html">read_fasta()</a></code>

        </dt>
        <dd>Parse a fasta file into a named character vector.</dd>

        <dt>

          <code><a href="read_tree.html">read_tree()</a></code>

        </dt>
        <dd>Read a newick formatted phylogenetic tree.</dd>

        <dt>

          <code><a href="export.html">write_mothur()</a></code> <code><a href="export.html">write_qiime2()</a></code>

        </dt>
        <dd>Export data to QIIME 2 or mothur.</dd>

        <dt>

          <code><a href="write_biom.html">write_biom()</a></code> <code><a href="write_biom.html">write_metadata()</a></code> <code><a href="write_biom.html">write_counts()</a></code> <code><a href="write_biom.html">write_taxonomy()</a></code> <code><a href="write_biom.html">write_fasta()</a></code> <code><a href="write_biom.html">write_tree()</a></code> <code><a href="write_biom.html">write_xlsx()</a></code>

        </dt>
        <dd>Save an rbiom object to a file.</dd>

        <dt>

          <code><a href="convert_to.html">convert_to_SE()</a></code> <code><a href="convert_to.html">convert_to_TSE()</a></code> <code><a href="convert_to.html">convert_to_phyloseq()</a></code> <code><a href="convert_to.html">convert_to_animalcules()</a></code>

        </dt>
        <dd>Convert biom data to an external package class.</dd>
      </dl></div><div class="section level2">
      <h2 id="the-rbiom-object">The rbiom Object<a class="anchor" aria-label="anchor" href="#the-rbiom-object"></a></h2>

      <div class="section-desc"><p>The <a href="rbiom_objects.html">rbiom object</a> itself includes many methods, including <code>$counts</code>, <code>$metadata</code>, <code>$taxonomy</code>, $samples, $n_samples and more.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="rbiom_objects.html">rbiom_objects</a></code>

        </dt>
        <dd>Working with rbiom Objects.</dd>

        <dt>

          <code><a href="as.list.rbiom.html">as.list(<i>&lt;rbiom&gt;</i>)</a></code>

        </dt>
        <dd>Convert an rbiom object to a base R list.</dd>

        <dt>

          <code><a href="as.matrix.rbiom.html">as.matrix(<i>&lt;rbiom&gt;</i>)</a></code>

        </dt>
        <dd>Convert an rbiom object to a simple count matrix.</dd>
      </dl></div><div class="section level2">
      <h2 id="sample-metadata">Sample Metadata<a class="anchor" aria-label="anchor" href="#sample-metadata"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="pull.rbiom.html">pull(<i>&lt;rbiom&gt;</i>)</a></code>

        </dt>
        <dd>Map sample names to metadata field values.</dd>

        <dt>

          <code><a href="with.html">with(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="with.html">within(<i>&lt;rbiom&gt;</i>)</a></code>

        </dt>
        <dd>Evaluate expressions on metadata.</dd>

        <dt>

          <code><a href="modify_metadata.html">mutate(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="modify_metadata.html">rename(<i>&lt;rbiom&gt;</i>)</a></code>

        </dt>
        <dd>Create, modify, and delete metadata fields.</dd>

        <dt>

          <code><a href="glimpse.rbiom.html">glimpse(<i>&lt;rbiom&gt;</i>)</a></code>

        </dt>
        <dd>Get a glimpse of your metadata.</dd>
      </dl></div><div class="section level2">
      <h2 id="subsetting">Subsetting<a class="anchor" aria-label="anchor" href="#subsetting"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="subset.html">subset(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="subset.html">`[`(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="subset.html">na.omit(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="subset.html">subset_taxa()</a></code>

        </dt>
        <dd>Subset an rbiom object by sample names, OTU names, metadata, or taxonomy.</dd>

        <dt>

          <code><a href="slice_metadata.html">slice(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="slice_metadata.html">slice_head(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="slice_metadata.html">slice_tail(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="slice_metadata.html">slice_min(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="slice_metadata.html">slice_max(<i>&lt;rbiom&gt;</i>)</a></code> <code><a href="slice_metadata.html">slice_sample(<i>&lt;rbiom&gt;</i>)</a></code>

        </dt>
        <dd>Subset to a specific number of samples.</dd>
      </dl></div><div class="section level2">
      <h2 id="taxa-abundance">Taxa Abundance<a class="anchor" aria-label="anchor" href="#taxa-abundance"></a></h2>

      <div class="section-desc"><p>Map OTUs to higher order taxonomic ranks, and compare those abundances to metadata.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="taxa_boxplot.html">taxa_boxplot()</a></code>

        </dt>
        <dd>Visualize BIOM data with boxplots.</dd>

        <dt>

          <code><a href="taxa_clusters.html">taxa_clusters()</a></code>

        </dt>
        <dd>Cluster samples by taxa abundances k-means.</dd>

        <dt>

          <code><a href="taxa_corrplot.html">taxa_corrplot()</a></code>

        </dt>
        <dd>Visualize taxa abundance with scatterplots and trendlines.</dd>

        <dt>

          <code><a href="taxa_heatmap.html">taxa_heatmap()</a></code>

        </dt>
        <dd>Display taxa abundances as a heatmap.</dd>

        <dt>

          <code><a href="taxa_map.html">taxa_map()</a></code>

        </dt>
        <dd>Map OTUs names to taxa names at a given rank.</dd>

        <dt>

          <code><a href="taxa_matrix.html">taxa_table()</a></code> <code><a href="taxa_matrix.html">taxa_matrix()</a></code>

        </dt>
        <dd>Taxa abundances per sample.</dd>

        <dt>

          <code><a href="taxa_stacked.html">taxa_stacked()</a></code>

        </dt>
        <dd>Display taxa abundances as a stacked bar graph.</dd>

        <dt>

          <code><a href="taxa_stats.html">taxa_stats()</a></code>

        </dt>
        <dd>Test taxa abundances for associations with metadata.</dd>

        <dt>

          <code><a href="taxa_sums.html">taxa_sums()</a></code> <code><a href="taxa_sums.html">taxa_means()</a></code> <code><a href="taxa_sums.html">taxa_apply()</a></code>

        </dt>
        <dd>Get summary taxa abundances.</dd>
      </dl></div><div class="section level2">
      <h2 id="alpha-diversity">Alpha Diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a></h2>

      <div class="section-desc"><p>Examine the diversity of OTUs present in each individual sample, and how that diversity correlates with metadata.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="adiv_boxplot.html">adiv_boxplot()</a></code>

        </dt>
        <dd>Visualize alpha diversity with boxplots.</dd>

        <dt>

          <code><a href="adiv_corrplot.html">adiv_corrplot()</a></code>

        </dt>
        <dd>Visualize alpha diversity with scatterplots and trendlines.</dd>

        <dt>

          <code><a href="adiv_stats.html">adiv_stats()</a></code>

        </dt>
        <dd>Test alpha diversity for associations with metadata.</dd>

        <dt>

          <code><a href="adiv_table.html">adiv_table()</a></code> <code><a href="adiv_table.html">adiv_matrix()</a></code> <code><a href="adiv_table.html">adiv_vector()</a></code>

        </dt>
        <dd>Calculate the alpha diversity of each sample.</dd>

        <dt>

          <code><a href="sample_sums.html">sample_sums()</a></code> <code><a href="sample_sums.html">sample_apply()</a></code>

        </dt>
        <dd>Summarize the taxa observations in each sample.</dd>
      </dl></div><div class="section level2">
      <h2 id="beta-diversity">Beta Diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a></h2>

      <div class="section-desc"><p>See how similiar samples are to each other, and what metadata/taxa influence clustering.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="bdiv_boxplot.html">bdiv_boxplot()</a></code>

        </dt>
        <dd>Visualize BIOM data with boxplots.</dd>

        <dt>

          <code><a href="bdiv_clusters.html">bdiv_clusters()</a></code>

        </dt>
        <dd>Cluster samples by beta diversity k-means.</dd>

        <dt>

          <code><a href="bdiv_corrplot.html">bdiv_corrplot()</a></code>

        </dt>
        <dd>Visualize beta diversity with scatterplots and trendlines.</dd>

        <dt>

          <code><a href="bdiv_heatmap.html">bdiv_heatmap()</a></code>

        </dt>
        <dd>Display beta diversities in an all vs all grid.</dd>

        <dt>

          <code><a href="bdiv_ord_plot.html">bdiv_ord_plot()</a></code>

        </dt>
        <dd>Ordinate samples and taxa on a 2D plane based on beta diversity distances.</dd>

        <dt>

          <code><a href="bdiv_ord_table.html">bdiv_ord_table()</a></code>

        </dt>
        <dd>Calculate PCoA and other ordinations, including taxa biplots and statistics.</dd>

        <dt>

          <code><a href="bdiv_stats.html">bdiv_stats()</a></code>

        </dt>
        <dd>Test beta diversity for associations with metadata.</dd>

        <dt>

          <code><a href="bdiv_table.html">bdiv_table()</a></code> <code><a href="bdiv_table.html">bdiv_matrix()</a></code> <code><a href="bdiv_table.html">bdiv_distmat()</a></code>

        </dt>
        <dd>Distance / dissimilarity between samples.</dd>
      </dl></div><div class="section level2">
      <h2 id="rarefaction">Rarefaction<a class="anchor" aria-label="anchor" href="#rarefaction"></a></h2>

      <div class="section-desc"><p>Improve signal-to-noise in analyses by ensuring all samples have an equal number of observations.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="rare_corrplot.html">rare_corrplot()</a></code>

        </dt>
        <dd>Visualize rarefaction curves with scatterplots and trendlines.</dd>

        <dt>

          <code><a href="rare_multiplot.html">rare_multiplot()</a></code>

        </dt>
        <dd>Combines rare_corrplot and rare_stacked into a single figure.</dd>

        <dt>

          <code><a href="rare_stacked.html">rare_stacked()</a></code>

        </dt>
        <dd>Visualize the number of observations per sample.</dd>

        <dt>

          <code><a href="rarefy.html">rarefy()</a></code>

        </dt>
        <dd>Rarefy OTU counts.</dd>
      </dl></div><div class="section level2">
      <h2 id="low-level-functions">Low Level Functions<a class="anchor" aria-label="anchor" href="#low-level-functions"></a></h2>

      <div class="section-desc"><p>Most functions operate on rbiom objects. These let you use arbitrary data.frames, distance matrices, matrices, and phylo objects.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="distmat_ord_table.html">distmat_ord_table()</a></code>

        </dt>
        <dd>Run ordinations on a distance matrix.</dd>

        <dt>

          <code><a href="distmat_stats.html">distmat_stats()</a></code>

        </dt>
        <dd>Run statistics on a distance matrix vs a categorical or numeric variable.</dd>

        <dt>

          <code><a href="matrix_ops.html">mtx_rarefy()</a></code> <code><a href="matrix_ops.html">mtx_percent()</a></code> <code><a href="matrix_ops.html">mtx_rescale()</a></code> <code><a href="matrix_ops.html">rarefy_cols()</a></code> <code><a href="matrix_ops.html">rescale_rows()</a></code> <code><a href="matrix_ops.html">rescale_cols()</a></code>

        </dt>
        <dd>Transform a counts matrix.</dd>

        <dt>

          <code><a href="stats_boxplot.html">stats_boxplot()</a></code>

        </dt>
        <dd>Visualize categorical metadata effects on numeric values.</dd>

        <dt>

          <code><a href="stats_corrplot.html">stats_corrplot()</a></code>

        </dt>
        <dd>Visualize regression with scatterplots and trendlines.</dd>

        <dt>

          <code><a href="stats_table.html">stats_table()</a></code>

        </dt>
        <dd>Run non-parametric statistics on a data.frame.</dd>

        <dt>

          <code><a href="tree_subset.html">tree_subset()</a></code>

        </dt>
        <dd>Create a subtree by specifying tips to keep.</dd>

        <dt>

          <code><a href="plot_heatmap.html">plot_heatmap()</a></code>

        </dt>
        <dd>Create a heatmap with tracks and dendrograms from any matrix.</dd>
      </dl></div><div class="section level2">
      <h2 id="advanced-operations">Advanced Operations<a class="anchor" aria-label="anchor" href="#advanced-operations"></a></h2>



      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="bdply.html">bdply()</a></code> <code><a href="bdply.html">blply()</a></code>

        </dt>
        <dd>Apply a function to each subset of an rbiom object.</dd>

        <dt>

          <code><a href="biom_merge.html">biom_merge()</a></code>

        </dt>
        <dd>Combine several rbiom objects into one.</dd>
      </dl></div><div class="section level2">
      <h2 id="datasets">Datasets<a class="anchor" aria-label="anchor" href="#datasets"></a></h2>

      <div class="section-desc"><p>Example datasets included with rbiom.</p></div>

      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="hmp50.html">hmp50</a></code>

        </dt>
        <dd>Human Microbiome Project - demo dataset (n = 50)</dd>

        <dt>

          <code><a href="gems.html">gems</a></code>

        </dt>
        <dd>Global Enteric Multicenter Study (n = 1,006)</dd>

        <dt>

          <code><a href="babies.html">babies</a></code>

        </dt>
        <dd>Longitudinal Stool Samples from Infants (n = 2,684)</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

