<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>rbiom: Read/Write, Transform, and Summarize BIOM Data — rbiom-package • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="rbiom: Read/Write, Transform, and Summarize BIOM Data — rbiom-package"><meta name="description" content="A toolkit for working with Biological Observation Matrix (BIOM) files.
Features include reading/writing all BIOM formats, rarefaction, alpha
diversity, beta diversity (including UniFrac), summarizing counts by
taxonomic level, and sample subsetting. Standalone functions for
reading, writing, and subsetting phylogenetic trees are also provided.
All CPU intensive operations are encoded in C with multi-thread support."><meta property="og:description" content="A toolkit for working with Biological Observation Matrix (BIOM) files.
Features include reading/writing all BIOM formats, rarefaction, alpha
diversity, beta diversity (including UniFrac), summarizing counts by
taxonomic level, and sample subsetting. Standalone functions for
reading, writing, and subsetting phylogenetic trees are also provided.
All CPU intensive operations are encoded in C with multi-thread support."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.13.9005</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>rbiom: Read/Write, Transform, and Summarize BIOM Data</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/R/rbiom-package.R" class="external-link"><code>R/rbiom-package.R</code></a></small>
      <div class="d-none name"><code>rbiom-package.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A toolkit for working with Biological Observation Matrix (BIOM) files.
Features include reading/writing all BIOM formats, rarefaction, alpha
diversity, beta diversity (including UniFrac), summarizing counts by
taxonomic level, and sample subsetting. Standalone functions for
reading, writing, and subsetting phylogenetic trees are also provided.
All CPU intensive operations are encoded in C with multi-thread support.</p>
    </div>


    <div class="section level2">
    <h2 id="multithreading">Multithreading<a class="anchor" aria-label="anchor" href="#multithreading"></a></h2>


<p>Many rbiom functions support multithreading:</p>
<p>The default behavior of these function is to run on as many cores as are
available in the local compute environment. If you wish to limit the number
of simultaneous threads, set <code>RcppParallel</code>'s <code>numThreads</code> option.
For instance:</p>
<div class="sourceCode"><pre><code><span></span>
<span>    <span class="fu">RcppParallel</span><span class="fu">::</span><span class="fu">setThreadOptions</span><span class="op">(</span>numThreads <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span></code></pre></div>

    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://cmmr.github.io/rbiom/">https://cmmr.github.io/rbiom/</a></p></li>
<li><p><a href="https://github.com/cmmr/rbiom" class="external-link">https://github.com/cmmr/rbiom</a></p></li>
<li><p>Report bugs at <a href="https://github.com/cmmr/rbiom/issues" class="external-link">https://github.com/cmmr/rbiom/issues</a></p></li>
</ul></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p><strong>Maintainer</strong>: Daniel P. Smith <a href="mailto:dansmith01@gmail.com">dansmith01@gmail.com</a> (<a href="https://orcid.org/0000-0002-2479-2044" class="external-link">ORCID</a>)</p>
<p>Other contributors:</p><ul><li><p>Alkek Center for Metagenomics and Microbiome Research [copyright holder, funder]</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

