<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Speed Ups. — speed • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Speed Ups. — speed"><meta name="description" content="When working with very large datasets, you can make use of these tips and
tricks to speed up operations on rbiom objects."><meta property="og:description" content="When working with very large datasets, you can make use of these tips and
tricks to speed up operations on rbiom objects."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Speed Ups.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/R/speed_ups.r" class="external-link"><code>R/speed_ups.r</code></a></small>
      <div class="d-none name"><code>speed.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>When working with very large datasets, you can make use of these tips and
tricks to speed up operations on rbiom objects.</p>
    </div>


    <div class="section level2">
    <h2 id="skip-cloning">Skip Cloning<a class="anchor" aria-label="anchor" href="#skip-cloning"></a></h2>



<p>Functions that modify rbiom objects, like <code><a href="subset.html">subset()</a></code> and <code><a href="rarefy.html">rarefy()</a></code>, will
automatically clone the object before modifying it. This is to make these
functions behave as most R users would expect - but at a performance trade
off.</p>
<p>Rather than:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">biom</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span></span></code></pre><p></p></div>
<p>Modify <code>biom</code> in place like this:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">biom</span>, clone <span class="op">=</span> <span class="cn">FALSE</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">biom</span>, clone <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or:</span></span>
<span><span class="va">biom</span><span class="op">$</span><span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">...</span><span class="op">)</span></span>
<span><span class="va">biom</span><span class="op">$</span><span class="va">counts</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu"><a href="../reference/rarefy_cols.html">rarefy_cols</a></span><span class="op">(</span><span class="op">)</span></span></code></pre><p></p></div>
    </div>
    <div class="section level2">
    <h2 id="drop-components">Drop Components<a class="anchor" aria-label="anchor" href="#drop-components"></a></h2>


<div class="section">
<h3 id="sequences">Sequences<a class="anchor" aria-label="anchor" href="#sequences"></a></h3>


<p>Reference sequences for OTUs will be imported along with the rest of your
dataset and stored in <code>$sequences</code>. However, rbiom doesn't currently use
these sequences for anything (except writing them back out with
<code><a href="write_biom.html">write_biom()</a></code> or <code><a href="write_biom.html">write_fasta()</a></code>).</p>
<p>You can delete them from your rbiom object with:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="va">biom</span><span class="op">$</span><span class="va">sequences</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre><p></p></div>
</div>

<div class="section">
<h3 id="tree">Tree<a class="anchor" aria-label="anchor" href="#tree"></a></h3>


<p>The phylogenetic reference tree for OTUs is only used for calculating
UniFrac distances. If you aren't using UniFrac, the tree can be dropped
from the rbiom object with:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="va">biom</span><span class="op">$</span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span></code></pre><p></p></div>
<p>Alternatively, you can store the tree separately from the rbiom object and
provide it to just the functions that use it. For example:</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="va">biom</span><span class="op">$</span><span class="va">tree</span></span>
<span><span class="va">biom</span><span class="op">$</span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdiv_table.html">bdiv_distmat</a></span><span class="op">(</span><span class="va">biom</span>, <span class="st">'unifrac'</span>, tree <span class="op">=</span> <span class="va">tree</span><span class="op">)</span></span></code></pre><p></p></div>
</div>

    </div>
    <div class="section level2">
    <h2 id="increase-caching">Increase Caching<a class="anchor" aria-label="anchor" href="#increase-caching"></a></h2>



<p>Caching is enabled by default - up to 20 MB per R session.</p>
<p>For large datasets, increasing the cache size can help. The size is
specified in bytes by an R option or environment variable.</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>rbiom.cache_size<span class="op">=</span><span class="fl">200</span> <span class="op">*</span> <span class="fl">1024</span> <span class="op">^</span> <span class="fl">2</span><span class="op">)</span> <span class="co"># 200 MB</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>RBIOM_CACHE_SIZE<span class="op">=</span><span class="fl">1024</span> <span class="op">^</span> <span class="fl">3</span><span class="op">)</span>    <span class="co"># 1 GB</span></span></code></pre><p></p></div>
<p>You can also specify a cache directory where results can be preserved from
one R session to the next.</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>rbiom.cache_dir<span class="op">=</span><span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/userdir.html" class="external-link">R_user_dir</a></span><span class="op">(</span><span class="st">"rbiom"</span>, <span class="st">"cache"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html" class="external-link">Sys.setenv</a></span><span class="op">(</span>RBIOM_CACHE_DIR<span class="op">=</span><span class="st">"~/rbiom_cache"</span><span class="op">)</span></span></code></pre><p></p></div>
<p>Other quick notes about caching:</p><ul><li><p>Setting the cache directory to <code>"FALSE"</code> will disable caching.</p></li>
<li><p>R options will override environment variables.</p></li>
<li><p>The key hash algorithm can be set with <code>options(rbiom.cache_hash=rlang::hash)</code>.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="summary-layers">Summary Layers<a class="anchor" aria-label="anchor" href="#summary-layers"></a></h2>



<p>The figure-generating functions allow you to display every data point.
However, when you have thousands of data points, rendering every single one
can be slow. Instead, set the <code>layers</code> parameter to use other options.</p>
<p></p><div class="sourceCode r"><pre><code><span><span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, layers <span class="op">=</span> <span class="st">"bl"</span><span class="op">)</span>  <span class="co"># bar, linerange</span></span>
<span><span class="fu"><a href="../reference/adiv_corrplot.html">adiv_corrplot</a></span><span class="op">(</span><span class="va">biom</span>, layers <span class="op">=</span> <span class="st">"tc"</span><span class="op">)</span> <span class="co"># trend, confidence</span></span>
<span><span class="fu"><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot</a></span><span class="op">(</span><span class="va">biom</span>, layers <span class="op">=</span> <span class="st">"e"</span><span class="op">)</span>  <span class="co"># ellipse</span></span></code></pre><p></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

