<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Visualize regression with scatterplots and trendlines. — stats_corrplot • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Visualize regression with scatterplots and trendlines. — stats_corrplot"><meta name="description" content="Visualize regression with scatterplots and trendlines."><meta property="og:description" content="Visualize regression with scatterplots and trendlines."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.13.9004</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visualize regression with scatterplots and trendlines.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/R/corrplot.r" class="external-link"><code>R/corrplot.r</code></a></small>
      <div class="d-none name"><code>stats_corrplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Visualize regression with scatterplots and trendlines.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">stats_corrplot</span><span class="op">(</span></span>
<span>  <span class="va">df</span>,</span>
<span>  <span class="va">x</span>,</span>
<span>  y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">df</span>, <span class="st">"response"</span><span class="op">)</span>,</span>
<span>  layers <span class="op">=</span> <span class="st">"tc"</span>,</span>
<span>  stat.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  facet.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  shapes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  test <span class="op">=</span> <span class="st">"emmeans"</span>,</span>
<span>  fit <span class="op">=</span> <span class="st">"gam"</span>,</span>
<span>  at <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  level <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  p.adj <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  p.top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  alt <span class="op">=</span> <span class="st">"!="</span>,</span>
<span>  mu <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  caption <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-df">df<a class="anchor" aria-label="anchor" href="#arg-df"></a></dt>
<dd><p>The dataset (data.frame or tibble object). "Dataset fields"
mentioned below should match column names in <code>df</code>. Required.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Dataset field with the x-axis values. Equivalent to the <code>regr</code>
argument in <code><a href="stats_table.html">stats_table()</a></code>. Required.</p></dd>


<dt id="arg-y">y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>A numeric metadata column name to use for the y-axis.
Default: <code>attr(df, 'response')</code></p></dd>


<dt id="arg-layers">layers<a class="anchor" aria-label="anchor" href="#arg-layers"></a></dt>
<dd><p>One or more of
<code>c("trend", "confidence", "point", "name", "residual")</code>. Single
letter abbreviations are also accepted. For instance,
<code>c("trend", "point")</code> is equivalent to <code>c("t", "p")</code> and <code>"tp"</code>.
Default: <code>"tc"</code></p></dd>


<dt id="arg-stat-by">stat.by<a class="anchor" aria-label="anchor" href="#arg-stat-by"></a></dt>
<dd><p>Dataset field with the statistical groups. Must be
categorical. Default: <code>NULL</code></p></dd>


<dt id="arg-facet-by">facet.by<a class="anchor" aria-label="anchor" href="#arg-facet-by"></a></dt>
<dd><p>Dataset field(s) to use for faceting. Must be categorical.
Default: <code>NULL</code></p></dd>


<dt id="arg-colors">colors<a class="anchor" aria-label="anchor" href="#arg-colors"></a></dt>
<dd><p>How to color the groups. Options are:</p><dl><dt><code>TRUE</code> - </dt>
<dd><p>Automatically select colorblind-friendly colors.</p></dd>

<dt><code>FALSE</code> or <code>NULL</code> - </dt>
<dd><p>Don't use colors.</p></dd>

<dt>a palette name - </dt>
<dd><p>Auto-select colors from this set. E.g. <code>"okabe"</code></p></dd>

<dt>character vector - </dt>
<dd><p>Custom colors to use. E.g. <code>c("red", "#00FF00")</code></p></dd>

<dt>named character vector - </dt>
<dd><p>Explicit mapping. E.g. <code>c(Male = "blue", Female = "red")</code></p></dd>


</dl><p>See "Aesthetics" section below for additional information.
Default: <code>TRUE</code></p></dd>


<dt id="arg-shapes">shapes<a class="anchor" aria-label="anchor" href="#arg-shapes"></a></dt>
<dd><p>Shapes for each group.
Options are similar to <code>colors</code>'s: <code>TRUE</code>, <code>FALSE</code>, <code>NULL</code>, shape
names (typically integers 0 - 17), or a named vector mapping
groups to specific shape names.
See "Aesthetics" section below for additional information.
Default: <code>TRUE</code></p></dd>


<dt id="arg-test">test<a class="anchor" aria-label="anchor" href="#arg-test"></a></dt>
<dd><p>Method for computing p-values: <code>'none'</code>, <code>'emmeans'</code>, or
<code>'emtrends'</code>. Default: <code>'emmeans'</code></p></dd>


<dt id="arg-fit">fit<a class="anchor" aria-label="anchor" href="#arg-fit"></a></dt>
<dd><p>How to fit the trendline. <code>'lm'</code>, <code>'log'</code>, or <code>'gam'</code>.
Default: <code>'gam'</code></p></dd>


<dt id="arg-at">at<a class="anchor" aria-label="anchor" href="#arg-at"></a></dt>
<dd><p>Position(s) along the x-axis where the means or slopes should be
evaluated. Default: <code>NULL</code>, which samples 100 evenly spaced positions
and selects the position where the p-value is most significant.</p></dd>


<dt id="arg-level">level<a class="anchor" aria-label="anchor" href="#arg-level"></a></dt>
<dd><p>The confidence level for calculating a confidence interval.
Default: <code>0.95</code></p></dd>


<dt id="arg-p-adj">p.adj<a class="anchor" aria-label="anchor" href="#arg-p-adj"></a></dt>
<dd><p>Method to use for multiple comparisons adjustment of
p-values. Run <code>p.adjust.methods</code> for a list of available
options. Default: <code>"fdr"</code></p></dd>


<dt id="arg-p-top">p.top<a class="anchor" aria-label="anchor" href="#arg-p-top"></a></dt>
<dd><p>Only display taxa with the most significant differences in
abundance. If <code>p.top</code> is &gt;= 1, then the <code>p.top</code> most
significant taxa are displayed. If <code>p.top</code> is less than one, all
taxa with an adjusted p-value &lt;= <code>p.top</code> are displayed.
Recommended to be used in combination with the <code>taxa</code> parameter
to set a lower bound on the mean abundance of considered taxa.
Default: <code>Inf</code></p></dd>


<dt id="arg-alt">alt<a class="anchor" aria-label="anchor" href="#arg-alt"></a></dt>
<dd><p>Alternative hypothesis direction. Options are <code>'!='</code>
(two-sided; not equal to <code>mu</code>), <code>'&lt;'</code> (less than <code>mu</code>), or <code>'&gt;'</code>
(greater than <code>mu</code>). Default: <code>'!='</code></p></dd>


<dt id="arg-mu">mu<a class="anchor" aria-label="anchor" href="#arg-mu"></a></dt>
<dd><p>Reference value to test against. Default: <code>0</code></p></dd>


<dt id="arg-caption">caption<a class="anchor" aria-label="anchor" href="#arg-caption"></a></dt>
<dd><p>Add methodology caption beneath the plot.
Default: <code>TRUE</code></p></dd>


<dt id="arg-check">check<a class="anchor" aria-label="anchor" href="#arg-check"></a></dt>
<dd><p>Generate additional plots to aid in assessing data normality.
Default: <code>FALSE</code></p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional parameters to pass along to ggplot2 functions.
Prefix a parameter name with a layer name to pass it to only that
layer. For instance, <code>p.size = 2</code> ensures only the points have their
size set to <code>2</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>ggplot2</code> plot. The computed data points, ggplot2 command,
stats table, and stats table commands are available as <code>$data</code>,
<code>$code</code>, <code>$stats</code>, and <code>$stats$code</code>, respectively.</p>
    </div>
    <div class="section level2">
    <h2 id="aesthetics">Aesthetics<a class="anchor" aria-label="anchor" href="#aesthetics"></a></h2>



<p>All built-in color palettes are colorblind-friendly. The available
categorical palette names are: <code>"okabe"</code>, <code>"carto"</code>, <code>"r4"</code>,
<code>"polychrome"</code>, <code>"tol"</code>, <code>"bright"</code>, <code>"light"</code>,
<code>"muted"</code>, <code>"vibrant"</code>, <code>"tableau"</code>, <code>"classic"</code>,
<code>"alphabet"</code>, <code>"tableau20"</code>, <code>"kelly"</code>, and <code>"fishy"</code>.</p>
<p>Shapes can be given as per base R - numbers 0 through 17 for various shapes,
or the decimal value of an ascii character, e.g. a-z = 65:90; A-Z = 97:122 to use
letters instead of shapes on the plot. Character strings may used as well.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other visualization:
<code><a href="adiv_boxplot.html">adiv_boxplot</a>()</code>,
<code><a href="adiv_corrplot.html">adiv_corrplot</a>()</code>,
<code><a href="bdiv_boxplot.html">bdiv_boxplot</a>()</code>,
<code><a href="bdiv_corrplot.html">bdiv_corrplot</a>()</code>,
<code><a href="bdiv_heatmap.html">bdiv_heatmap</a>()</code>,
<code><a href="bdiv_ord_plot.html">bdiv_ord_plot</a>()</code>,
<code><a href="plot_heatmap.html">plot_heatmap</a>()</code>,
<code><a href="rare_corrplot.html">rare_corrplot</a>()</code>,
<code><a href="rare_multiplot.html">rare_multiplot</a>()</code>,
<code><a href="rare_stacked.html">rare_stacked</a>()</code>,
<code><a href="stats_boxplot.html">stats_boxplot</a>()</code>,
<code><a href="taxa_boxplot.html">taxa_boxplot</a>()</code>,
<code><a href="taxa_corrplot.html">taxa_corrplot</a>()</code>,
<code><a href="taxa_heatmap.html">taxa_heatmap</a>()</code>,
<code><a href="taxa_stacked.html">taxa_stacked</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="subset.html">subset</a></span><span class="op">(</span><span class="va">hmp50</span>, <span class="va">`Body Site`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Saliva'</span>, <span class="st">'Stool'</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="va">df</span>   <span class="op">&lt;-</span> <span class="fu"><a href="adiv_table.html">adiv_table</a></span><span class="op">(</span><span class="fu"><a href="rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu">stats_corrplot</span><span class="op">(</span><span class="va">df</span>, <span class="st">"age"</span>, stat.by <span class="op">=</span> <span class="st">"body"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="stats_corrplot-1.png" alt="" width="768" height="384"></span>
<span class="r-in"><span>    <span class="fu">stats_corrplot</span><span class="op">(</span></span></span>
<span class="r-in"><span>      df       <span class="op">=</span> <span class="va">df</span>, </span></span>
<span class="r-in"><span>      x        <span class="op">=</span> <span class="st">"Age"</span>, </span></span>
<span class="r-in"><span>      stat.by  <span class="op">=</span> <span class="st">"Body Site"</span>, </span></span>
<span class="r-in"><span>      facet.by <span class="op">=</span> <span class="st">"Sex"</span>, </span></span>
<span class="r-in"><span>      layers   <span class="op">=</span> <span class="st">"trend"</span> <span class="op">)</span></span></span>
<span class="r-plt img"><img src="stats_corrplot-2.png" alt="" width="768" height="384"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

