<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Visualize BIOM data with boxplots. — taxa_boxplot • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Visualize BIOM data with boxplots. — taxa_boxplot"><meta name="description" content="Visualize BIOM data with boxplots."><meta property="og:description" content="Visualize BIOM data with boxplots."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.13</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visualize BIOM data with boxplots.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/v2.0.13/R/boxplot.r" class="external-link"><code>R/boxplot.r</code></a></small>
      <div class="d-none name"><code>taxa_boxplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Visualize BIOM data with boxplots.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">taxa_boxplot</span><span class="op">(</span></span>
<span>  <span class="va">biom</span>,</span>
<span>  x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rank <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  layers <span class="op">=</span> <span class="st">"x"</span>,</span>
<span>  taxa <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  unc <span class="op">=</span> <span class="st">"singly"</span>,</span>
<span>  other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  p.top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  stat.by <span class="op">=</span> <span class="va">x</span>,</span>
<span>  facet.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  colors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  shapes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  patterns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  flip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  stripe <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ci <span class="op">=</span> <span class="st">"ci"</span>,</span>
<span>  level <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  p.adj <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  outliers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xlab.angle <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  p.label <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  transform <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  y.transform <span class="op">=</span> <span class="st">"sqrt"</span>,</span>
<span>  caption <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-biom">biom<a class="anchor" aria-label="anchor" href="#arg-biom"></a></dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A categorical metadata column name to use for the x-axis. Or
<code>NULL</code>, which puts taxa along the x-axis. Default: <code>NULL</code></p></dd>


<dt id="arg-rank">rank<a class="anchor" aria-label="anchor" href="#arg-rank"></a></dt>
<dd><p>What rank(s) of taxa to display. E.g. <code>"Phylum"</code>,
<code>"Genus"</code>, <code>".otu"</code>, etc. An integer vector can also be
given, where <code>1</code> is the highest rank, <code>2</code> is the second
highest, <code>-1</code> is the lowest rank, <code>-2</code> is the second
lowest, and <code>0</code> is the OTU "rank". Run <code>biom$ranks</code> to
see all options for a given rbiom object. Default: <code>-1</code>.</p></dd>


<dt id="arg-layers">layers<a class="anchor" aria-label="anchor" href="#arg-layers"></a></dt>
<dd><p>One or more of
<code>c("bar", "box" ("x"), "violin", "dot", "strip", "crossbar", "errorbar", "linerange", "pointrange")</code>.
Single letter abbreviations are also accepted. For instance,
<code>c("box", "dot")</code> is equivalent to <code>c("x", "d")</code> and <code>"xd"</code>.
Default: <code>"x"</code></p></dd>


<dt id="arg-taxa">taxa<a class="anchor" aria-label="anchor" href="#arg-taxa"></a></dt>
<dd><p>Which taxa to display. An integer value will show the top n
most abundant taxa. A value 0 &lt;= n &lt; 1 will show any taxa with that
mean abundance or greater (e.g. <code>0.1</code> implies &gt;= 10%). A
character vector of taxa names will show only those named taxa.
Default: <code>6</code>.</p></dd>


<dt id="arg-unc">unc<a class="anchor" aria-label="anchor" href="#arg-unc"></a></dt>
<dd><p>How to handle unclassified, uncultured, and similarly ambiguous
taxa names. Options are:</p><dl><dt><code>"singly"</code> - </dt>
<dd><p>Replaces them with the OTU name.</p></dd>

<dt><code>"grouped"</code> - </dt>
<dd><p>Replaces them with a higher rank's name.</p></dd>

<dt><code>"drop"</code> - </dt>
<dd><p>Excludes them from the result.</p></dd>

<dt><code>"asis"</code> - </dt>
<dd><p>To not check/modify any taxa names.</p></dd>


</dl><p>Abbreviations are allowed. Default: <code>"singly"</code></p></dd>


<dt id="arg-other">other<a class="anchor" aria-label="anchor" href="#arg-other"></a></dt>
<dd><p>Sum all non-itemized taxa into an "Other" taxa. When
<code>FALSE</code>, only returns taxa matched by the <code>taxa</code>
argument. Specifying <code>TRUE</code> adds "Other" to the returned set.
A string can also be given to imply <code>TRUE</code>, but with that
value as the name to use instead of "Other".
Default: <code>FALSE</code></p></dd>


<dt id="arg-p-top">p.top<a class="anchor" aria-label="anchor" href="#arg-p-top"></a></dt>
<dd><p>Only display taxa with the most significant differences in
abundance. If <code>p.top</code> is &gt;= 1, then the <code>p.top</code> most
significant taxa are displayed. If <code>p.top</code> is less than one, all
taxa with an adjusted p-value &lt;= <code>p.top</code> are displayed.
Recommended to be used in combination with the <code>taxa</code> parameter
to set a lower bound on the mean abundance of considered taxa.
Default: <code>Inf</code></p></dd>


<dt id="arg-stat-by">stat.by<a class="anchor" aria-label="anchor" href="#arg-stat-by"></a></dt>
<dd><p>Dataset field with the statistical groups. Must be
categorical. Default: <code>NULL</code></p></dd>


<dt id="arg-facet-by">facet.by<a class="anchor" aria-label="anchor" href="#arg-facet-by"></a></dt>
<dd><p>Dataset field(s) to use for faceting. Must be categorical.
Default: <code>NULL</code></p></dd>


<dt id="arg-colors">colors<a class="anchor" aria-label="anchor" href="#arg-colors"></a></dt>
<dd><p>How to color the groups. Options are:</p><dl><dt><code>TRUE</code> - </dt>
<dd><p>Automatically select colorblind-friendly colors.</p></dd>

<dt><code>FALSE</code> or <code>NULL</code> - </dt>
<dd><p>Don't use colors.</p></dd>

<dt>a palette name - </dt>
<dd><p>Auto-select colors from this set. E.g. <code>"okabe"</code></p></dd>

<dt>character vector - </dt>
<dd><p>Custom colors to use. E.g. <code>c("red", "#00FF00")</code></p></dd>

<dt>named character vector - </dt>
<dd><p>Explicit mapping. E.g. <code>c(Male = "blue", Female = "red")</code></p></dd>


</dl><p>See "Aesthetics" section below for additional information.
Default: <code>TRUE</code></p></dd>


<dt id="arg-shapes">shapes<a class="anchor" aria-label="anchor" href="#arg-shapes"></a></dt>
<dd><p>Shapes for each group.
Options are similar to <code>colors</code>'s: <code>TRUE</code>, <code>FALSE</code>, <code>NULL</code>, shape
names (typically integers 0 - 17), or a named vector mapping
groups to specific shape names.
See "Aesthetics" section below for additional information.
Default: <code>TRUE</code></p></dd>


<dt id="arg-patterns">patterns<a class="anchor" aria-label="anchor" href="#arg-patterns"></a></dt>
<dd><p>Patterns for each group.
Options are similar to <code>colors</code>'s: <code>TRUE</code>, <code>FALSE</code>, <code>NULL</code>, pattern
names (<code>"brick"</code>, <code>"chevron"</code>, <code>"fish"</code>, <code>"grid"</code>, etc), or a named
vector mapping groups to specific pattern names.
See "Aesthetics" section below for additional information.
Default: <code>FALSE</code></p></dd>


<dt id="arg-flip">flip<a class="anchor" aria-label="anchor" href="#arg-flip"></a></dt>
<dd><p>Transpose the axes, so that taxa are present as rows instead
of columns. Default: <code>FALSE</code></p></dd>


<dt id="arg-stripe">stripe<a class="anchor" aria-label="anchor" href="#arg-stripe"></a></dt>
<dd><p>Shade every other x position. Default: <em>same as flip</em></p></dd>


<dt id="arg-ci">ci<a class="anchor" aria-label="anchor" href="#arg-ci"></a></dt>
<dd><p>How to calculate min/max of the <b>crossbar</b>,
<b>errorbar</b>, <b>linerange</b>, and <b>pointrange</b> layers.
Options are: <code>"ci"</code> (confidence interval), <code>"range"</code>,
<code>"sd"</code> (standard deviation), <code>"se"</code> (standard error), and
<code>"mad"</code> (median absolute deviation).
The center mark of <b>crossbar</b> and <b>pointrange</b> represents
the mean, except for <code>"mad"</code> in which case it represents the median.
Default: <code>"ci"</code></p></dd>


<dt id="arg-level">level<a class="anchor" aria-label="anchor" href="#arg-level"></a></dt>
<dd><p>The confidence level for calculating a confidence interval.
Default: <code>0.95</code></p></dd>


<dt id="arg-p-adj">p.adj<a class="anchor" aria-label="anchor" href="#arg-p-adj"></a></dt>
<dd><p>Method to use for multiple comparisons adjustment of
p-values. Run <code>p.adjust.methods</code> for a list of available
options. Default: <code>"fdr"</code></p></dd>


<dt id="arg-outliers">outliers<a class="anchor" aria-label="anchor" href="#arg-outliers"></a></dt>
<dd><p>Show boxplot outliers? <code>TRUE</code> to always show.
<code>FALSE</code> to always hide. <code>NULL</code> to only hide them when
overlaying a dot or strip chart.  Default: <code>NULL</code></p></dd>


<dt id="arg-xlab-angle">xlab.angle<a class="anchor" aria-label="anchor" href="#arg-xlab-angle"></a></dt>
<dd><p>Angle of the labels at the bottom of the plot.
Options are <code>"auto"</code>, <code>'0'</code>, <code>'30'</code>, and <code>'90'</code>.
Default: <code>"auto"</code>.</p></dd>


<dt id="arg-p-label">p.label<a class="anchor" aria-label="anchor" href="#arg-p-label"></a></dt>
<dd><p>Minimum adjusted p-value to display on the plot with a
bracket.</p><dl><dt><code>p.label = 0.05</code> - </dt>
<dd><p>Show p-values that are &lt;= 0.05.</p></dd>

<dt><code>p.label = 0</code> - </dt>
<dd><p>Don't show any p-values on the plot.</p></dd>

<dt><code>p.label = 1</code> - </dt>
<dd><p>Show all p-values on the plot.</p></dd>


</dl><p>If a numeric vector with more than one value is
provided, they will be used as breaks for asterisk notation.
Default: <code>0.05</code></p></dd>


<dt id="arg-transform">transform<a class="anchor" aria-label="anchor" href="#arg-transform"></a></dt>
<dd><p>Transformation to apply. Options are:
<code>c("none", "rank", "log", "log1p", "sqrt", "percent")</code>. <code>"rank"</code> is
useful for correcting for non-normally distributions before applying
regression statistics. Default: <code>"none"</code></p></dd>


<dt id="arg-y-transform">y.transform<a class="anchor" aria-label="anchor" href="#arg-y-transform"></a></dt>
<dd><p>The transformation to apply to the y-axis. Visualizing
differences of both high- and low-abundance taxa is best done with
a non-linear axis. Options are:</p><dl><dt><code>"sqrt"</code> - </dt>
<dd><p>square-root transformation</p></dd>

<dt><code>"log1p"</code> - </dt>
<dd><p>log(y + 1) transformation</p></dd>

<dt><code>"none"</code> - </dt>
<dd><p>no transformation</p></dd>


</dl><p>These methods allow visualization of both high- and low-abundance
taxa simultaneously, without complaint about 'zero' count
observations. Default: <code>"sqrt"</code>
Use <code>xaxis.transform</code> or <code>yaxis.transform</code> to pass custom values
directly to ggplot2's <code>scale_*</code> functions.</p></dd>


<dt id="arg-caption">caption<a class="anchor" aria-label="anchor" href="#arg-caption"></a></dt>
<dd><p>Add methodology caption beneath the plot.
Default: <code>TRUE</code></p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional parameters to pass along to ggplot2 functions.
Prefix a parameter name with a layer name to pass it to only that
layer. For instance, <code>d.size = 2</code> ensures only the points on the
<b>dot</b> layer have their size set to <code>2</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>ggplot2</code> plot. The computed data points, ggplot2 command,
stats table, and stats table commands are available as <code>$data</code>,
<code>$code</code>, <code>$stats</code>, and <code>$stats$code</code>, respectively.</p>
    </div>
    <div class="section level2">
    <h2 id="aesthetics">Aesthetics<a class="anchor" aria-label="anchor" href="#aesthetics"></a></h2>



<p>All built-in color palettes are colorblind-friendly. The available
categorical palette names are: <code>"okabe"</code>, <code>"carto"</code>, <code>"r4"</code>,
<code>"polychrome"</code>, <code>"tol"</code>, <code>"bright"</code>, <code>"light"</code>,
<code>"muted"</code>, <code>"vibrant"</code>, <code>"tableau"</code>, <code>"classic"</code>,
<code>"alphabet"</code>, <code>"tableau20"</code>, <code>"kelly"</code>, and <code>"fishy"</code>.</p>
<p>Patterns are added using the fillpattern R package. Options are <code>"brick"</code>,
<code>"chevron"</code>, <code>"fish"</code>, <code>"grid"</code>, <code>"herringbone"</code>, <code>"hexagon"</code>, <code>"octagon"</code>,
<code>"rain"</code>, <code>"saw"</code>, <code>"shingle"</code>, <code>"rshingle"</code>, <code>"stripe"</code>, and <code>"wave"</code>,
optionally abbreviated and/or suffixed with modifiers. For example,
<code>"hex10_sm"</code> for the hexagon pattern rotated 10 degrees and shrunk by 2x.
See <code><a href="https://cmmr.github.io/fillpattern/reference/fill_pattern.html" class="external-link">fillpattern::fill_pattern()</a></code> for complete documentation of options.</p>
<p>Shapes can be given as per base R - numbers 0 through 17 for various shapes,
or the decimal value of an ascii character, e.g. a-z = 65:90; A-Z = 97:122 to use
letters instead of shapes on the plot. Character strings may used as well.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other taxa_abundance:
<code><a href="sample_sums.html">sample_sums</a>()</code>,
<code><a href="taxa_clusters.html">taxa_clusters</a>()</code>,
<code><a href="taxa_corrplot.html">taxa_corrplot</a>()</code>,
<code><a href="taxa_heatmap.html">taxa_heatmap</a>()</code>,
<code><a href="taxa_stacked.html">taxa_stacked</a>()</code>,
<code><a href="taxa_stats.html">taxa_stats</a>()</code>,
<code><a href="taxa_sums.html">taxa_sums</a>()</code>,
<code><a href="taxa_matrix.html">taxa_table</a>()</code></p>
<p>Other visualization:
<code><a href="adiv_boxplot.html">adiv_boxplot</a>()</code>,
<code><a href="adiv_corrplot.html">adiv_corrplot</a>()</code>,
<code><a href="bdiv_boxplot.html">bdiv_boxplot</a>()</code>,
<code><a href="bdiv_corrplot.html">bdiv_corrplot</a>()</code>,
<code><a href="bdiv_heatmap.html">bdiv_heatmap</a>()</code>,
<code><a href="bdiv_ord_plot.html">bdiv_ord_plot</a>()</code>,
<code><a href="plot_heatmap.html">plot_heatmap</a>()</code>,
<code><a href="rare_corrplot.html">rare_corrplot</a>()</code>,
<code><a href="rare_multiplot.html">rare_multiplot</a>()</code>,
<code><a href="rare_stacked.html">rare_stacked</a>()</code>,
<code><a href="stats_boxplot.html">stats_boxplot</a>()</code>,
<code><a href="stats_corrplot.html">stats_corrplot</a>()</code>,
<code><a href="taxa_corrplot.html">taxa_corrplot</a>()</code>,
<code><a href="taxa_heatmap.html">taxa_heatmap</a>()</code>,
<code><a href="taxa_stacked.html">taxa_stacked</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">taxa_boxplot</span><span class="op">(</span><span class="va">biom</span>, stat.by <span class="op">=</span> <span class="st">"Body Site"</span>, stripe <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="taxa_boxplot-1.png" alt="" width="768" height="384"></span>
<span class="r-in"><span>    <span class="fu">taxa_boxplot</span><span class="op">(</span><span class="va">biom</span>, layers <span class="op">=</span> <span class="st">"bed"</span>, rank <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Genus"</span><span class="op">)</span>, flip <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="taxa_boxplot-2.png" alt="" width="768" height="384"></span>
<span class="r-in"><span>    <span class="fu">taxa_boxplot</span><span class="op">(</span></span></span>
<span class="r-in"><span>      biom    <span class="op">=</span> <span class="fu"><a href="subset.html">subset</a></span><span class="op">(</span><span class="va">biom</span>, <span class="va">`Body Site`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Saliva'</span>, <span class="st">'Stool'</span><span class="op">)</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>      taxa    <span class="op">=</span> <span class="fl">3</span>, </span></span>
<span class="r-in"><span>      layers  <span class="op">=</span> <span class="st">"ps"</span>, </span></span>
<span class="r-in"><span>      stat.by <span class="op">=</span> <span class="st">"Body Site"</span>,</span></span>
<span class="r-in"><span>      colors  <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Saliva'</span> <span class="op">=</span> <span class="st">"blue"</span>, <span class="st">'Stool'</span> <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">)</span></span></span>
<span class="r-plt img"><img src="taxa_boxplot-3.png" alt="" width="768" height="384"></span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

