<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Display taxa abundances as a heatmap. — taxa_heatmap • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Display taxa abundances as a heatmap. — taxa_heatmap"><meta name="description" content="Display taxa abundances as a heatmap."><meta property="og:description" content="Display taxa abundances as a heatmap."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Display taxa abundances as a heatmap.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/R/taxa_heatmap.r" class="external-link"><code>R/taxa_heatmap.r</code></a></small>
      <div class="d-none name"><code>taxa_heatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Display taxa abundances as a heatmap.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">taxa_heatmap</span><span class="op">(</span></span>
<span>  <span class="va">biom</span>,</span>
<span>  rank <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  taxa <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  tracks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  grid <span class="op">=</span> <span class="st">"bilbao"</span>,</span>
<span>  other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  unc <span class="op">=</span> <span class="st">"singly"</span>,</span>
<span>  lineage <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rescale <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  trees <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clust <span class="op">=</span> <span class="st">"complete"</span>,</span>
<span>  dist <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  asp <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  tree_height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  track_height <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legend <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  xlab.angle <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-biom">biom<a class="anchor" aria-label="anchor" href="#arg-biom"></a></dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt id="arg-rank">rank<a class="anchor" aria-label="anchor" href="#arg-rank"></a></dt>
<dd><p>What rank(s) of taxa to display. E.g. <code>"Phylum"</code>,
<code>"Genus"</code>, <code>".otu"</code>, etc. An integer vector can also be
given, where <code>1</code> is the highest rank, <code>2</code> is the second
highest, <code>-1</code> is the lowest rank, <code>-2</code> is the second
lowest, and <code>0</code> is the OTU "rank". Run <code>biom$ranks</code> to
see all options for a given rbiom object. Default: <code>-1</code>.</p></dd>


<dt id="arg-taxa">taxa<a class="anchor" aria-label="anchor" href="#arg-taxa"></a></dt>
<dd><p>Which taxa to display. An integer value will show the top n
most abundant taxa. A value 0 &lt;= n &lt; 1 will show any taxa with that
mean abundance or greater (e.g. <code>0.1</code> implies &gt;= 10%). A
character vector of taxa names will show only those named taxa.
Default: <code>6</code>.</p></dd>


<dt id="arg-tracks">tracks<a class="anchor" aria-label="anchor" href="#arg-tracks"></a></dt>
<dd><p>A character vector of metadata fields to display as tracks
at the top of the plot. Or, a list as expected by the <code>tracks</code>
argument of <code><a href="plot_heatmap.html">plot_heatmap()</a></code>. Default: <code>NULL</code></p></dd>


<dt id="arg-grid">grid<a class="anchor" aria-label="anchor" href="#arg-grid"></a></dt>
<dd><p>Color palette name, or a list as expected <code><a href="plot_heatmap.html">plot_heatmap()</a></code>.
Default: <code>"bilbao"</code></p></dd>


<dt id="arg-other">other<a class="anchor" aria-label="anchor" href="#arg-other"></a></dt>
<dd><p>Sum all non-itemized taxa into an "Other" taxa. When
<code>FALSE</code>, only returns taxa matched by the <code>taxa</code>
argument. Specifying <code>TRUE</code> adds "Other" to the returned set.
A string can also be given to imply <code>TRUE</code>, but with that
value as the name to use instead of "Other".
Default: <code>FALSE</code></p></dd>


<dt id="arg-unc">unc<a class="anchor" aria-label="anchor" href="#arg-unc"></a></dt>
<dd><p>How to handle unclassified, uncultured, and similarly ambiguous
taxa names. Options are:</p><dl><dt><code>"singly"</code> - </dt>
<dd><p>Replaces them with the OTU name.</p></dd>

<dt><code>"grouped"</code> - </dt>
<dd><p>Replaces them with a higher rank's name.</p></dd>

<dt><code>"drop"</code> - </dt>
<dd><p>Excludes them from the result.</p></dd>

<dt><code>"asis"</code> - </dt>
<dd><p>To not check/modify any taxa names.</p></dd>


</dl><p>Abbreviations are allowed. Default: <code>"singly"</code></p></dd>


<dt id="arg-lineage">lineage<a class="anchor" aria-label="anchor" href="#arg-lineage"></a></dt>
<dd><p>Include all ranks in the name of the taxa. For instance,
setting to <code>TRUE</code> will produce
<code>Bacteria; Actinobacteria; Coriobacteriia; Coriobacteriales</code>.
Otherwise the taxa name will simply be <code>Coriobacteriales</code>. You
want to set this to TRUE when <code>unc = "asis"</code> and you have taxa
names (such as <em>Incertae_Sedis</em>) that map to multiple higher
level ranks. Default: <code>FALSE</code></p></dd>


<dt id="arg-label">label<a class="anchor" aria-label="anchor" href="#arg-label"></a></dt>
<dd><p>Label the matrix rows and columns. You can supply a list
or logical vector of length two to control row labels and column
labels separately, for example
<code>label = c(rows = TRUE, cols = FALSE)</code>, or simply
<code>label = c(TRUE, FALSE)</code>. Other valid options are <code>"rows"</code>,
<code>"cols"</code>, <code>"both"</code>, <code>"bottom"</code>, <code>"right"</code>,
and <code>"none"</code>.
Default: <code>TRUE</code></p></dd>


<dt id="arg-label-size">label_size<a class="anchor" aria-label="anchor" href="#arg-label-size"></a></dt>
<dd><p>The font size to use for the row and column labels. You
can supply a numeric vector of length two to control row label sizes
and column label sizes separately, for example
<code>c(rows = 20, cols = 8)</code>, or simply <code>c(20, 8)</code>.
Default: <code>NULL</code>, which computes:
<code>pmax(8, pmin(20, 100 / dim(mtx)))</code></p></dd>


<dt id="arg-rescale">rescale<a class="anchor" aria-label="anchor" href="#arg-rescale"></a></dt>
<dd><p>Rescale rows or columns to all have a common min/max.
Options: <code>"none"</code>, <code>"rows"</code>, or <code>"cols"</code>.
Default: <code>"none"</code></p></dd>


<dt id="arg-trees">trees<a class="anchor" aria-label="anchor" href="#arg-trees"></a></dt>
<dd><p>Draw a dendrogram for rows (left) and columns (top). You can
supply a list or logical vector of length two to control the row tree
and column tree separately, for example
<code>trees = c(rows = TRUE, cols = FALSE)</code>,
or simply <code>trees = c(TRUE, FALSE)</code>.
Other valid options are <code>"rows"</code>, <code>"cols"</code>, <code>"both"</code>,
<code>"left"</code>, <code>"top"</code>, and <code>"none"</code>.
Default: <code>TRUE</code></p></dd>


<dt id="arg-clust">clust<a class="anchor" aria-label="anchor" href="#arg-clust"></a></dt>
<dd><p>Clustering algorithm for reordering the rows and columns by
similarity. You can supply a list or character vector of length two to
control the row and column clustering separately, for example
<code>clust = c(rows = "complete", cols = NA)</code>, or simply
<code>clust = c("complete", NA)</code>. Options are:</p><dl><dt><code>FALSE</code> or <code>NA</code> - </dt>
<dd><p>Disable reordering.</p></dd>

<dt>An <code>hclust</code> class object</dt>
<dd><p>E.g. from <code><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">stats::hclust()</a></code>.</p></dd>

<dt>A method name - </dt>
<dd><p><code>"ward.D"</code>,
<code>"ward.D2"</code>, <code>"single"</code>, <code>"complete"</code>,
<code>"average"</code>, <code>"mcquitty"</code>, <code>"median"</code>, or
<code>"centroid"</code>.</p></dd>


</dl><p>Default: <code>"complete"</code></p></dd>


<dt id="arg-dist">dist<a class="anchor" aria-label="anchor" href="#arg-dist"></a></dt>
<dd><p>Distance algorithm to use when reordering the rows and columns
by similarity. You can supply a list or character vector of length
two to control the row and column clustering separately, for example
<code>dist = c(rows = "euclidean", cols = "maximum")</code>, or simply
<code>dist = c("euclidean", "maximum")</code>. Options are:</p><dl><dt>A <code>dist</code> class object</dt>
<dd><p>E.g. from <code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">stats::dist()</a></code> or <code><a href="bdiv_table.html">bdiv_distmat()</a></code>.</p></dd>

<dt>A method name - </dt>
<dd><p><code>"euclidean"</code>,
<code>"maximum"</code>, <code>"manhattan"</code>, <code>"canberra"</code>,
<code>"binary"</code>, or <code>"minkowski"</code>.</p></dd>


</dl><p>Default: <code>"euclidean"</code></p></dd>


<dt id="arg-asp">asp<a class="anchor" aria-label="anchor" href="#arg-asp"></a></dt>
<dd><p>Aspect ratio (height/width) for entire grid.
Default: <code>1</code> (square)</p></dd>


<dt id="arg-tree-height-track-height">tree_height, track_height<a class="anchor" aria-label="anchor" href="#arg-tree-height-track-height"></a></dt>
<dd><p>The height of the dendrogram or annotation
tracks as a percentage of the overall grid size. Use a numeric vector
of length two to assign <code>c(top, left)</code> independently.
Default: <code>10</code> (10% of the grid's height)</p></dd>


<dt id="arg-legend">legend<a class="anchor" aria-label="anchor" href="#arg-legend"></a></dt>
<dd><p>Where to place the legend. Options are: <code>"right"</code> or
<code>"bottom"</code>. Default: <code>"right"</code></p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>Plot title. Set to <code>TRUE</code> for a default title, <code>NULL</code> for
no title, or any character string. Default: <code>TRUE</code></p></dd>


<dt id="arg-xlab-angle">xlab.angle<a class="anchor" aria-label="anchor" href="#arg-xlab-angle"></a></dt>
<dd><p>Angle of the labels at the bottom of the plot.
Options are <code>"auto"</code>, <code>'0'</code>, <code>'30'</code>, and <code>'90'</code>.
Default: <code>"auto"</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments to pass on to ggplot2::theme().</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A <code>ggplot2</code> plot. The computed data points and ggplot
command are available as <code>$data</code> and <code>$code</code>,
respectively.</p>
    </div>
    <div class="section level2">
    <h2 id="annotation-tracks">Annotation Tracks<a class="anchor" aria-label="anchor" href="#annotation-tracks"></a></h2>



<p>Metadata can be displayed as colored tracks above the heatmap. Common use
cases are provided below, with more thorough documentation available at
https://cmmr.github.io/rbiom .</p>
<div class="sourceCode"><pre><code><span><span class="co">## Categorical ----------------------------</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="st">"Body Site"</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'Body Site'</span> <span class="op">=</span> <span class="st">"bright"</span><span class="op">)</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'Body Site'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Stool'</span> <span class="op">=</span> <span class="st">"blue"</span>, <span class="st">'Saliva'</span> <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Numeric --------------------------------</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="st">"Age"</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'Age'</span> <span class="op">=</span> <span class="st">"reds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Multiple Tracks ------------------------</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Body Site"</span>, <span class="st">"Age"</span><span class="op">)</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'Body Site'</span> <span class="op">=</span> <span class="st">"bright"</span>, <span class="st">'Age'</span> <span class="op">=</span> <span class="st">"reds"</span><span class="op">)</span></span>
<span><span class="va">tracks</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">'Body Site'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Stool'</span> <span class="op">=</span> <span class="st">"blue"</span>, <span class="st">'Saliva'</span> <span class="op">=</span> <span class="st">"green"</span><span class="op">)</span>,</span>
<span>  <span class="st">'Age'</span>       <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'colors'</span> <span class="op">=</span> <span class="st">"reds"</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>

<p>The following entries in the track definitions are understood:</p>
<dl><dt><code>colors</code> - </dt>
<dd><p>A pre-defined palette name or custom set of colors to map to.</p></dd>

<dt><code>range</code> - </dt>
<dd><p>The c(min,max) to use for scale values.</p></dd>

<dt><code>label</code> - </dt>
<dd><p>Label for this track. Defaults to the name of this list element.</p></dd>

<dt><code>side</code> - </dt>
<dd><p>Options are <code>"top"</code> (default) or <code>"left"</code>.</p></dd>

<dt><code>na.color</code> - </dt>
<dd><p>The color to use for <code>NA</code> values.</p></dd>

<dt><code>bins</code> - </dt>
<dd><p>Bin a gradient into this many bins/steps.</p></dd>

<dt><code>guide</code> - </dt>
<dd><p>A list of arguments for guide_colorbar() or guide_legend().</p></dd>


</dl><p>All built-in color palettes are colorblind-friendly.</p>
<p>Categorical palette names: <code>"okabe"</code>, <code>"carto"</code>, <code>"r4"</code>,
<code>"polychrome"</code>, <code>"tol"</code>, <code>"bright"</code>, <code>"light"</code>,
<code>"muted"</code>, <code>"vibrant"</code>, <code>"tableau"</code>, <code>"classic"</code>,
<code>"alphabet"</code>, <code>"tableau20"</code>, <code>"kelly"</code>, and <code>"fishy"</code>.</p>
<p>Numeric palette names: <code>"reds"</code>, <code>"oranges"</code>, <code>"greens"</code>,
<code>"purples"</code>, <code>"grays"</code>, <code>"acton"</code>, <code>"bamako"</code>,
<code>"batlow"</code>, <code>"bilbao"</code>, <code>"buda"</code>, <code>"davos"</code>,
<code>"devon"</code>, <code>"grayC"</code>, <code>"hawaii"</code>, <code>"imola"</code>,
<code>"lajolla"</code>, <code>"lapaz"</code>, <code>"nuuk"</code>, <code>"oslo"</code>,
<code>"tokyo"</code>, <code>"turku"</code>, <code>"bam"</code>, <code>"berlin"</code>,
<code>"broc"</code>, <code>"cork"</code>, <code>"lisbon"</code>, <code>"roma"</code>,
<code>"tofino"</code>, <code>"vanimo"</code>, and <code>"vik"</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other taxa_abundance:
<code><a href="sample_sums.html">sample_sums</a>()</code>,
<code><a href="taxa_boxplot.html">taxa_boxplot</a>()</code>,
<code><a href="taxa_clusters.html">taxa_clusters</a>()</code>,
<code><a href="taxa_corrplot.html">taxa_corrplot</a>()</code>,
<code><a href="taxa_stacked.html">taxa_stacked</a>()</code>,
<code><a href="taxa_stats.html">taxa_stats</a>()</code>,
<code><a href="taxa_sums.html">taxa_sums</a>()</code>,
<code><a href="taxa_matrix.html">taxa_table</a>()</code></p>
<p>Other visualization:
<code><a href="adiv_boxplot.html">adiv_boxplot</a>()</code>,
<code><a href="adiv_corrplot.html">adiv_corrplot</a>()</code>,
<code><a href="bdiv_boxplot.html">bdiv_boxplot</a>()</code>,
<code><a href="bdiv_corrplot.html">bdiv_corrplot</a>()</code>,
<code><a href="bdiv_heatmap.html">bdiv_heatmap</a>()</code>,
<code><a href="bdiv_ord_plot.html">bdiv_ord_plot</a>()</code>,
<code><a href="plot_heatmap.html">plot_heatmap</a>()</code>,
<code><a href="rare_corrplot.html">rare_corrplot</a>()</code>,
<code><a href="rare_multiplot.html">rare_multiplot</a>()</code>,
<code><a href="rare_stacked.html">rare_stacked</a>()</code>,
<code><a href="stats_boxplot.html">stats_boxplot</a>()</code>,
<code><a href="stats_corrplot.html">stats_corrplot</a>()</code>,
<code><a href="taxa_boxplot.html">taxa_boxplot</a>()</code>,
<code><a href="taxa_corrplot.html">taxa_corrplot</a>()</code>,
<code><a href="taxa_stacked.html">taxa_stacked</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="co"># Keep and rarefy the 10 most deeply sequenced samples.</span></span></span>
<span class="r-in"><span>    <span class="va">hmp10</span> <span class="op">&lt;-</span> <span class="fu"><a href="rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">taxa_heatmap</span><span class="op">(</span><span class="va">hmp10</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span>, tracks <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="taxa_heatmap-1.png" alt="" width="768" height="384"></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">taxa_heatmap</span><span class="op">(</span><span class="va">hmp10</span>, rank <span class="op">=</span> <span class="st">"Genus"</span>, tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"bo"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="taxa_heatmap-2.png" alt="" width="768" height="384"></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">taxa_heatmap</span><span class="op">(</span><span class="va">hmp10</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span>, tracks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>      <span class="st">'Sex'</span>       <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>m <span class="op">=</span> <span class="st">"#0000FF"</span>, f <span class="op">=</span> <span class="st">"violetred"</span><span class="op">)</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>      <span class="st">'Body Site'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>colors <span class="op">=</span> <span class="st">"muted"</span>, label <span class="op">=</span> <span class="st">"Source"</span><span class="op">)</span> <span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="taxa_heatmap-3.png" alt="" width="768" height="384"></span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

