<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Save an rbiom object to a file. — write_biom • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.9/font.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Save an rbiom object to a file. — write_biom"><meta name="description" content="Automatically creates directories and adds compression based on file name.
write_biom() -
According to BIOM format specification.

write_xlsx() -
Raw data and summary tables in Excel file format. See details.

write_fasta() -
Sequences only in fasta format. biom may also be a named character vector.

write_tree() -
Phylogenetic tree only in newick format. biom may also be a phylo object.

write_counts(), write_metadata(), write_taxonomy() -
Tab-separated values.



"><meta property="og:description" content="Automatically creates directories and adds compression based on file name.
write_biom() -
According to BIOM format specification.

write_xlsx() -
Raw data and summary tables in Excel file format. See details.

write_fasta() -
Sequences only in fasta format. biom may also be a named character vector.

write_tree() -
Phylogenetic tree only in newick format. biom may also be a phylo object.

write_counts(), write_metadata(), write_taxonomy() -
Tab-separated values.



"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.13.9009</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Save an rbiom object to a file.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/R/write_biom.r" class="external-link"><code>R/write_biom.r</code></a>, <a href="https://github.com/cmmr/rbiom/blob/master/R/write_xlsx.r" class="external-link"><code>R/write_xlsx.r</code></a></small>
      <div class="d-none name"><code>write_biom.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Automatically creates directories and adds compression based on file name.</p><dl><dt><code>write_biom()</code> - </dt>
<dd><p>According to <a href="http://biom-format.org/documentation/" class="external-link">BIOM format</a> specification.</p></dd>

<dt><code>write_xlsx()</code> - </dt>
<dd><p>Raw data and summary tables in Excel file format. See details.</p></dd>

<dt><code>write_fasta()</code> - </dt>
<dd><p>Sequences only in fasta format. <code>biom</code> may also be a named character vector.</p></dd>

<dt><code>write_tree()</code> - </dt>
<dd><p>Phylogenetic tree only in newick format. <code>biom</code> may also be a phylo object.</p></dd>

<dt><code>write_counts()</code>, <code>write_metadata()</code>, <code>write_taxonomy()</code> - </dt>
<dd><p>Tab-separated values.</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">write_biom</span><span class="op">(</span><span class="va">biom</span>, <span class="va">file</span>, format <span class="op">=</span> <span class="st">"json"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">write_metadata</span><span class="op">(</span><span class="va">biom</span>, <span class="va">file</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">write_counts</span><span class="op">(</span><span class="va">biom</span>, <span class="va">file</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">write_taxonomy</span><span class="op">(</span><span class="va">biom</span>, <span class="va">file</span>, quote <span class="op">=</span> <span class="cn">FALSE</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, <span class="va">...</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">write_fasta</span><span class="op">(</span><span class="va">biom</span>, file <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">write_tree</span><span class="op">(</span><span class="va">biom</span>, file <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">write_xlsx</span><span class="op">(</span><span class="va">biom</span>, <span class="va">file</span>, depth <span class="op">=</span> <span class="fl">0.1</span>, n <span class="op">=</span> <span class="cn">NULL</span>, seed <span class="op">=</span> <span class="fl">0</span>, unc <span class="op">=</span> <span class="st">"singly"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-biom">biom<a class="anchor" aria-label="anchor" href="#arg-biom"></a></dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt id="arg-file">file<a class="anchor" aria-label="anchor" href="#arg-file"></a></dt>
<dd><p>Path to the output file. File names ending in <code>.gz</code> or
<code>.bz2</code> will be compressed accordingly.
Setting <code>file=NULL</code> for <code>write_fasta()</code>, <code>write_tree()</code>, and
<code>write_biom(format='json')</code>, and returns a string of the output which
would have been written. For <code>write_biom(format='tab')</code>, <code>file=NULL</code>
returns the tibble that would have been written.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>Options are <code>"tab"</code>, <code>"json"</code>, and <code>"hdf5"</code>,
corresponding to classic tabular format, BIOM format version 1.0 and
biom version 2.1, respectively. NOTE: to write HDF5 formatted BIOM
files, the BioConductor R package <code>rhdf5</code> must be installed.
Default: <code>"json"</code></p></dd>


<dt id="arg-quote-sep-">quote, sep, ...<a class="anchor" aria-label="anchor" href="#arg-quote-sep-"></a></dt>
<dd><p>Parameters passed on to <code><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table()</a></code>.
Default: <code>quote=FALSE, sep="\t"</code></p></dd>


<dt id="arg-depth-n">depth, n<a class="anchor" aria-label="anchor" href="#arg-depth-n"></a></dt>
<dd><p>Passed on to <code><a href="rarefy_cols.html">rarefy_cols()</a></code>. For <code>write_xlsx()</code> only,
<code>depth=0</code> disables rarefaction. Default: <code>depth=0.1, n=NULL</code></p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed to use in rarefying. See <code><a href="rarefy_cols.html">rarefy_cols()</a></code> function
for details. Must be a non-negative integer. Default: <code>0</code></p></dd>


<dt id="arg-unc">unc<a class="anchor" aria-label="anchor" href="#arg-unc"></a></dt>
<dd><p>How to handle unclassified, uncultured, and similarly ambiguous
taxa names. Options are:</p><dl><dt><code>"singly"</code> - </dt>
<dd><p>Replaces them with the OTU name.</p></dd>

<dt><code>"grouped"</code> - </dt>
<dd><p>Replaces them with a higher rank's name.</p></dd>

<dt><code>"drop"</code> - </dt>
<dd><p>Excludes them from the result.</p></dd>

<dt><code>"asis"</code> - </dt>
<dd><p>To not check/modify any taxa names.</p></dd>


</dl><p>Abbreviations are allowed. Default: <code>"singly"</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The normalized filepath that was written to (invisibly), unless
<code>file=NULL</code> (see <code>file</code> argument above).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For <code>write_xlsx()</code>, <code>attributes(biom)</code> are saved as additional worksheets if
the attribute is a data frame, matrix, or dist -class object. An attribute
named 'Reads Per Step' is treated specially and merged with the usual 'Reads
Per Sample' tab.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">write_tree</span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/substr.html" class="external-link">substr</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">50</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "(((((((((((((EschC738:0.03627,(((Unc92490:0.05748,"</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>      <span class="va">hmp10</span>        <span class="op">&lt;-</span> <span class="va">hmp50</span><span class="op">$</span><span class="fu">clone</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>      <span class="va">hmp10</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="va">hmp10</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="rarefy_cols.html">rarefy_cols</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span>      </span></span>
<span class="r-in"><span>      <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">hmp10</span>, <span class="st">"Weighted UniFrac"</span><span class="op">)</span>   <span class="op">&lt;-</span> <span class="fu"><a href="bdiv_table.html">bdiv_distmat</a></span><span class="op">(</span><span class="va">hmp10</span>, <span class="st">'unifrac'</span><span class="op">)</span></span></span>
<span class="r-in"><span>      <span class="fu"><a href="https://rdrr.io/r/base/attr.html" class="external-link">attr</a></span><span class="op">(</span><span class="va">hmp10</span>, <span class="st">"Unweighted Jaccard"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="bdiv_table.html">bdiv_distmat</a></span><span class="op">(</span><span class="va">hmp10</span>, <span class="st">'jaccard'</span>, weighted<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span>      </span></span>
<span class="r-in"><span>      <span class="va">outfile</span> <span class="op">&lt;-</span> <span class="fu">write_xlsx</span><span class="op">(</span><span class="va">hmp10</span>, <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>fileext <span class="op">=</span> <span class="st">".xlsx"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

