<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Benchmark • rbiom</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Benchmark">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rbiom</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/benchmark.html">Benchmark</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Benchmark</h1>
                        <h4 class="author">Daniel Smith, PhD</h4>
            
            <h4 class="date">7/2/2018</h4>
      
      
      <div class="hidden name"><code>benchmark.Rmd</code></div>

    </div>

    
    
<div id="overview" class="section level2">
<h2 class="hasAnchor">
<a href="#overview" class="anchor"></a>Overview</h2>
<p>Here we will compare rbiom to similar packages by measuring the time taken to complete common operations on a representative dataset.</p>
</div>
<div id="create-test-environment" class="section level2">
<h2 class="hasAnchor">
<a href="#create-test-environment" class="anchor"></a>Create Test Environment</h2>
<p>For our tests, we’ll use a virtual machine on Amazon Web Service’s Elastic Compute Cloud. Specifically: * US-West region * AMI 482e6430 (<a href="https://s3-us-west-2.amazonaws.com/qiime2-data/distro/core/aws-amis.txt">QIIME 2 Core</a>) * c3.8xlarge instance type (32 virtual CPUs) * 20 GiB General Purpose SSD (GP2) Storage on Root Volume</p>
<p>Both the username and password are ‘qiime2’. On this instance, qiime2 is already configured. The following will download and install rbiom, phyloseq, and usearch as well as the test dataset.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="fu">sudo</span> apt-get install build-essential gfortran libssl-dev unzip</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="bu">export</span> <span class="va">TAR=</span>/bin/tar</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="ex">R</span> -e <span class="st">'source("http://bioconductor.org/biocLite.R"); biocLite(c("phyloseq", "rhdf5"))'</span></a>
<a class="sourceLine" id="cb1-5" data-line-number="5"><span class="ex">R</span> -e <span class="st">'install.packages(c("devtools", "doParallel"), repos="http://cran.us.r-project.org");'</span></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="ex">R</span> -e <span class="st">'devtools::install_git("https://github.com/cmmr/rbiom.git")'</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"></a>
<a class="sourceLine" id="cb1-8" data-line-number="8"><span class="fu">wget</span> https://s3.amazonaws.com/jplab/share/30d/usearch10.0.240</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"><span class="fu">chmod</span> a+x usearch10.0.240</a>
<a class="sourceLine" id="cb1-10" data-line-number="10"></a>
<a class="sourceLine" id="cb1-11" data-line-number="11"><span class="fu">wget</span> https://raw.githubusercontent.com/cmmr/rbiom/master/vignettes/data/hmp500.biom</a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="ex">R</span> -e <span class="st">'x &lt;- rbiom::read.biom("hmp500.biom"); rbiom::write.tree(x$phylogeny, "hmp500.tre")'</span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="ex">R</span> -e <span class="st">'x &lt;- rbiom::read.biom("hmp500.biom"); x$taxonomy &lt;- NULL; rbiom::write.biom(x, "hmp500.tab", "tab")'</span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="ex">biom</span> convert -i hmp500.tab -o hmp500.json --table-type=<span class="st">"OTU table"</span> --to-json</a></code></pre></div>
</div>
<div id="prepare-input-data" class="section level2">
<h2 class="hasAnchor">
<a href="#prepare-input-data" class="anchor"></a>Prepare input data</h2>
<p>The example data is taken from the publicly available Human Microbiome Project. The master data set contains 4,468 samples with 4,468,000 observations of 1,140 OTUs. These samples have been rarefied to 1000 obervations per sample. The below script will randomly subset the master file to n samples, where n begins at 100 samples and is incremented by 100 all the way to 4,400. This will allow us to see how programs handle inputs of varying sizes.</p>
<p>prep_data.r:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(rbiom)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw">dir.create</span>(<span class="st">"in"</span>)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw">dir.create</span>(<span class="st">"out"</span>)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="kw">setwd</span>(<span class="st">"in"</span>)</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"></a>
<a class="sourceLine" id="cb2-7" data-line-number="7">x &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read.biom.html">read.biom</a></span>(<span class="st">"https://s3.amazonaws.com/jplab/projects/hmp/hmp4468.biom"</span>)</a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">44</span>) {</a>
<a class="sourceLine" id="cb2-9" data-line-number="9">  n &lt;-<span class="st"> </span>i <span class="op">*</span><span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10">  y &lt;-<span class="st"> </span><span class="kw"><a href="../reference/select.html">select</a></span>(x, <span class="kw">sample.int</span>(<span class="dv">4468</span>, n))</a>
<a class="sourceLine" id="cb2-11" data-line-number="11">  y<span class="op">$</span>taxonomy &lt;-<span class="st"> </span><span class="ot">NULL</span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12">  <span class="kw"><a href="../reference/write.biom.html">write.biom</a></span>(y, <span class="kw">paste0</span>(<span class="st">"hmp"</span>, n, <span class="st">".tab"</span>), <span class="st">"tab"</span>)</a>
<a class="sourceLine" id="cb2-13" data-line-number="13">  <span class="kw"><a href="../reference/write.tree.html">write.tree</a></span>(y<span class="op">$</span>phylogeny, <span class="kw">paste0</span>(<span class="st">"hmp"</span>, n, <span class="st">".tre"</span>))</a>
<a class="sourceLine" id="cb2-14" data-line-number="14">  cmd &lt;-<span class="st"> </span><span class="kw">sprintf</span>(<span class="st">'biom convert -i hmp%i.tab -o hmp%i.json --table-type="OTU table" --to-json'</span>, n, n)</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">  <span class="kw">system</span>(<span class="dt">command =</span> cmd, <span class="dt">wait =</span> <span class="ot">FALSE</span>, <span class="dt">ignore.stdout =</span> <span class="ot">TRUE</span>, <span class="dt">ignore.stderr =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-16" data-line-number="16">}</a></code></pre></div>
</div>
<div id="weighted-unifrac" class="section level2">
<h2 class="hasAnchor">
<a href="#weighted-unifrac" class="anchor"></a>Weighted UniFrac</h2>
<p>Create four scripts, one per program. These scripts will be the minimal steps needs to read in a biom file and newick, compute a distance matrix, then save the results to a tsv file. Each script will accept three arguments: (1) the number of cpu cores to use, (2) a single character ‘w’ or ‘u’ indicating whether to run weighted or unweighted unifrac, and (3) the infile size being tested, in number of samples.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="fu">cat</span> <span class="op">&gt;</span> rbiom.r <span class="op">&lt;&lt; EOF</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">args &lt;- commandArgs(trailingOnly=TRUE)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">options('rbiom.max.threads' = args[1])</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"></a>
<a class="sourceLine" id="cb3-6" data-line-number="6">bFile &lt;- paste0("in/hmp", args[3],".json")</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">tFile &lt;- paste0("in/hmp", args[3],".tre")</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">oFile &lt;- paste0("out/rbiom_dm", args[3],".tsv")</a>
<a class="sourceLine" id="cb3-9" data-line-number="9"></a>
<a class="sourceLine" id="cb3-10" data-line-number="10">dm &lt;- rbiom::unifrac(</a>
<a class="sourceLine" id="cb3-11" data-line-number="11">  biom     = rbiom::read.biom(bFile),</a>
<a class="sourceLine" id="cb3-12" data-line-number="12">  tree     = rbiom::read.tree(tFile), </a>
<a class="sourceLine" id="cb3-13" data-line-number="13">  weighted = identical(args[2], 'w') )</a>
<a class="sourceLine" id="cb3-14" data-line-number="14">write.table(as.matrix(dm), oFile, sep="\t", quote=FALSE)</a>
<a class="sourceLine" id="cb3-15" data-line-number="15"><span class="op">EOF</span></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"></a>
<a class="sourceLine" id="cb3-17" data-line-number="17"></a>
<a class="sourceLine" id="cb3-18" data-line-number="18"><span class="fu">cat</span> <span class="op">&gt;</span> phyloseq.r <span class="op">&lt;&lt; EOF</span></a>
<a class="sourceLine" id="cb3-19" data-line-number="19">args &lt;- commandArgs(trailingOnly=TRUE)</a>
<a class="sourceLine" id="cb3-20" data-line-number="20">doParallel::registerDoParallel(parallel::makeCluster(args[1]))</a>
<a class="sourceLine" id="cb3-21" data-line-number="21"></a>
<a class="sourceLine" id="cb3-22" data-line-number="22">bFile &lt;- paste0("in/hmp", args[3],".json")</a>
<a class="sourceLine" id="cb3-23" data-line-number="23">tFile &lt;- paste0("in/hmp", args[3],".tre")</a>
<a class="sourceLine" id="cb3-24" data-line-number="24">oFile &lt;- paste0("out/phyloseq_dm", args[3],".tsv")</a>
<a class="sourceLine" id="cb3-25" data-line-number="25"></a>
<a class="sourceLine" id="cb3-26" data-line-number="26">dm &lt;- phyloseq::UniFrac(</a>
<a class="sourceLine" id="cb3-27" data-line-number="27">  phy        = phyloseq::import_biom(bFile, tFile), </a>
<a class="sourceLine" id="cb3-28" data-line-number="28">  weighted   = identical(args[2], 'w'), </a>
<a class="sourceLine" id="cb3-29" data-line-number="29">  normalized = FALSE, </a>
<a class="sourceLine" id="cb3-30" data-line-number="30">  parallel   = TRUE )</a>
<a class="sourceLine" id="cb3-31" data-line-number="31">write.table(as.matrix(dm), oFile, sep="\t", quote=FALSE)</a>
<a class="sourceLine" id="cb3-32" data-line-number="32"><span class="op">EOF</span></a>
<a class="sourceLine" id="cb3-33" data-line-number="33"></a>
<a class="sourceLine" id="cb3-34" data-line-number="34"></a>
<a class="sourceLine" id="cb3-35" data-line-number="35"><span class="fu">cat</span> <span class="op">&gt;</span> qiime2.sh <span class="op">&lt;&lt; EOF</span></a>
<a class="sourceLine" id="cb3-36" data-line-number="36">if [ <span class="va">$2</span> = "u" ]; then metric="unweighted_unifrac"; else metric="weighted_unifrac"; fi</a>
<a class="sourceLine" id="cb3-37" data-line-number="37">qiime tools import --input-path in/hmp<span class="va">$3</span>.json --output-path json --type FeatureTable[Frequency] --source-format BIOMV100Format</a>
<a class="sourceLine" id="cb3-38" data-line-number="38">qiime tools import --input-path in/hmp<span class="va">$3</span>.tre --output-path tree --type Phylogeny[Rooted] --source-format NewickFormat</a>
<a class="sourceLine" id="cb3-39" data-line-number="39">qiime diversity beta-phylogenetic --i-table json.qza --i-phylogeny tree.qza --p-metric <span class="va">$metric</span> --output-dir qiime_out --p-n-jobs <span class="va">$2</span></a>
<a class="sourceLine" id="cb3-40" data-line-number="40">qiime tools export --output-dir qiime_out qiime_out/distance_matrix.qza</a>
<a class="sourceLine" id="cb3-41" data-line-number="41">mv qiime_out/distance-matrix.tsv out/qiime2_dm<span class="va">$3</span>.tsv</a>
<a class="sourceLine" id="cb3-42" data-line-number="42">rm -rf qiime_out</a>
<a class="sourceLine" id="cb3-43" data-line-number="43"><span class="op">EOF</span></a>
<a class="sourceLine" id="cb3-44" data-line-number="44"></a>
<a class="sourceLine" id="cb3-45" data-line-number="45"></a>
<a class="sourceLine" id="cb3-46" data-line-number="46"><span class="fu">cat</span> <span class="op">&gt;</span> usearch.sh <span class="op">&lt;&lt; EOF</span></a>
<a class="sourceLine" id="cb3-47" data-line-number="47">if [ <span class="va">$2</span> = "u" ]; then metric="unifrac_binary"; else metric="unifrac"; fi</a>
<a class="sourceLine" id="cb3-48" data-line-number="48">./usearch10.0.240 -beta_div in/hmp<span class="va">$3</span>.tab -metrics <span class="va">$metric</span> -tree in/hmp<span class="va">$3</span>.tre -quiet</a>
<a class="sourceLine" id="cb3-49" data-line-number="49">mv <span class="va">$metric</span>.txt out/unifrac_dm<span class="va">$3</span>.tsv</a>
<a class="sourceLine" id="cb3-50" data-line-number="50">rm <span class="va">$metric</span>.sorted.txt</a>
<a class="sourceLine" id="cb3-51" data-line-number="51">rm <span class="va">$metric</span>.tree</a>
<a class="sourceLine" id="cb3-52" data-line-number="52"><span class="op">EOF</span></a></code></pre></div>
<p>Finally, execute and time each script.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="ex">/usr/bin/time</span> -f <span class="st">'Time: %E  Memory: %M\n'</span> Rscript rbiom.r    8 w 500</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="ex">/usr/bin/time</span> -f <span class="st">'Time: %E  Memory: %M\n'</span> Rscript phyloseq.r 8 w 500</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="ex">/usr/bin/time</span> -f <span class="st">'Time: %E  Memory: %M\n'</span> /bin/sh qiime2.sh  8 w 500</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="ex">/usr/bin/time</span> -f <span class="st">'Time: %E  Memory: %M\n'</span> /bin/sh usearch.sh 8 w 500</a></code></pre></div>
<p>Results on an 8-core machine:</p>
<table class="table">
<thead><tr class="header">
<th>Package</th>
<th align="center">Time (MM:SS)</th>
<th align="right">Memory (kb)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>rbiom</td>
<td align="center">0:03.56</td>
<td align="right">109736</td>
</tr>
<tr class="even">
<td>qiime2</td>
<td align="center">0:28.17</td>
<td align="right">225896</td>
</tr>
<tr class="odd">
<td>usearch</td>
<td align="center">0:40.91</td>
<td align="right">7140</td>
</tr>
<tr class="even">
<td>phyloseq</td>
<td align="center">1:59.55</td>
<td align="right">445384</td>
</tr>
</tbody>
</table>
<p>Qiime2 says: <code>Weighted UniFrac is not parallelizable</code> Weighted Output matrices: rbiom == phyloseq != UniFrac. Unweighted Output matrices: rbiom == phyloseq == qiime2 != uniFrac. And unifrac_binary also doesn’t equal normalized unweighted phyloseq</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#overview">Overview</a></li>
      <li><a href="#create-test-environment">Create Test Environment</a></li>
      <li><a href="#prepare-input-data">Prepare input data</a></li>
      <li><a href="#weighted-unifrac">Weighted UniFrac</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Daniel P. Smith.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
