<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rbiom">
<title>Regression • rbiom</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Regression">
<meta property="og:description" content="rbiom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9125</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rbiom.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6>
    <a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a>
    <a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a>
    <a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a>
    <a class="dropdown-item" href="../articles/diversity.html">Diversity Metrics</a>
    <a class="dropdown-item" href="../articles/statistics.html">Statistics</a>
    <a class="dropdown-item" href="../reference/speed.html">Speed Ups</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Plotting</h6>
    <a class="dropdown-item" href="../articles/plots.html">Plot Types</a>
    <a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a>
    <a class="dropdown-item" href="../articles/scales.html">Scales and Transformations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Regression</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/HEAD/vignettes/articles/regression.Rmd" class="external-link"><code>vignettes/articles/regression.Rmd</code></a></small>
      <div class="d-none name"><code>regression.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Regression functions are provided for examining the association
between two continuous variables.</p>
<p>Examples of continuous variables are alpha diversity metrics, taxa
abundances, and sample metadata such as age or BMI. These all have a
numeric range in which a data point can take on any value in that
range.</p>
</div>
<div class="section level2">
<h2 id="children-heights">Children Heights<a class="anchor" aria-label="anchor" href="#children-heights"></a>
</h2>
<p>To start with a simple example, consider the <code>children</code>
dataset from the R package <code>npregfast</code>. It contains the age
and height measurements of 2500 children aged 5 through 19 years, split
by sex (1292 females and 1208 males).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span><span class="op">)</span></span>
<span><span class="co">#&gt;      sex height   age</span></span>
<span><span class="co">#&gt; 1   male 150.77 13.25</span></span>
<span><span class="co">#&gt; 2 female 170.59 14.17</span></span>
<span><span class="co">#&gt; 3 female 167.31 15.17</span></span>
<span><span class="co">#&gt; 4 female 165.72 16.58</span></span>
<span><span class="co">#&gt; 5 female 171.67 15.17</span></span>
<span><span class="co">#&gt; 6 female 143.74 12.67</span></span></code></pre></div>
<div class="section level3">
<h3 id="height-vs-age">Height vs Age<a class="anchor" aria-label="anchor" href="#height-vs-age"></a>
</h3>
<p>The most straight forward plot is putting ‘age’ on the x-axis,
‘height’ on the y-axis, and drawing a straight line that minimizes the
distances between the observed data points and the linear trendline.</p>
<p>With this plot, we’re testing the hypothesis “as you get older, your
height changes”.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tp'</span>,                 <span class="co"># show trendline and points</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emtrends'</span>,           <span class="co"># run stats on the slope</span></span>
<span>  pt.size  <span class="op">=</span> <span class="fl">0.2</span>,                  <span class="co"># make points smaller</span></span>
<span>  pt.alpha <span class="op">=</span> <span class="fl">0.2</span> <span class="op">)</span>                 <span class="co"># and semi-transparent</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-3-1.png" width="768"></p>
<p>As you’d expect, the trendline’s slope is positive (p-value = 0).
Therefore it’s safe to say that height changes with age.</p>
</div>
<div class="section level3">
<h3 id="grouped-by-sex">Grouped by Sex<a class="anchor" aria-label="anchor" href="#grouped-by-sex"></a>
</h3>
<p>Is the rate of growth influenced by sex? We can add
<code>stat.by = "sex"</code> to draw separate lines for each sex.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tp'</span>,                 <span class="co"># show trendline and points</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emtrends'</span>,           <span class="co"># run stats on the slopes</span></span>
<span>  pt.size  <span class="op">=</span> <span class="fl">0.2</span>,                  <span class="co"># make points smaller</span></span>
<span>  pt.alpha <span class="op">=</span> <span class="fl">0.2</span> <span class="op">)</span>                 <span class="co"># and semi-transparent</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-4-1.png" width="768"></p>
<p>If you look at the captions of the last two plots, you’ll notice that
the p-value test automatically switches. When
<code>stat.by = NULL</code>, <code>test = "emtrends"</code> will test if
the slope is zero. When <code>stat.by</code> is not <code>NULL</code>,
<code>test = "emtrends"</code> will test if the trendline slopes are
equal.</p>
<p>Here, a p-value of 3.9e-51 indicates that males and females do have
different growth rates.</p>
</div>
<div class="section level3">
<h3 id="smoothed-fit">Smoothed Fit<a class="anchor" aria-label="anchor" href="#smoothed-fit"></a>
</h3>
<p>Linear trendlines probably aren’t the best way to model growth rates.
Setting <code>fit = "gam"</code> will use a generalized additive model
which fits several sub-ranges of age with independent splines.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'gam'</span>,                <span class="co"># smoothed trendline</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tp'</span>,                 <span class="co"># show trendline and points</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emtrends'</span>,           <span class="co"># run stats on the slopes</span></span>
<span>  pt.size  <span class="op">=</span> <span class="fl">0.2</span>,                  <span class="co"># make points smaller</span></span>
<span>  pt.alpha <span class="op">=</span> <span class="fl">0.2</span> <span class="op">)</span>                 <span class="co"># and semi-transparent</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<p>This gives us a much better idea of the moving average over time.</p>
<p>We still have <code>test = "emtrends"</code>, so the p-value is
reported where the differences is slope is most significant - in this
case, at age = 5.83 years.</p>
</div>
<div class="section level3">
<h3 id="difference-in-means">Difference in Means<a class="anchor" aria-label="anchor" href="#difference-in-means"></a>
</h3>
<p>Let’s set <code>test = "emmeans"</code> instead, and show the 95%
confidence interval instead of all the data points.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'gam'</span>,                <span class="co"># smoothed trendline</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tc'</span>,                 <span class="co"># show trendline with conf. int.</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emmeans'</span> <span class="op">)</span>           <span class="co"># run stats on the means</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>Differences in height are most significant at age = 16.8 (p-value =
8.9e-40).</p>
</div>
<div class="section level3">
<h3 id="stats-table">Stats Table<a class="anchor" aria-label="anchor" href="#stats-table"></a>
</h3>
<p>The complete statistical output is attached to the returned plot as
<code>$stats</code>. Though not covered in this article, you can also
access <code>$code</code> and <code>$stats$code</code> to see the
ggplot2 plotting command and statistics R code, respectively.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'gam'</span>,                <span class="co"># smoothed trendline</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tc'</span>,                 <span class="co"># show trendline with conf. int.</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emmeans'</span> <span class="op">)</span>           <span class="co"># run stats on the means</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    gam(height ~ s(age, by = sex, bs = "cs") + sex, method = "REML")</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 14</span></span></span>
<span><span class="co">#&gt;     age sex    .mean.diff .h1     .p.val   .adj.p .effect.size   .se    .n   .df</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>  16.8 femal…      -<span style="color: #BB0000;">7.73</span> != 0  8.88<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span> 8.88<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span>        -<span style="color: #BB0000;">1.21</span> 0.575  <span style="text-decoration: underline;">2</span>500  <span style="text-decoration: underline;">2</span>485</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more variables: .t.ratio &lt;dbl&gt;, .aic &lt;dbl&gt;, .bic &lt;dbl&gt;, .loglik &lt;dbl&gt;</span></span></span></code></pre></div>
<p>This is the same table you can generate with the <a href="#stats_table">stats_table()</a> function, and that reference page
has an explanation of each field.</p>
</div>
</div>
<div class="section level2">
<h2 id="babies-dataset">Babies Dataset<a class="anchor" aria-label="anchor" href="#babies-dataset"></a>
</h2>
<p>Now that we have the concepts down, let’s switch to the
<code>babies</code> dataset included with rbiom. It has microbiome data
from 2,684 stool samples from 12 infants. The original Nature article by
Eric J. de Muinck and Pal Trosvik for this data set is available online
for free at <a href="https://www.nature.com/articles/s41467-018-04641-7" class="external-link">Individuality
and Convergence of the Infant Gut Microbiota During the First Year of
Life</a>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">babies</span></span>
<span><span class="co">#&gt; DailyBaby (2018-06-08)</span></span>
<span><span class="co">#&gt; Data from [Muinck,</span></span>
<span><span class="co">#&gt; 2018](https://www.nature.com/articles/s41467-018-04641-7).</span></span>
<span><span class="co">#&gt; Retrieved from</span></span>
<span><span class="co">#&gt; [MicrobiomeHD](https://doi.org/10.1038/s41467-017-01973-8). </span></span>
<span><span class="co">#&gt; -----------</span></span>
<span><span class="co">#&gt;    2684 Samples:  SRR7044380, SRR7058817, SRR7039319, ...</span></span>
<span><span class="co">#&gt;     506 OTUs:     ASV085, ASV022, ASV251, ..., and ASV500</span></span>
<span><span class="co">#&gt;       8 Ranks:    .otu, Kingdom, Phylum, ..., and Species</span></span>
<span><span class="co">#&gt;      10 Metadata: .sample, Subject ID, Sex, ...</span></span>
<span><span class="co">#&gt;         Tree:     &lt;absent&gt;</span></span>
<span></span>
<span></span>
<span><span class="fu">glimpse</span><span class="op">(</span><span class="va">babies</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 2,684</span></span>
<span><span class="co">#&gt; Columns: 10</span></span>
<span><span class="co">#&gt; $ .sample                       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "SRR7044380", "SRR7058817", "SRR7039319"…</span></span>
<span><span class="co">#&gt; $ `Subject ID`                  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> ID3, ID12, ID1, ID6, ID8, ID11, ID11, ID…</span></span>
<span><span class="co">#&gt; $ Sex                           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Female, Female, Female, Male, Female, Fe…</span></span>
<span><span class="co">#&gt; $ `Age (days)`                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 255, 31, 222, 41, 79, 228, 52, 32, 178, …</span></span>
<span><span class="co">#&gt; $ `Child's diet`                <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Breast milk, Breast milk and formula, Br…</span></span>
<span><span class="co">#&gt; $ `Sample collection`           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Stored in alcohol, Frozen upon collectio…</span></span>
<span><span class="co">#&gt; $ `Antibiotic exposure`         <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> No, No, No, No, No, No, No, No, No, No, …</span></span>
<span><span class="co">#&gt; $ `Antifungal exposure`         <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> No, No, No, No, No, No, No, No, No, No, …</span></span>
<span><span class="co">#&gt; $ `Delivery mode`               <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Vaginal, Vaginal, Vaginal, Vaginal, Vagi…</span></span>
<span><span class="co">#&gt; $ `Solid food introduced (Age)` <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 165, 122, 186, 121, 247, 173, 173, 121, …</span></span></code></pre></div>
<p>We’ll use the <code>age</code>, <code>sex</code>,
<code>delivery</code>, and <code>subject</code> metadata fields in the
following examples. <code>age</code> is a continuous variable while the
rest are categorical.</p>
</div>
<div class="section level2">
<h2 id="normality">Normality<a class="anchor" aria-label="anchor" href="#normality"></a>
</h2>
<p>Many common statistical tools rely on the assumption that your data
is normally distributed.</p>
<p>In the children dataset, this assumption is accurate.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">performance</span><span class="fu">::</span><span class="fu"><a href="https://easystats.github.io/performance/reference/check_model.html" class="external-link">check_model</a></span><span class="op">(</span></span>
<span>  check <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pp_check"</span>, <span class="st">"linearity"</span><span class="op">)</span>,</span>
<span>  x     <span class="op">=</span> <span class="fu">mgcv</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/mgcv/man/gam.html" class="external-link">gam</a></span><span class="op">(</span></span>
<span>    data    <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,</span>
<span>    formula <span class="op">=</span> <span class="va">height</span> <span class="op">~</span> <span class="fu">s</span><span class="op">(</span><span class="va">age</span>, bs <span class="op">=</span> <span class="st">'cs'</span><span class="op">)</span>, </span>
<span>    method  <span class="op">=</span> <span class="st">'REML'</span> <span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
<p>However, microbiome data is usually non-normally distributed.</p>
<p>performance::check_model( check = c(“pp_check”, “linearity”), x =
mgcv::gam( data = adiv_table(babies) %&gt;% rename(age = ‘Age (days)’),
formula = .diversity ~ s(age, bs = ‘cs’), method = ‘REML’ ) )</p>
<pre><code>









# Draft Text




You can use the formula `height ~ age` as a way of saying "as you get older, your height changes". You can get fancier using a formula like `height ~ log(age)` which says that larger values of age have less of an effect on height, or `height ~ log(age) * sex` which also accounts for different growth trends of males and females. If you're of the opinion that height and age are linked, but not linearly or logarithmically, you can write `height ~ s(age)` to partition age into sub-ranges and fit splines to each independently.

That's as complex as it gets for rbiom! And again, rbiom insulates you from needing to write any formulas yourself. All you need to do is point the functions to the relevant variables.


&lt;img src="D:/Dropbox/Baylor/GitHub/rbiom/docs/articles/regression_files/figure-html/unnamed-chunk-10-1.png" width="768" /&gt;



## Models and Trendlines

Before we go further, let's define a couple terms.

* **Model Family** - A generalized equation, e.g. `y = mx + b`.
* **Fitted Model** - An equation for a specific line, e.g. `y = 1.2x + 10`.
* **Coefficients** - In the above fitted model, `1.2` and `10`.

For more on this topic, see the online book [R for Data Science](https://r4ds.had.co.nz/model-basics.html) by Hadley Wickham and Garrett Grolemund.

You can tell rbiom which model family to use with the `fit` parameter:
* `fit = "lm"` - Fit a linear `y ~ x` model.
* `fit = "log"` - Fit a logarithmic `y ~ log(x)` model.
* `fit = "gam"` - Fit a generalized additive `y ~ s(x)` model.

One important thing to know about rbiom is that **model fitting is done AFTER splitting the dataset**. For example, if your formula is `height ~ age * sex`, then the data will be split into two datasets - one of just males and one of just females. Each new dataset and the formula of `height ~ age` is passed on to the underlying statistical f



## Optimal Model/Trendline

So, how do you decide whether to use a linear, logarithmic, or smoothed trendline?

Looking at the different fits with your eye is an excellent place to start. There are also some metrics and charts to help you to really evaluate the goodness of fit.

Below (and in the figures above) data is sourced from the R package dataset `npregfast::children`, which has the age and height measurements of 2500 children aged 5 to 19 years, split by sex (1292 females and 1208 males). 

To start, we can use [broom::glance()] or rbiom's `stats_table()` to calculate R&lt;sup&gt;2&lt;/sup&gt;, AIC, and other goodness of fit metrics.


```r
children &lt;- as_tibble(npregfast::children)
glimpse(children)
#&gt; Rows: 2,500
#&gt; Columns: 3
#&gt; $ sex     [3m [38;5;246m&lt;fct&gt; [39m [23m male, female, female, female, female, female, female, male, fem…
#&gt; $ height  [3m [38;5;246m&lt;dbl&gt; [39m [23m 150.77, 170.59, 167.31, 165.72, 171.67, 143.74, 157.93, 157.73,…
#&gt; $ age     [3m [38;5;246m&lt;dbl&gt; [39m [23m 13.25, 14.17, 15.17, 16.58, 15.17, 12.67, 12.75, 17.17, 16.67, …


# Using broom::glance() directly:
models &lt;- with(children, list(
  'height ~ age'      = stats::lm(height ~ age),
  'height ~ log(age)' = stats::lm(height ~ log(age)),
  'height ~ s(age)'   = mgcv::gam(height ~ s(age, bs = 'cs'), method = 'REML') ))

plyr::ldply(models, broom::glance) %&gt;% 
  as_tibble() %&gt;% 
  select(model = .id, r.squared, adj.r.squared, logLik:BIC, p.value)
#&gt;  [38;5;246m# A tibble: 3 × 7 [39m
#&gt;   model             r.squared adj.r.squared logLik    AIC    BIC p.value
#&gt;    [3m [38;5;246m&lt;chr&gt; [39m [23m                  [3m [38;5;246m&lt;dbl&gt; [39m [23m          [3m [38;5;246m&lt;dbl&gt; [39m [23m   [3m [38;5;246m&lt;dbl&gt; [39m [23m   [3m [38;5;246m&lt;dbl&gt; [39m [23m   [3m [38;5;246m&lt;dbl&gt; [39m [23m    [3m [38;5;246m&lt;dbl&gt; [39m [23m
#&gt;  [38;5;250m1 [39m height ~ age          0.770         0.770 - [31m [4m8 [24m58 [39m [31m2 [39m [31m. [39m  [4m1 [24m [4m7 [24m171.  [4m1 [24m [4m7 [24m188.       0
#&gt;  [38;5;250m2 [39m height ~ log(age)     0.783         0.783 - [31m [4m8 [24m50 [39m [31m7 [39m [31m. [39m  [4m1 [24m [4m7 [24m020.  [4m1 [24m [4m7 [24m037.       0
#&gt;  [38;5;250m3 [39m height ~ s(age)       [31mNA [39m             [31mNA [39m     - [31m [4m8 [24m40 [39m [31m2 [39m [31m. [39m  [4m1 [24m [4m6 [24m821.  [4m1 [24m [4m6 [24m872.       [31mNA [39m


# Or equivalently with stats_table()
stats_table(children, resp = "height", regr = "age", fit = "lm") %&gt;% 
  select(.r.sqr:.fit.p)
#&gt;  [38;5;246m# Model:    lm(height ~ age) [39m
#&gt;  [38;5;246m# A tibble: 1 × 6 [39m
#&gt;   .r.sqr .adj.r   .aic   .bic .loglik .fit.p
#&gt;     [3m [38;5;246m&lt;dbl&gt; [39m [23m   [3m [38;5;246m&lt;dbl&gt; [39m [23m   [3m [38;5;246m&lt;dbl&gt; [39m [23m   [3m [38;5;246m&lt;dbl&gt; [39m [23m    [3m [38;5;246m&lt;dbl&gt; [39m [23m   [3m [38;5;246m&lt;dbl&gt; [39m [23m
#&gt;  [38;5;250m1 [39m  0.770  0.770  [4m1 [24m [4m7 [24m171.  [4m1 [24m [4m7 [24m188.  - [31m [4m8 [24m58 [39m [31m2 [39m [31m. [39m      0</code></pre>
<p>Here’s how to interpret this output:</p>
<ul>
<li>
<code>.r.sqr</code> - Coefficient of Determination
(<i>R</i><sup>2</sup>); range: 0-1, lower is better. Percent of
variation explained by the model.</li>
<li>
<code>.adj.r</code> - <i>R</i><sup>2</sup>, taking degrees of
freedom into account.</li>
<li>
<code>.aic</code> - Akaike Information Criterion; lower is better.
Preferred when using model for prediction.</li>
<li>
<code>.bic</code> - Bayesian Information Criterion; lower is better.
Preferred when using model for interpretation.</li>
<li>
<code>.loglik</code> - Log-Likelihood (negative values); higher
values (closer to zero) are better.</li>
<li>
<code>.fit.p</code> - P-value for observing this fit by chance;
range: 0-1, lower is better.</li>
</ul>
<p>Based on these metrics, the smoothed trendline is optimal.</p>
<p>We can also visualize the fit with the <code>performance</code> R
package. Below, compare the graphs for the linear vs smoothed
trendline.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/with.html">with</a></span><span class="op">(</span><span class="va">children</span>, <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">height</span> <span class="op">~</span> <span class="va">age</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">performance</span><span class="fu">::</span><span class="fu"><a href="https://easystats.github.io/performance/reference/check_model.html" class="external-link">check_model</a></span><span class="op">(</span>check <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pp_check"</span>, <span class="st">"linearity"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/with.html">with</a></span><span class="op">(</span><span class="va">children</span>, <span class="fu">mgcv</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/mgcv/man/gam.html" class="external-link">gam</a></span><span class="op">(</span><span class="va">height</span> <span class="op">~</span> <span class="fu">s</span><span class="op">(</span><span class="va">age</span>, bs <span class="op">=</span> <span class="st">'cs'</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">'REML'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">performance</span><span class="fu">::</span><span class="fu"><a href="https://easystats.github.io/performance/reference/check_model.html" class="external-link">check_model</a></span><span class="op">(</span>check <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pp_check"</span>, <span class="st">"linearity"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
<p>Omitting the <code>checks =</code> argument will generate even more
plots, and adding in the <code>* sex</code> interaction term will
produce even more plots, but I’ll leave that as an exercise for the
reader. Just an FYI - the gam formula with an interaction will look like
<code>height ~ s(age, by=sex, bs='cs') + sex</code>. Aren’t you glad
that rbiom doesn’t need you to know formulas?</p>
<p>From these checks we can conclude that our model fits the
<code>children</code> dataset very well and can give credence to any
conclusions reached by interpretations of that model.</p>
<div class="section level3">
<h3 id="stats_table">stats_table()<a class="anchor" aria-label="anchor" href="#stats_table"></a>
</h3>
<p>We’ll use <a href="#stats_table">stats_table()</a> in this
article.</p>
<p><a href="#stats_table">stats_table()</a> is called by [adiv_stats()],
[bdiv_stats()], and [taxa_stats()], which in turn are called by the
<code>*_corrplot()</code> plotting functions. Therefore, once you
understand how to use <a href="#stats_table">stats_table()</a>, you’ll
understand all the other regression functions too!</p>
<p>Here are the first few arguments for <a href="#stats_table">stats_table()</a>, explained in more detail than on
the reference page.</p>
<ul>
<li>
<code>df</code> - A data.frame with all your data. Typically, this
will be the data.frame returned by [adiv_stats()], [bdiv_stats()], or
[taxa_stats()], but a custom-made one is fine too. It doesn’t even have
to be microbiome data. A nice thing about the <code>_stats()</code>
functions is they set
e.g. <code>attr(df, 'response') &lt;- ".abundance"</code> so you don’t
need to worry about setting <code>stats_table(resp = )</code>.</li>
<li>
<code>regr</code> - The independent variable. Such as
<code>"age"</code> in the <code>babies</code> dataset.</li>
<li>
<code>resp</code> - The dependent variable. Will be
<code>".abundance"</code>, <code>".diversity"</code>, or
<code>".distance"</code> if calculated by the <code>_stats()</code>
functions, or <code>"height"</code> from the earlier section.</li>
</ul>
<p>As this article goes on, we’ll go over how and when to use additional
<a href="#stats_table">stats_table()</a> arguments.</p>
</div>
<div class="section level3">
<h3 id="linear-trends">Linear Trends<a class="anchor" aria-label="anchor" href="#linear-trends"></a>
</h3>
<p>Let’s explore Firmicutes abundance in the <code>babies</code>
dataset.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxa_matrix.html">taxa_table</a></span><span class="op">(</span><span class="va">babies</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span>, taxa <span class="op">=</span> <span class="st">"Firmicutes"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span></span>
<span>    <span class="va">.sample</span>, <span class="va">.taxa</span>, <span class="va">.abundance</span>, </span>
<span>    age      <span class="op">=</span> <span class="va">`Age (days)`</span>, </span>
<span>    sex      <span class="op">=</span> <span class="va">Sex</span>, </span>
<span>    delivery <span class="op">=</span> <span class="va">`Delivery mode`</span>, </span>
<span>    subject  <span class="op">=</span> <span class="va">`Subject ID`</span> <span class="op">)</span></span>
<span></span>
<span><span class="va">df</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2,684 × 7</span></span></span>
<span><span class="co">#&gt;   .sample    .taxa      .abundance   age sex    delivery subject</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> SRR7044380 Firmicutes        132   255 Female Vaginal  ID3    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> SRR7058817 Firmicutes         11    31 Female Vaginal  ID12   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> SRR7039319 Firmicutes         55   222 Female Vaginal  ID1    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> SRR7053418 Firmicutes        397    41 Male   Vaginal  ID6    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> SRR7054009 Firmicutes        103    79 Female Vaginal  ID8    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2,679 more rows</span></span></span></code></pre></div>
<p>The simplest way to find the abundance is to take the average across
all samples.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">.abundance</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 183.3104</span></span>
<span></span>
<span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">$</span><span class="va">.mean</span></span>
<span><span class="co">#&gt; [1] 183.3104</span></span></code></pre></div>
<p>Regression comes into play when we suspect that abundance changes
over time.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/with.html">with</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">age</span>, <span class="va">.abundance</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, fit <span class="op">=</span> <span class="st">"lm"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">.slope</span>, <span class="va">.h1</span>, <span class="va">.p.val</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ age)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;   .slope .h1     .p.val</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> -<span style="color: #BB0000;">0.127</span> != 0  0.000<span style="text-decoration: underline;">272</span></span></span></code></pre></div>
<p>Above, we’re modeling <code>.abundance ~ age</code> and using
[emmeans::emtrends()] behind the scenes see if the resulting trendline
is non-horizontal.</p>
<p>Recall from your statistic classes that you are testing a null
hypothesis H<sub>0</sub> against an alternate hypothesis H<sub>1</sub>.
In this case our null hypothesis is that the trendline is flat (slope =
0), indicating no relationship between age and Firmicutes abundance. The
p-value is the probability that the null hypothesis is correct (a
p-value of 0.6 is interpreted as a 60% chance that there’s no
relationship between age and abundance). When the p-value is below a
certain value (usually 0.05) we accept the alternative hypothesis
instead - in this case that there IS a relationship between age and
Firmicutes abundance.</p>
<p>In the output we see that <code>.slope</code> is -0.127 and our
alternative hypothesis (<code>.h1</code>) of <code>.slope != 0</code> is
acceptable since <code>.p.val</code> is less than 0.05. Therefore, the
mean abundance of Firmicutes decreases over time in babies.</p>
<p>The un-<code>select()</code>ed output contains columns for confidence
intervals, goodness of fit measures, and intermediary statistic values
that we’ll get to later.</p>
</div>
<div class="section level3">
<h3 id="data-normality">Data Normality<a class="anchor" aria-label="anchor" href="#data-normality"></a>
</h3>
<p>In statistics, it is crucial to know if your data is normally
distributed.</p>
<p>The <code>performance</code> R package has excellent utilities for
assessing normality and similar characteristics.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/with.html">with</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">.abundance</span> <span class="op">~</span> <span class="va">age</span> <span class="op">*</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">performance</span><span class="fu">::</span><span class="fu"><a href="https://easystats.github.io/performance/reference/check_model.html" class="external-link">check_model</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>The “Normality of Residuals” plots clearly show we’re dealing with
non-normal data. To handle this, we need to use non-parametric methods
or manually apply a rank transformation.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span><span class="op">$</span><span class="va">rt_abundance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rank.html" class="external-link">rank</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">.abundance</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/with.html">with</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">rt_abundance</span> <span class="op">~</span> <span class="va">age</span> <span class="op">*</span> <span class="va">sex</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">performance</span><span class="fu">::</span><span class="fu"><a href="https://easystats.github.io/performance/reference/check_model.html" class="external-link">check_model</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-18-1.png" width="768"></p>
</div>
<div class="section level3">
<h3 id="trends-and-engines">Trends and Engines<a class="anchor" aria-label="anchor" href="#trends-and-engines"></a>
</h3>
<p>Now’s a good time to explain two more <code><a href="../reference/stats_table.html">stats_table()</a></code>
parameters: <code>fit</code> and <code>test</code>.</p>
<p><code>fit</code> is the shape of the trendline you want to fit (see
intro section). Options are: * <code>'lm'</code> - Linear (straight
line), using <code>stats::lm(y ~ x)</code>. * <code>'log'</code> -
Logarithmic, using <code>stats::lm(y ~ log(x))</code>. *
<code>'gam'</code> - Generalized additive model, using
<code>mgcv::gam(y ~ s(x, bs = 'cs'), method = 'REML')</code>.</p>
<p><code>test</code> is the top-most operator the model. Options are: *
<code>'emmeans'</code> - To compute means. * <code>'emtrends'</code> -
To compute trendline slopes.</p>
</div>
<div class="section level3">
<h3 id="marginal-means">Marginal Means<a class="anchor" aria-label="anchor" href="#marginal-means"></a>
</h3>
<p>The default test for <code><a href="../reference/stats_table.html">stats_table()</a></code> is
<code>'emmeans'</code>, which computes estimated marginal means (EMMs),
also known as least-squares means, using the [emmeans::emmeans]
package.</p>
<p>Generally speaking, EMMs are means extracted from a statistical
model. This allows EMMs to take into account more complex associations
and produce confidence intervals in addition to estimates of the mean.
For a more information on EMMs, these resources are recommended:</p>
<ul>
<li><a href="https://www.andrewheiss.com/blog/2022/05/20/marginalia/#marginal-effects-at-the-mean-the-default-in-emmeans" class="external-link">Marginalia:
A Guide to Figuring Out What the Heck Marginal Effects Are</a></li>
<li><a href="https://stats.oarc.ucla.edu/r/seminars/interactions-r" class="external-link">Decomposing,
Probing, and Plotting Interactions in R</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="model-selection">Model Selection<a class="anchor" aria-label="anchor" href="#model-selection"></a>
</h3>
<p>As per our discussion of age vs height, a straight (linear) trendline
might not be the best fit for a dataset.</p>
<p>To evaluate different models, change the <code>fit</code> parameter
and look at the resultant goodness of fit values.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span>, fit <span class="op">=</span> <span class="st">"lm"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">.r.sqr</span><span class="op">:</span><span class="va">.fit.p</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ age)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 6</span></span></span>
<span><span class="co">#&gt;    .r.sqr  .adj.r   .aic   .bic .loglik   .fit.p</span></span>
<span><span class="co">#&gt;     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 0.004<span style="text-decoration: underline;">93</span> 0.004<span style="text-decoration: underline;">56</span> <span style="text-decoration: underline;">33</span>519. <span style="text-decoration: underline;">33</span>537. -<span style="color: #BB0000; text-decoration: underline;">16</span><span style="color: #BB0000;">757.</span> 0.000<span style="text-decoration: underline;">272</span></span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span>, fit <span class="op">=</span> <span class="st">"log"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">.r.sqr</span><span class="op">:</span><span class="va">.fit.p</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ log(age))</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 6</span></span></span>
<span><span class="co">#&gt;    .r.sqr  .adj.r   .aic   .bic .loglik .fit.p</span></span>
<span><span class="co">#&gt;     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 0.002<span style="text-decoration: underline;">32</span> 0.001<span style="text-decoration: underline;">95</span> <span style="text-decoration: underline;">33</span>526. <span style="text-decoration: underline;">33</span>544. -<span style="color: #BB0000; text-decoration: underline;">16</span><span style="color: #BB0000;">760.</span> 0.012<span style="text-decoration: underline;">5</span></span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span>, fit <span class="op">=</span> <span class="st">"gam"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">.aic</span><span class="op">:</span><span class="va">.loglik</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    gam(.abundance ~ s(age, bs = "cs"), method = "REML")</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;     .aic   .bic .loglik</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> <span style="text-decoration: underline;">33</span>520. <span style="text-decoration: underline;">33</span>542. -<span style="color: #BB0000; text-decoration: underline;">16</span><span style="color: #BB0000;">757.</span></span></span></code></pre></div>
<p>These values are calculated by the [broom::glance()] function.</p>
<ul>
<li>
<code>.r.sqr</code> - Coefficient of Determination
(<i>R</i><sup>2</sup>); range: 0-1, lower is better. Percent of
variation explained by the model.</li>
<li>
<code>.adj.r</code> - <i>R</i><sup>2</sup>, taking degrees of
freedom into account.</li>
<li>
<code>.aic</code> - Akaike Information Criterion; lower is better.
Preferred when using model for prediction.</li>
<li>
<code>.bic</code> - Bayesian Information Criterion; lower is better.
Preferred when using model for interpretation.</li>
<li>
<code>.loglik</code> - Log-Likelihood (negative values); higher
values (closer to zero) are better.</li>
<li>
<code>.fit.p</code> - P-value for observing this fit by chance;
range: 0-1, lower is better.</li>
</ul>
<p>Based on these metrics, the <code>fit = "lm"</code> model is
optimal.</p>
</div>
<div class="section level3">
<h3 id="means-at-specific-timepoints">Means at Specific Timepoints<a class="anchor" aria-label="anchor" href="#means-at-specific-timepoints"></a>
</h3>
<p>The previous section showed us that Firmicutes abundance decreases
with age (slope = -0.127). We can use <code>stats_table(at = ...)</code>
to find the estimated marginal mean of <code>.abundance</code> at any
value of age.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">90</span>, <span class="fl">150</span>, <span class="fl">210</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">age</span>, <span class="va">.mean</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ age)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 2</span></span></span>
<span><span class="co">#&gt;     age .mean</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>    30  194.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>    90  187.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>   150  179.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>   210  171.</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pairwise-contrasts">Pairwise Contrasts<a class="anchor" aria-label="anchor" href="#pairwise-contrasts"></a>
</h3>
<p>Let’s check if the abundance of Firmicutes is different for babies
born vaginally vs via Cesarean.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, stat.by <span class="op">=</span> <span class="st">"delivery"</span>, at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">30</span>, <span class="fl">90</span>, <span class="fl">150</span>, <span class="fl">210</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">age</span>, <span class="va">delivery</span>, <span class="va">.mean.diff</span>, <span class="va">.h1</span>, <span class="va">.effect.size</span>, <span class="va">.p.val</span>, <span class="va">.adj.p</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ age * delivery)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 7</span></span></span>
<span><span class="co">#&gt;     age delivery           .mean.diff .h1   .effect.size   .p.val   .adj.p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>    30 Cesarean - Vaginal     -<span style="color: #BB0000;">33.3</span>  != 0       -<span style="color: #BB0000;">0.268</span>  0.000<span style="text-decoration: underline;">167</span> 0.000<span style="text-decoration: underline;">666</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>    90 Cesarean - Vaginal     -<span style="color: #BB0000;">17.4</span>  != 0       -<span style="color: #BB0000;">0.140</span>  0.003<span style="text-decoration: underline;">14</span>  0.006<span style="text-decoration: underline;">28</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>   210 Cesarean - Vaginal      14.6  != 0        0.117  0.121    0.161   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>   150 Cesarean - Vaginal      -<span style="color: #BB0000;">1.40</span> != 0       -<span style="color: #BB0000;">0.011</span><span style="color: #BB0000; text-decoration: underline;">2</span> 0.820    0.820</span></span></code></pre></div>
<p>Here we see why the <code>at</code> parameter is so important; at day
30, the difference is very significant, while at day 150, the difference
is negligible.</p>
<p>A few new columns are introduced here. <code>.mean.diff</code> is the
differences in mean abundance. The negative values indicate there is
more Firmicutes in the babies delivered vaginally. See how the pair is
listed as <code>Cesarean - Vaginal</code>? You can think “Cesarean
abundance” <strong>minus</strong> “Vaginal abundance”. In this table,
the <code>.h1</code> column tells us the alternate hypothesis is that
the <code>.mean.diff</code> is not equal to zero.</p>
<p>The <code>.adj.p</code> column is the <code>.p.val</code> corrected
for multiple comparisons. Since we looked at four timepoints, the
chances of encountering a randomly significant difference is higher than
if we looked at just one timepoint. The default
<code>stats_table(p.adj = 'fdr')</code> uses the Benjamini &amp;
Hochberg (1995) correction algorithm.</p>
<p>The <code>.effect.size</code> column gives us an idea of the
practical significance of <code>.mean.diff</code>. Calculated as
<code>.mean.diff / standard_deviation</code>, a rough interpretation is
0.2 = small effect, 0.5 = moderate effect, and 0.8 = large effect.</p>
<p>Considering this, we can say that babies born vaginally had
significantly more Firmicutes in their stool than babies born via
Cesarean. However, this difference is fairly small.</p>
</div>
<div class="section level3">
<h3 id="pairwise-trends">Pairwise Trends<a class="anchor" aria-label="anchor" href="#pairwise-trends"></a>
</h3>
<p>The above differences were actually the result of calculating
separate models for each level of delivery. These underlying trendlines
can be interesting in their own right as well. The slopes might be
different, or they might be the same but with different y-intercepts. We
can find out like this:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, stat.by <span class="op">=</span> <span class="st">"delivery"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">delivery</span>, <span class="va">.slope.diff</span>, <span class="va">.h1</span>, <span class="va">.p.val</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ age * delivery)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 4</span></span></span>
<span><span class="co">#&gt;   delivery           .slope.diff .h1     .p.val</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Cesarean - Vaginal       0.266 != 0  0.000<span style="text-decoration: underline;">995</span></span></span></code></pre></div>
<p>This shows us that the linear trendline slopes are significantly
different. The new column <code>.slope.diff</code> is the difference in
slopes for the two delivery group trendlines.</p>
<p>But what if we want to know the slope of those two trendlines by
themselves? We can find those by changing
<code>stats_table(stat.by = ...)</code> to
<code>stats_table(split.by = ...)</code>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, split.by <span class="op">=</span> <span class="st">"delivery"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">delivery</span>, <span class="va">.slope</span>, <span class="va">.h1</span>, <span class="va">.adj.p</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ age)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 4</span></span></span>
<span><span class="co">#&gt;   delivery  .slope .h1       .adj.p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Vaginal  -<span style="color: #BB0000;">0.192</span>  != 0  0.000<span style="text-decoration: underline;">009</span>55</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Cesarean  0.074<span style="text-decoration: underline;">4</span> != 0  0.212</span></span></code></pre></div>
<p>This reveals that when babies are born via Cesarean, their Firmicutes
abundance does NOT change as they get older. But babies born vaginally
DO see a decrease over time in Firmicutes abundance. Interesting!</p>
<p>We can check the y-intercepts of the trendlines by checking the mean
abundances at age = 0.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="va">df</span>, regr <span class="op">=</span> <span class="st">"age"</span>, split.by <span class="op">=</span> <span class="st">"delivery"</span>, at <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">age</span>, <span class="va">delivery</span>, <span class="va">.mean</span>, <span class="va">.lower</span>, <span class="va">.upper</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    lm(.abundance ~ age)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 5</span></span></span>
<span><span class="co">#&gt;     age delivery .mean .lower .upper</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>     0 Vaginal   208.   197.   219.</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>     0 Cesarean  167.   151.   183.</span></span></code></pre></div>
<p>Here we see that the means (and their 95% confidence interval,
<code>.lower</code> - <code>.upper</code>) are quite different.
Therefore, at birth, babies born vaginally have more Firmcutes in their
stool than babies born via Cesarean.</p>
<ul>
<li>
<code>stat.by</code> - An optional categorical variable defining the
statistical groups. If set, this triggers a pairwise comparison of all
<code>stat.by</code> groups.</li>
<li>
<code>split.by</code> - Optional field(s) to pre-partition the
dataset by.</li>
<li>
<code>test</code> - The options are <code>'fits'</code>,
<code>'means'</code>, or <code>'slopes'</code>. <code>'fits'</code> will
return goodness-of-fit metrics. <code>'means'</code> calculates the
estimated marginal mean of <code>resp</code> at <code>at</code>.
<code>'means'</code> estimates the trendline slope at <code>at</code>.
More detail on these later.</li>
<li>
<code>fit</code> - For regression, the choices are
<code>'lm'</code>, <code>'log'</code>, or <code>'gam'</code>. These fit
straight trend-lines, log-curved trend-lines, and multi-spline
trend-lines to the data, respectively.</li>
<li>
<code>level</code> - The confidence level for calculating a
confidence interval.</li>
<li>
<code>alt</code> - Alternative hypothesis direction. Options are
<code>'!='</code> (two-sided; not equal to <code>mu</code>),
<code>'&lt;'</code> (less than <code>mu</code>), or <code>'&gt;'</code>
(greater than <code>mu</code>).</li>
<li>
<code>mu</code> - Reference value to test against.</li>
<li>
<code>at</code> - Calculate means or slopes at this
<code>regr</code> value, or <code>NULL</code> to use the median of
<code>regr</code>’s values.</li>
<li>
<code>p.adj</code> - Method to use for multiple comparisons
adjustment of p-values. Run <code>p.adjust.methods</code> for a list of
available options.</li>
</ul>
<p>Rbiom has three pre-defined regression models:</p>
<p><img src="regression_files/figure-html/unnamed-chunk-25-1.png" width="768"></p>
<p>p(“lm”, y ~ x, “stats::lm(y ~ x)”), p(“log”, y ~ log(x), “stats::lm(y
~ log(x))”), p(“gam”, y ~ s(x, bs = “cs”), “mgcv::gam(y ~ s(x, bs =
‘cs’),= ‘REML’)”)</p>
<p>Visualizations are one of the best ways to identify correlations in
your dataset. If you can see a trend with your eyes, then you’re on the
right track. Rbiom’s plotting functions are</p>
<p>The <code>*_boxplot()</code>, <code>*_corrplot()</code>, and
<code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code> functions will automatically add p-values
to your figures whenever possible. The ggplot object they return has
<code>$data</code>, <code>$code</code>, <code>$stats</code>, and
<code>$stats$code</code> attributes you can use to automate, reproduce,
and customize your figures.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
