<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rbiom">
<title>Statistics • rbiom</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Statistics">
<meta property="og:description" content="rbiom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9117</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rbiom.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6>
    <a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a>
    <a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a>
    <a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a>
    <a class="dropdown-item" href="../articles/diversity.html">Diversity Metrics</a>
    <a class="dropdown-item" href="../articles/statistics.html">Statistics</a>
    <a class="dropdown-item" href="../reference/speed.html">Speed Ups</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Plotting</h6>
    <a class="dropdown-item" href="../articles/plots.html">Plot Types</a>
    <a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a>
    <a class="dropdown-item" href="../articles/scales.html">Scales and Transformations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Statistics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/HEAD/vignettes/articles/statistics.Rmd" class="external-link"><code>vignettes/articles/statistics.Rmd</code></a></small>
      <div class="d-none name"><code>statistics.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Visualizations are one of the best ways to identify correlations in
your dataset. If you can see a trend with your eyes, then you’re on the
right track. Rbiom’s plotting functions are</p>
<table class="table table-borderless">
<tr>
<td>
<code><a href="../reference/adiv_boxplot.html">adiv_boxplot()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_corrplot.html">bdiv_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/rare_barplot.html">rare_barplot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_boxplot.html">taxa_boxplot()</a></code>
</td>
</tr>
<tr>
<td>
<code><a href="../reference/adiv_corrplot.html">adiv_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_heatmap.html">bdiv_heatmap()</a></code>
</td>
<td>
<code><a href="../reference/rare_corrplot.html">rare_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_corrplot.html">taxa_corrplot()</a></code>
</td>
</tr>
<tr>
<td>
<code><a href="../reference/bdiv_boxplot.html">bdiv_boxplot()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_barplot.html">taxa_barplot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_heatmap.html">taxa_heatmap()</a></code>
</td>
</tr>
</table>
<p>The <code>*_boxplot()</code>, <code>*_corrplot()</code>, and
<code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code> functions will automatically add p-values
to your figures whenever possible. The ggplot object they return has
<code>$data</code>, <code>$code</code>, <code>$stats</code>, and
<code>$stats$code</code> attributes you can use to automate, reproduce,
and customize your figures.</p>
<p>To generate statistics without creating a plot, use one of the
following:</p>
<table class="table table-borderless"><tr>
<td>
<code><a href="../reference/adiv_stats.html">adiv_stats()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_stats.html">bdiv_stats()</a></code>
</td>
<td>
<code><a href="../reference/distmat_stats.html">distmat_stats()</a></code>
</td>
<td>
<code><a href="../reference/taxa_stats.html">taxa_stats()</a></code>
</td>
</tr></table>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>Your metadata field of interest and microbiome property will
determine which rbiom function to use.</p>
<table class="table">
<tr>
<th rowspan="2" class="border-0 border-end border-bottom align-middle">
Metadata<br>Property
</th>
<th colspan="3" class="border-0" align="center">
Microbiome Property
</th>
</tr>
<tr>
<th class="border-bottom">
Alpha Diversity<br><span class="eg">Shannon, Simpson</span>
</th>
<th class="border-bottom">
Beta Diversity<br><span class="eg">UniFrac, Jaccard</span>
</th>
<th class="border-bottom">
Taxa Abundance<br><span class="eg">Phylum, Genus</span>
</th>
</tr>
<tr>
<th class="border-end">
Categorical<br><span class="eg">Sex, Body Site</span>
</th>
<td>
<code><a href="../reference/adiv_boxplot.html">adiv_boxplot()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_boxplot.html">bdiv_boxplot()</a></code><br><code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_boxplot.html">taxa_boxplot()</a></code>
</td>
</tr>
<tr>
<th class="border-end">
Numeric<br><span class="eg">Age, BMI</span>
</th>
<td>
<code><a href="../reference/adiv_corrplot.html">adiv_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_corrplot.html">bdiv_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_corrplot.html">taxa_corrplot()</a></code>
</td>
</tr>
<tr>
<th class="border-end">
<i>Any</i>
</th>
<td>
<code><a href="../reference/adiv_stats.html">adiv_stats()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_stats.html">bdiv_stats()</a></code>
</td>
<td>
<code><a href="../reference/taxa_stats.html">taxa_stats()</a></code>
</td>
</tr>
</table>
<p>For instance, to explore the effect of Body Site (a categorical
metadata field) on Shannon Diversity (an alpha diversity metric), we’d
use <code><a href="../reference/adiv_boxplot.html">adiv_boxplot()</a></code> to produce a plot with statistics, or
<code><a href="../reference/adiv_stats.html">adiv_stats()</a></code> if we only want the stats.</p>
<p>Although all the functions have important differences, the
statistical methods they employ can be grouped by the three plot types:
ordination plot, box plot, and correlation plot.</p>
</div>
<div class="section level2">
<h2 id="ordination-plots">Ordination Plots<a class="anchor" aria-label="anchor" href="#ordination-plots"></a>
</h2>
<p>Statistics for ordination plots are the most straight-forward. Set a
categorical metadata field to the <code>color.by</code> parameter to
test whether inter-sample distances are correlated with that
variable.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot</a></span><span class="op">(</span></span>
<span>  biom     <span class="op">=</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span>, </span>
<span>  color.by <span class="op">=</span> <span class="st">"Body Site"</span>, </span>
<span>  bdiv     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Jaccard"</span>, <span class="st">"Bray-Curtis"</span><span class="op">)</span>, </span>
<span>  ord      <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PCoA"</span>, <span class="st">"UMAP"</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="statistics_files/figure-html/unnamed-chunk-4-1.png" width="768"></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Test:     adonis2 ~ `Body Site`. 999 permutations.</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 7</span></span></span>
<span><span class="co">#&gt;   .weighted .bdiv          .n .stat    .z .p.val .adj.p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> TRUE      Jaccard        49  11.3  51.6  0.001  0.001</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> TRUE      Bray-Curtis    49  19.5  72.0  0.001  0.001</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">stats</span><span class="op">$</span><span class="va">code</span></span>
<span><span class="co">#&gt; iters   <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">list</span><span style="color: #BBBB00;">(</span>weighted = <span style="color: #BB00BB;">TRUE</span>, bdiv = <span style="color: #00BBBB;">c</span><span style="color: #BB00BB;">(</span><span style="color: #BBBB00;">"Jaccard"</span>, <span style="color: #BBBB00;">"Bray-Curtis"</span><span style="color: #BB00BB;">)</span><span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; dm_list <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">blply</span><span style="color: #BBBB00;">(</span>biom, <span style="color: #BB00BB; font-weight: bold;">NULL</span>, bdiv_distmat, iters = iters, prefix = <span style="color: #BB00BB;">TRUE</span><span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; stats   <span style="color: #00BB00;">&lt;-</span> plyr::<span style="color: #00BBBB;">ldply</span><span style="color: #BBBB00;">(</span>dm_list, <span style="color: #BB0000;">function</span> <span style="color: #BB00BB;">(</span>dm<span style="color: #BB00BB;">)</span> <span style="color: #BB00BB;">{</span></span></span>
<span><span class="co">#&gt;   groups <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">pull(</span>biom, <span style="color: #BBBB00;">"Body Site"</span><span style="color: #00BBBB;">)[attr</span><span style="color: #BBBB00;">(</span>dm, <span style="color: #BBBB00;">'Labels')</span><span style="color: #00BBBB;">]</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">set.seed(</span><span style="color: #BB00BB;">0</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt;   ptest  <span style="color: #00BB00;">&lt;-</span> vegan::<span style="color: #00BBBB;">adonis2(</span>formula = dm <span style="color: #00BB00;">~</span> groups, permutations = <span style="color: #BB00BB;">999</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt;   pstats <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">summary(</span>vegan::<span style="color: #00BBBB;">permustats</span><span style="color: #BBBB00;">(</span>ptest<span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">with(</span>pstats, <span style="color: #00BBBB;">data.frame</span><span style="color: #BBBB00;">(</span>statistic, z, p<span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt; <span style="color: #BB00BB;">}</span><span style="color: #BBBB00;">)</span></span></span></code></pre></div>
<p>The plot subtitles have the summary statistics. Additionally,
<code>p$stats</code> contains a tibble data.frame with the full
statistics table, and <code>p$stats$code</code> shows the R commands for
reproducing the statistics outside of rbiom.</p>
<p>Note that the ordination statistics are not dependent on the
ordination, only the distance metric. This is because the statistics are
based on beta diversity distances which are computed prior to
ordination.</p>
<p>By default, <code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code> applies the perMANOVA test.
You can change this to MRPP by specifying <code>test="mrpp"</code>.
Details on the available tests are below.</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="20%">
<col width="68%">
</colgroup>
<thead><tr class="header">
<th>Test</th>
<th>Function</th>
<th>Method</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>adonis2</code></td>
<td><code><a href="https://rdrr.io/pkg/vegan/man/adonis.html" class="external-link">vegan::adonis2()</a></code></td>
<td>Permutational Multivariate Analysis of Variance (perMANOVA)</td>
</tr>
<tr class="even">
<td><code>mrpp</code></td>
<td><code><a href="https://rdrr.io/pkg/vegan/man/mrpp.html" class="external-link">vegan::mrpp()</a></code></td>
<td>Multiple Response Permutation Procedure (MRPP)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="box-plots">Box Plots<a class="anchor" aria-label="anchor" href="#box-plots"></a>
</h2>
<p>Statistics on box plots will automatically toggle between pairwise
and group-wise statistics based on the values of <code>x</code> and
<code>color.by</code>: <code>x</code> controls pairwise and
<code>color.by</code> controls group-wise. You can set <code>x</code>
and <code>color.by</code> to the same categorical metadata field to get
colored pairwise statistics, or set them to different categorical
metadata fields to get multiple group-wise statistics per plot.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">`Body Site`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Saliva'</span>, <span class="st">'Stool'</span>, <span class="st">'Buccal mucosa'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, color.by <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"Body Site"</span>, color.by <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"Sex"</span>, color.by <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">p1</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'x = "Body Site"'</span><span class="op">)</span>, </span>
<span>  <span class="va">p2</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'color.by = "Body Site"'</span><span class="op">)</span>, </span>
<span>  <span class="va">p3</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'x = color.by = "Body Site"'</span><span class="op">)</span>, </span>
<span>  <span class="va">p4</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'x = "Sex", color.by = "Body Site"'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">`+`</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span>, caption <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="statistics_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<p>Above, the two plots on the left are annotated with pairwise
statistics while the two on the right have group-wise statistics. As
with other plots, you can find the full statistics tables and
reproducible R code in the plot attributes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Test:     wilcox.test(.diversity ~ `Body Site`).</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 3 × 8</span></span></span>
<span><span class="co">#&gt;   `Body Site`               .n .stat .estimate .lower .upper    .p.val    .adj.p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>                  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Buccal mucosa - Saliva    20     1    -<span style="color: #BB0000;">1.68</span>  -<span style="color: #BB0000;">2.19</span>  -<span style="color: #BB0000;">0.982</span> 0.000<span style="text-decoration: underline;">021</span>7 0.000<span style="text-decoration: underline;">065</span>1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Buccal mucosa - Stool     19     8    -<span style="color: #BB0000;">1.16</span>  -<span style="color: #BB0000;">1.7</span>   -<span style="color: #BB0000;">0.447</span> 0.001<span style="text-decoration: underline;">45</span>   0.002<span style="text-decoration: underline;">17</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Saliva - Stool            19    80     0.526  0.186  0.824 0.002<span style="text-decoration: underline;">99</span>   0.002<span style="text-decoration: underline;">99</span></span></span>
<span></span>
<span><span class="va">p2</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Test:     kruskal.test(.diversity ~ `Body Site`).</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 5</span></span></span>
<span><span class="co">#&gt;      .n .stat   .df    .p.val    .adj.p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>    29  19.9     2 0.000<span style="text-decoration: underline;">046</span>8 0.000<span style="text-decoration: underline;">046</span>8</span></span>
<span></span>
<span><span class="va">p2</span><span class="op">$</span><span class="va">stats</span><span class="op">$</span><span class="va">code</span></span>
<span><span class="co">#&gt; data  <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">adiv_table</span><span style="color: #BBBB00;">(</span>biom<span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; stats <span style="color: #00BB00;">&lt;-</span> stats::<span style="color: #00BBBB;">kruskal.test</span><span style="color: #BBBB00;">(</span>.diversity <span style="color: #00BB00;">~</span> `Body Site`, data = data<span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">with</span><span style="color: #BBBB00;">(</span><span style="color: #00BBBB;">data.frame</span><span style="color: #BB00BB;">(</span></span></span>
<span><span class="co">#&gt;    .n     = <span style="color: #00BBBB;">nrow(</span>data<span style="color: #00BBBB;">)</span>, </span></span>
<span><span class="co">#&gt;    .stat  = statistic, </span></span>
<span><span class="co">#&gt;    .df    = parameter, </span></span>
<span><span class="co">#&gt;    .p.val = p.value <span style="color: #BB00BB;">)</span><span style="color: #BBBB00;">)</span></span></span></code></pre></div>
<p>Internally, rbiom uses the non-parametric functions listed below.</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="25%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>Test</th>
<th>Function</th>
<th>Method</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>pairwise</td>
<td><code><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">stats::wilcox.test()</a></code></td>
<td>Two-sample Wilcoxon Rank Sum Test, aka Mann-Whitney Test</td>
</tr>
<tr class="even">
<td>group-wise</td>
<td><code><a href="https://rdrr.io/r/stats/kruskal.test.html" class="external-link">stats::kruskal.test()</a></code></td>
<td>Kruskal-Wallis Rank Sum Test</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<div class="section level3">
<h3 id="normality">Normality<a class="anchor" aria-label="anchor" href="#normality"></a>
</h3>
<p>A normal distribution is visualized as a “bell curve”, where values
further from the mean are observed less often. Microbial abundances do
not follow this pattern; it’s common to observe high or low abundances
more often than a “medium” abundance.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>  widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Normal Distribution"</span><span class="op">)</span>,</span>
<span>    </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="../reference/taxa_matrix.html">taxa_table</a></span><span class="op">(</span><span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span>, taxa <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">.abundance</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="st">".taxa"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Genera Abundance Distributions"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="statistics_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p>To compensate for this non-normality, rbiom uses the following
non-parametric tests for categorical variables that are based on ranking
or permutations.</p>
<table class="table">
<colgroup>
<col width="31%">
<col width="31%">
<col width="36%">
</colgroup>
<thead><tr class="header">
<th>Test</th>
<th>Function</th>
<th>Used For</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Wilcoxon Rank-Sum</td>
<td><code><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">stats::wilcox.test()</a></code></td>
<td>Pairwise boxplot</td>
</tr>
<tr class="even">
<td>Kruskal-Wallis Rank Sum</td>
<td><code><a href="https://rdrr.io/r/stats/kruskal.test.html" class="external-link">stats::kruskal.test()</a></code></td>
<td>Groupwise boxplot</td>
</tr>
<tr class="odd">
<td>Permutational MANOVA</td>
<td><code><a href="https://rdrr.io/pkg/vegan/man/adonis.html" class="external-link">vegan::adonis2()</a></code></td>
<td>
<code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code> clusters</td>
</tr>
</tbody>
</table>
<p>For correlation/regression analysis, rbiom uses estimated marginal
means (<code><a href="https://rdrr.io/pkg/emmeans/man/emmeans.html" class="external-link">emmeans::emmeans()</a></code>) on rank-transformed values.</p>
<p>Further reading:</p>
<ul>
<li>
<a href="https://uw.pressbooks.pub/appliedmultivariatestatistics/" class="external-link">Applied
Multivariate Statistics in R</a>: <a href="https://uw.pressbooks.pub/appliedmultivariatestatistics/chapter/permanova/" class="external-link">PERMANOVA</a>,
<a href="https://uw.pressbooks.pub/appliedmultivariatestatistics/chapter/comparison-of-techniques/" class="external-link">Comparison
of Techniques</a>
</li>
<li>
<a href="https://medium.com/datasciencearth/r-applications-part-1-simple-linear-regression-ef5a0e19a05d" class="external-link">R
Applications — Part 1: Simple Linear Regression</a>, particularly the
section on <a href="https://medium.com/datasciencearth/r-applications-part-1-simple-linear-regression-ef5a0e19a05d#a6f6" class="external-link">assumption
control</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="compositionality">Compositionality<a class="anchor" aria-label="anchor" href="#compositionality"></a>
</h3>
<p>Compositional data arises when the counts don’t represent the entire
population. In microbiome studies, the number of microbes that get
sequenced is far less than the number of microbes from where the sample
was collected. Articles by <a href="https://doi.org/10.3389/fmicb.2017.02224" class="external-link">Gloor et al</a> and <a href="https://doi.org/10.1371/journal.pcbi.1003531" class="external-link">McMurdie and
Holmes</a> propose the use of their analysis tools ( <a href="https://doi.org/doi:10.18129/B9.bioc.ALDEx2" class="external-link">ALDEx2</a> and <a href="https://doi.org/doi:10.18129/B9.bioc.metagenomeSeq" class="external-link">metagenomeSeq</a>,
respectively) to apply the proper statistical methods for this
situation. Conversely, rbiom does not correct for compositionality. This
is because correcting for compositionality introduces extra noise into
your dataset and severely limits your selection of metrics and
visualizations, typically without any significant benefit to
analysis.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Daniel P. Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
