<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rbiom">
<title>Statistics • rbiom</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Statistics">
<meta property="og:description" content="rbiom">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9134</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rbiom.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6>
    <a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a>
    <a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a>
    <a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a>
    <a class="dropdown-item" href="../articles/statistics.html">Statistics</a>
    <a class="dropdown-item" href="../articles/regression.html">Regression</a>
    <a class="dropdown-item" href="../reference/speed.html">Speed Ups</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Plotting</h6>
    <a class="dropdown-item" href="../articles/plots.html">Plot Types</a>
    <a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Statistics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/HEAD/vignettes/articles/statistics.Rmd" class="external-link"><code>vignettes/articles/statistics.Rmd</code></a></small>
      <div class="d-none name"><code>statistics.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quick-start">Quick Start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The rbiom package includes many statistical functions. If you have an
rbiom object, you can use dedicated functions for alpha diversity, beta
diversity, and taxa abundance. Otherwise, look to the generic functions
that operate on any data.frame or distance matrix.</p>
<div class="section level3">
<h3 id="rbiom-object-functions">rbiom Object Functions<a class="anchor" aria-label="anchor" href="#rbiom-object-functions"></a>
</h3>
<p>Your metadata field and microbiome property of interest will
determine which rbiom function to use.</p>
<table class="table">
<tr>
<th rowspan="2" class="border-0 border-end border-bottom align-middle">
Metadata<br>Field
</th>
<th colspan="3" class="border-0" align="center">
Microbiome Property
</th>
</tr>
<tr>
<th class="border-bottom">
Alpha Diversity<br><span class="eg">Shannon, Simpson</span>
</th>
<th class="border-bottom">
Beta Diversity<br><span class="eg">UniFrac, Jaccard</span>
</th>
<th class="border-bottom">
Taxa Abundance<br><span class="eg">Phylum, Genus</span>
</th>
</tr>
<tr>
<th class="border-end">
Categorical<br><span class="eg">Sex, Body Site</span>
</th>
<td>
<code><a href="../reference/adiv_boxplot.html">adiv_boxplot()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_boxplot.html">bdiv_boxplot()</a></code><br><code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_boxplot.html">taxa_boxplot()</a></code>
</td>
</tr>
<tr>
<th class="border-end">
Numeric<br><span class="eg">Age, BMI</span>
</th>
<td>
<code><a href="../reference/adiv_corrplot.html">adiv_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_corrplot.html">bdiv_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/taxa_corrplot.html">taxa_corrplot()</a></code>
</td>
</tr>
<tr>
<th class="border-end">
<i>Any</i>
</th>
<td>
<code><a href="../reference/adiv_stats.html">adiv_stats()</a></code>
</td>
<td>
<code><a href="../reference/bdiv_stats.html">bdiv_stats()</a></code>
</td>
<td>
<code><a href="../reference/taxa_stats.html">taxa_stats()</a></code>
</td>
</tr>
</table>
<p>For instance, to explore the effect of Body Site (a categorical
metadata field) on Shannon Diversity (an alpha diversity metric), we’d
use <code><a href="../reference/adiv_boxplot.html">adiv_boxplot()</a></code> to produce a plot with statistics, or
<code><a href="../reference/adiv_stats.html">adiv_stats()</a></code> if we only want the stats.</p>
</div>
<div class="section level3">
<h3 id="generic-functions">Generic Functions<a class="anchor" aria-label="anchor" href="#generic-functions"></a>
</h3>
<p>If your data is in a data.frame (or tibble), use:</p>
<table class="table table-borderless"><tr>
<td>
<code><a href="../reference/stats_boxplot.html">stats_boxplot()</a></code>
</td>
<td>
<code><a href="../reference/stats_corrplot.html">stats_corrplot()</a></code>
</td>
<td>
<code><a href="../reference/stats_table.html">stats_table()</a></code>
</td>
</tr></table>
<p>Or, for a distance matrix:</p>
<table class="table table-borderless"><tr>
<td>
<code><a href="../reference/distmat_stats.html">distmat_stats()</a></code>
</td>
</tr></table>
</div>
</div>
<div class="section level2">
<h2 id="statistics-table">Statistics Table<a class="anchor" aria-label="anchor" href="#statistics-table"></a>
</h2>
<p>The function <code><a href="../reference/stats_table.html">stats_table()</a></code> and functions ending in
“_stats” (e.g. <code><a href="../reference/adiv_stats.html">adiv_stats()</a></code>, <code><a href="../reference/distmat_stats.html">distmat_stats()</a></code>)
will return a statistics table. Functions with “plot” in their name
(e.g. <code><a href="../reference/adiv_boxplot.html">adiv_boxplot()</a></code>, <code><a href="../reference/stats_corrplot.html">stats_corrplot()</a></code>) will
return a plot object <code>p</code>. You can access the plot’s
associated statistics table with <code>p$stats</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span></span>
<span>  biom     <span class="op">=</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span>,           <span class="co"># Dataset as an rbiom object</span></span>
<span>  adiv     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Shannon"</span>, <span class="st">"Simpson"</span><span class="op">)</span>, <span class="co"># Alpha diversity metrics</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">"Body Site"</span>,             <span class="co"># Statistical groups</span></span>
<span>  facet.by <span class="op">=</span> <span class="st">"Sex"</span> <span class="op">)</span>                  <span class="co"># Split data prior to stats</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    kruskal.test(.diversity ~ `Body Site`)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 8</span></span></span>
<span><span class="co">#&gt;   Sex    .adiv   .stat .h1      .p.val   .adj.p    .n   .df</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Female Simpson  24.5 &gt; 0   0.000<span style="text-decoration: underline;">062</span>0 0.000<span style="text-decoration: underline;">156</span>    30     4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Female Shannon  24.1 &gt; 0   0.000<span style="text-decoration: underline;">077</span>9 0.000<span style="text-decoration: underline;">156</span>    30     4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Male   Simpson  13.8 &gt; 0   0.003<span style="text-decoration: underline;">14</span>   0.003<span style="text-decoration: underline;">25</span>     19     3</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Male   Shannon  13.8 &gt; 0   0.003<span style="text-decoration: underline;">25</span>   0.003<span style="text-decoration: underline;">25</span>     19     3</span></span></code></pre></div>
<p>This table has the following information:</p>
<div class="section level3">
<h3 id="model">Model<a class="anchor" aria-label="anchor" href="#model"></a>
</h3>
<p>In the upper-left corner, we see
<code>Model:    kruskal.test(.diversity ~ `Body Site`)</code>. This
tells us that the underlying test used here is base R’s
<code><a href="https://rdrr.io/r/stats/kruskal.test.html" class="external-link">kruskal.test()</a></code>, grouping values of <code>.diversity</code>
by <code>`Body Site`</code>. In rbiom, <code>.diversity</code> is the
default name for the column containing alpha diversity values.</p>
</div>
<div class="section level3">
<h3 id="multiple-comparisons">Multiple Comparisons<a class="anchor" aria-label="anchor" href="#multiple-comparisons"></a>
</h3>
<p>The above table has four rows, meaning we ran data through the model
four times. In this case, it was two alpha diversity metrics times two
<code>Sex</code> facets. The first two columns of the table
(<code>Sex</code> and <code>.adiv</code>) tell us which row is for each
combination. Additionally, the <code>.adj.p</code> column is the
<code>.p.val</code> adjusted for multiple comparisons.</p>
</div>
<div class="section level3">
<h3 id="hypothesis">Hypothesis<a class="anchor" aria-label="anchor" href="#hypothesis"></a>
</h3>
<p>The three columns <code>.stat</code>, <code>.h1</code>, and
<code>.p.val</code> together show the hypothesis and its outcome. Above,
we’re asking if <code>.stat</code> is greater than zero. When
<code>.p.val</code> is less than 0.05 we can say that it is.</p>
<p>That’s the very simplified explanation.</p>
<p>Recall from your statistic classes that you are testing a null
hypothesis H<sub>0</sub> against an alternate hypothesis H<sub>1</sub>.
In this case, our null hypothesis is that the Kruskal-Wallis statistic
is zero (<code>.stat == 0</code>), indicating all
<code>`Body Site`</code> groups have similar alpha diversity. The
p-value is the probability that the null hypothesis is correct (a
p-value of 0.6 is interpreted as a 60% chance that
<code>.stat == 0</code>). When the p-value is below a certain value
(usually 0.05) we accept the alternative hypothesis instead - in this
case that <code>.stat &gt; 0</code>, meaning <code>.diversity</code>
does vary by <code>`Body Site`</code>.</p>
</div>
<div class="section level3">
<h3 id="terms-of-interest">Terms of Interest<a class="anchor" aria-label="anchor" href="#terms-of-interest"></a>
</h3>
<p>The last two columns are:</p>
<ul>
<li>
<code>.n</code> - The number of samples. After rarefying, we have 30
females and 19 males.</li>
<li>
<code>.df</code> - Degrees of freedom. One less than the number of
groups. The HMP50 dataset has five body sites: Anterior nares, buccal
mucosa, mid vagina, saliva, and stool. However, the male facets don’t
have any mid vagina samples.</li>
</ul>
</div>
<div class="section level3">
<h3 id="data-and-code">Data and Code<a class="anchor" aria-label="anchor" href="#data-and-code"></a>
</h3>
<p>You can inspect the data passed to <code><a href="https://rdrr.io/r/stats/kruskal.test.html" class="external-link">kruskal.test()</a></code> by
looking at <code>p$data</code>. Additionally, the code used to generate
the statistics is available in <code>p$stats$code</code>. These
attributes can help you reproduce and customize your data analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">data</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 98 × 6</span></span></span>
<span><span class="co">#&gt;   .sample .depth .adiv   .diversity `Body Site`   Sex   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> HMP01     <span style="text-decoration: underline;">1</span>183 Shannon       1.76 Buccal mucosa Female</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> HMP02     <span style="text-decoration: underline;">1</span>183 Shannon       2.60 Buccal mucosa Male  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> HMP03     <span style="text-decoration: underline;">1</span>183 Shannon       2.92 Saliva        Male  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> HMP04     <span style="text-decoration: underline;">1</span>183 Shannon       3.26 Saliva        Male  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> HMP05     <span style="text-decoration: underline;">1</span>183 Shannon       1.40 Buccal mucosa Female</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 93 more rows</span></span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">stats</span><span class="op">$</span><span class="va">code</span></span>
<span><span class="co">#&gt; data <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">adiv_table</span><span style="color: #BBBB00;">(</span>biom, <span style="color: #00BBBB;">c</span><span style="color: #BB00BB;">(</span><span style="color: #BBBB00;">"Shannon"</span>, <span style="color: #BBBB00;">"Simpson"</span><span style="color: #BB00BB;">)</span>, <span style="color: #00BBBB;">c</span><span style="color: #BB00BB;">(</span><span style="color: #BBBB00;">"Body Site"</span>, <span style="color: #BBBB00;">"Sex"</span><span style="color: #BB00BB;">)</span><span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data <span style="color: #00BB00;">%&lt;&gt;%</span> dplyr::<span style="color: #00BBBB;">rename</span><span style="color: #BBBB00;">(</span></span></span>
<span><span class="co">#&gt;   .resp    = <span style="color: #BBBB00;">".diversity"</span>, </span></span>
<span><span class="co">#&gt;   .stat.by = <span style="color: #BBBB00;">"Body Site"</span> <span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; stats <span style="color: #00BB00;">&lt;-</span> plyr::<span style="color: #00BBBB;">ddply</span><span style="color: #BBBB00;">(</span>data, <span style="color: #00BBBB;">.</span><span style="color: #BB00BB;">(</span>Sex, .adiv<span style="color: #BB00BB;">)</span>, <span style="color: #BB0000;">function</span> <span style="color: #BB00BB;">(</span>data<span style="color: #BB00BB;">)</span> <span style="color: #BB00BB;">{</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">tryCatch(</span>error = <span style="color: #BB0000;">function</span> <span style="color: #BBBB00;">(</span>e<span style="color: #BBBB00;">)</span> <span style="color: #00BBBB;">data.frame</span><span style="color: #BBBB00;">()[</span><span style="color: #BB00BB;">1</span>,<span style="color: #BBBB00;">]</span>, <span style="color: #00BBBB;">suppressWarnings</span><span style="color: #BBBB00;">(</span><span style="color: #BB00BB;">{</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     data <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;       stats::<span style="color: #00BBBB;">kruskal.test(</span>.resp <span style="color: #00BB00;">~</span> .stat.by, .<span style="color: #00BBBB;">)</span> <span style="color: #00BB00;">%&gt;%</span></span></span>
<span><span class="co">#&gt;       <span style="color: #00BBBB;">with(tibble</span><span style="color: #BBBB00;">(</span></span></span>
<span><span class="co">#&gt;         .stat  = statistic, </span></span>
<span><span class="co">#&gt;         .h1    = <span style="color: #00BBBB;">factor</span><span style="color: #BB00BB;">(</span><span style="color: #BBBB00;">'&gt; 0'</span><span style="color: #BB00BB;">)</span>, </span></span>
<span><span class="co">#&gt;         .p.val = p.value, </span></span>
<span><span class="co">#&gt;         .n     = <span style="color: #00BBBB;">nrow</span><span style="color: #BB00BB;">(</span>data<span style="color: #BB00BB;">)</span>, </span></span>
<span><span class="co">#&gt;         .df    = parameter <span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #BB00BB;">}</span><span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt; <span style="color: #BB00BB;">}</span><span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">as_tibble</span><span style="color: #BBBB00;">()</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   dplyr::<span style="color: #00BBBB;">mutate</span><span style="color: #BBBB00;">(</span>.adj.p = <span style="color: #00BBBB;">p.adjust</span><span style="color: #BB00BB;">(</span>.p.val, <span style="color: #BBBB00;">'fdr'</span><span style="color: #BB00BB;">)</span>, .after = .p.val<span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   dplyr::<span style="color: #00BBBB;">arrange</span><span style="color: #BBBB00;">(</span>.p.val<span style="color: #BBBB00;">)</span></span></span></code></pre></div>
<p>When generating statistics with <code>st &lt;- adiv_stats()</code>,
<code><a href="../reference/bdiv_stats.html">bdiv_stats()</a></code>, or <code><a href="../reference/taxa_stats.html">taxa_stats()</a></code>, the data and
code are in <code>st$data</code> and <code>st$code</code>, respectively.
For <code>st &lt;- stats_table(df = df)</code>, the data and code are in
<code>df</code> and <code>st$code</code>, respectively.</p>
</div>
<div class="section level3">
<h3 id="column-reference">Column Reference<a class="anchor" aria-label="anchor" href="#column-reference"></a>
</h3>
<p>Care has been taken to keep rbiom’s statistics tables consistent
across functions. However, some tables will provide more information
when it is available from the underlying statistical function.</p>
<p>Below is a quick reference guide to all columns that may appear in an
rbiom statistics table.</p>
<table class="table">
<colgroup>
<col width="20%">
<col width="79%">
</colgroup>
<thead><tr class="header">
<th><strong>Field</strong></th>
<th><strong>Description</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>.stat</code></td>
<td>Wilcoxon or Kruskal-Wallis rank sum statistic.</td>
</tr>
<tr class="even">
<td><code>.mean</code></td>
<td>Estimated marginal mean. See <code><a href="https://rvlenth.github.io/emmeans/reference/emmeans.html" class="external-link">emmeans::emmeans()</a></code>.</td>
</tr>
<tr class="odd">
<td><code>.mean.diff</code></td>
<td>Difference in means.</td>
</tr>
<tr class="even">
<td><code>.slope</code></td>
<td>Trendline slope. See <code><a href="https://rvlenth.github.io/emmeans/reference/emtrends.html" class="external-link">emmeans::emtrends()</a></code>.</td>
</tr>
<tr class="odd">
<td><code>.slope.diff</code></td>
<td>Difference in slopes.</td>
</tr>
<tr class="even">
<td><code>.h1</code></td>
<td>Alternate hypothesis.</td>
</tr>
<tr class="odd">
<td><code>.p.val</code></td>
<td>Probability that null hypothesis is correct.</td>
</tr>
<tr class="even">
<td><code>.adj.p</code></td>
<td>
<code>.p.val</code> after adjusting for multiple comparisons.</td>
</tr>
<tr class="odd">
<td><code>.effect.size</code></td>
<td>Effect size. See <code><a href="https://rvlenth.github.io/emmeans/reference/eff_size.html" class="external-link">emmeans::eff_size()</a></code>.</td>
</tr>
<tr class="even">
<td><code>.lower</code></td>
<td>Confidence interval lower bound.</td>
</tr>
<tr class="odd">
<td><code>.upper</code></td>
<td>Confidence interval upper bound.</td>
</tr>
<tr class="even">
<td><code>.se</code></td>
<td>Standard error.</td>
</tr>
<tr class="odd">
<td><code>.n</code></td>
<td>Number of samples.</td>
</tr>
<tr class="even">
<td><code>.df</code></td>
<td>Degrees of freedom.</td>
</tr>
<tr class="odd">
<td><code>.t.ratio</code></td>
<td>
<em>(.mean, .mean.diff, .slope, or .slope.diff)</em> /
<code>.se</code>
</td>
</tr>
<tr class="even">
<td><code>.z</code></td>
<td>Std. effect size. See <code><a href="https://rdrr.io/pkg/vegan/man/permustats.html" class="external-link">vegan::summary.permustats()</a></code>.</td>
</tr>
<tr class="odd">
<td><code>.r.sqr</code></td>
<td>Percent of variation explained by the model.</td>
</tr>
<tr class="even">
<td><code>.adj.r</code></td>
<td>
<code>.r.sqr</code>, taking degrees of freedom into account.</td>
</tr>
<tr class="odd">
<td><code>.aic</code></td>
<td>Akaike Information Criterion (predictive models).</td>
</tr>
<tr class="even">
<td><code>.bic</code></td>
<td>Bayesian Information Criterion (descriptive models).</td>
</tr>
<tr class="odd">
<td><code>.loglik</code></td>
<td>Log-likelihood goodness-of-fit score.</td>
</tr>
<tr class="even">
<td><code>.fit.p</code></td>
<td>P-value for observing this fit by chance.</td>
</tr>
</tbody>
</table>
<p>The <code>.h1</code> field will always come immediately after the
column it is testing against.</p>
</div>
</div>
<div class="section level2">
<h2 id="plot-output">Plot Output<a class="anchor" aria-label="anchor" href="#plot-output"></a>
</h2>
<p>Visualizations are one of the best ways to identify correlations in
your dataset. If you can see a trend with your eyes, then you’re on the
right track. The statistics-supported plotting functions in rbiom are
ordination plots, box plots, and correlation plots.</p>
<div class="section level3">
<h3 id="ordination-plots">Ordination Plots<a class="anchor" aria-label="anchor" href="#ordination-plots"></a>
</h3>
<p>Statistics for ordination plots are the most straight-forward. Set a
categorical metadata field to the <code>stat.by</code> parameter to test
whether inter-sample distances are correlated with that variable.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot</a></span><span class="op">(</span></span>
<span>  biom    <span class="op">=</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span>, </span>
<span>  stat.by <span class="op">=</span> <span class="st">"Body Site"</span>, </span>
<span>  bdiv    <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Jaccard"</span>, <span class="st">"Bray-Curtis"</span><span class="op">)</span>, </span>
<span>  ord     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PCoA"</span>, <span class="st">"UMAP"</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="statistics_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Test:     adonis2 ~ `Body Site`. 999 permutations.</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 7</span></span></span>
<span><span class="co">#&gt;   .weighted .bdiv          .n .stat    .z .p.val .adj.p</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> TRUE      Jaccard        49  11.3  51.6  0.001  0.001</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> TRUE      Bray-Curtis    49  19.5  72.0  0.001  0.001</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">stats</span><span class="op">$</span><span class="va">code</span></span>
<span><span class="co">#&gt; iters   <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">list</span><span style="color: #BBBB00;">(</span>weighted = <span style="color: #BB00BB;">TRUE</span>, bdiv = <span style="color: #00BBBB;">c</span><span style="color: #BB00BB;">(</span><span style="color: #BBBB00;">"Jaccard"</span>, <span style="color: #BBBB00;">"Bray-Curtis"</span><span style="color: #BB00BB;">)</span><span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; dm_list <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">blply</span><span style="color: #BBBB00;">(</span>biom, <span style="color: #BB00BB; font-weight: bold;">NULL</span>, bdiv_distmat, iters = iters, prefix = <span style="color: #BB00BB;">TRUE</span><span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; stats   <span style="color: #00BB00;">&lt;-</span> plyr::<span style="color: #00BBBB;">ldply</span><span style="color: #BBBB00;">(</span>dm_list, <span style="color: #BB0000;">function</span> <span style="color: #BB00BB;">(</span>dm<span style="color: #BB00BB;">)</span> <span style="color: #BB00BB;">{</span></span></span>
<span><span class="co">#&gt;   groups <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">pull(</span>biom, <span style="color: #BBBB00;">"Body Site"</span><span style="color: #00BBBB;">)[attr</span><span style="color: #BBBB00;">(</span>dm, <span style="color: #BBBB00;">'Labels')</span><span style="color: #00BBBB;">]</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">set.seed(</span><span style="color: #BB00BB;">0</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt;   ptest  <span style="color: #00BB00;">&lt;-</span> vegan::<span style="color: #00BBBB;">adonis2(</span>formula = dm <span style="color: #00BB00;">~</span> groups, permutations = <span style="color: #BB00BB;">999</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt;   pstats <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">summary(</span>vegan::<span style="color: #00BBBB;">permustats</span><span style="color: #BBBB00;">(</span>ptest<span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">with(</span>pstats, <span style="color: #00BBBB;">data.frame</span><span style="color: #BBBB00;">(</span>statistic, z, p<span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt; <span style="color: #BB00BB;">}</span><span style="color: #BBBB00;">)</span></span></span></code></pre></div>
<p>The plot subtitles have the summary statistics. Additionally,
<code>p$stats</code> contains a tibble data.frame with the full
statistics table, and <code>p$stats$code</code> shows the R commands for
reproducing the statistics outside of rbiom.</p>
<p>Note that the ordination statistics are not dependent on the
ordination, only the distance metric. This is because the statistics are
based on beta diversity distances which are computed prior to
ordination.</p>
<p>By default, <code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code> applies the perMANOVA test.
You can change this to MRPP by specifying <code>test="mrpp"</code>.
Details on the available tests are below.</p>
<table class="table">
<colgroup>
<col width="10%">
<col width="20%">
<col width="68%">
</colgroup>
<thead><tr class="header">
<th>Test</th>
<th>Function</th>
<th>Method</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>adonis2</code></td>
<td><code><a href="https://rdrr.io/pkg/vegan/man/adonis.html" class="external-link">vegan::adonis2()</a></code></td>
<td>Permutational Multivariate Analysis of Variance (perMANOVA)</td>
</tr>
<tr class="even">
<td><code>mrpp</code></td>
<td><code><a href="https://rdrr.io/pkg/vegan/man/mrpp.html" class="external-link">vegan::mrpp()</a></code></td>
<td>Multiple Response Permutation Procedure (MRPP)</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="box-plots">Box Plots<a class="anchor" aria-label="anchor" href="#box-plots"></a>
</h3>
<p>Statistics on box plots will automatically toggle between pairwise
and group-wise statistics based on the values of <code>x</code> and
<code>stat.by</code>: <code>x</code> controls pairwise and
<code>stat.by</code> controls group-wise. You can set <code>x</code> and
<code>stat.by</code> to the same categorical metadata field to get
colored pairwise statistics, or set them to different categorical
metadata fields to get multiple group-wise statistics per plot.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">`Body Site`</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Saliva'</span>, <span class="st">'Stool'</span>, <span class="st">'Buccal mucosa'</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"Body Site"</span>, stat.by <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="cn">NULL</span>,        stat.by <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"Body Site"</span>, stat.by <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"Sex"</span>,       stat.by <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">p1</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'x = "Body Site", stat.by = NULL'</span><span class="op">)</span>, </span>
<span>  <span class="va">p2</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'x = NULL, stat.by = "Body Site"'</span><span class="op">)</span>, </span>
<span>  <span class="va">p3</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'x = "Body Site", stat.by = "Body Site"'</span><span class="op">)</span>, </span>
<span>  <span class="va">p4</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'x = "Sex", stat.by = "Body Site"'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">`+`</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span>, caption <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">`+`</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.subtitle <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="statistics_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p>Above, the two plots on the left are annotated with pairwise
statistics while the two on the right have group-wise statistics. As
with other plots, you can find the full statistics tables and
reproducible R code in the plot attributes.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p2</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    kruskal.test(.diversity ~ `Body Site`)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 6</span></span></span>
<span><span class="co">#&gt;   .stat .h1      .p.val    .adj.p    .n   .df</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>  19.9 &gt; 0   0.000<span style="text-decoration: underline;">046</span>8 0.000<span style="text-decoration: underline;">046</span>8    29     2</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">p2</span><span class="op">$</span><span class="va">stats</span><span class="op">$</span><span class="va">code</span></span>
<span><span class="co">#&gt; data <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">adiv_table</span><span style="color: #BBBB00;">(</span>biom, md = <span style="color: #BBBB00;">"Body Site")</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data <span style="color: #00BB00;">%&lt;&gt;%</span> dplyr::<span style="color: #00BBBB;">rename</span><span style="color: #BBBB00;">(</span></span></span>
<span><span class="co">#&gt;   .resp    = <span style="color: #BBBB00;">".diversity"</span>, </span></span>
<span><span class="co">#&gt;   .stat.by = <span style="color: #BBBB00;">"Body Site"</span> <span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; stats <span style="color: #00BB00;">&lt;-</span> data <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   stats::<span style="color: #00BBBB;">kruskal.test</span><span style="color: #BBBB00;">(</span>.resp <span style="color: #00BB00;">~</span> .stat.by, .<span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">with</span><span style="color: #BBBB00;">(</span><span style="color: #00BBBB;">tibble</span><span style="color: #BB00BB;">(</span></span></span>
<span><span class="co">#&gt;     .stat  = statistic, </span></span>
<span><span class="co">#&gt;     .h1    = <span style="color: #00BBBB;">factor(</span><span style="color: #BBBB00;">'&gt; 0'</span><span style="color: #00BBBB;">)</span>, </span></span>
<span><span class="co">#&gt;     .p.val = p.value, </span></span>
<span><span class="co">#&gt;     .n     = <span style="color: #00BBBB;">nrow(</span>data<span style="color: #00BBBB;">)</span>, </span></span>
<span><span class="co">#&gt;     .df    = parameter <span style="color: #BB00BB;">)</span><span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   dplyr::<span style="color: #00BBBB;">mutate</span><span style="color: #BBBB00;">(</span>.adj.p = <span style="color: #00BBBB;">p.adjust</span><span style="color: #BB00BB;">(</span>.p.val, <span style="color: #BBBB00;">'fdr'</span><span style="color: #BB00BB;">)</span>, .after = .p.val<span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   dplyr::<span style="color: #00BBBB;">arrange</span><span style="color: #BBBB00;">(</span>.p.val<span style="color: #BBBB00;">)</span></span></span></code></pre></div>
<p>Internally, rbiom uses the non-parametric functions listed below.</p>
<table class="table">
<colgroup>
<col width="14%">
<col width="25%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>Test</th>
<th>Function</th>
<th>Method</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>pairwise</td>
<td><code><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">stats::wilcox.test()</a></code></td>
<td>Two-sample Wilcoxon Rank Sum Test, aka Mann-Whitney Test</td>
</tr>
<tr class="even">
<td>group-wise</td>
<td><code><a href="https://rdrr.io/r/stats/kruskal.test.html" class="external-link">stats::kruskal.test()</a></code></td>
<td>Kruskal-Wallis Rank Sum Test</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="correlation-plots">Correlation Plots<a class="anchor" aria-label="anchor" href="#correlation-plots"></a>
</h3>
<p>For an in-depth description of correlation plots, see
vignette(“articles/regression”).</p>
<p>Depending on the arguments given to <code>stat.by</code> and
<code>test</code>, you can test:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="va">gems</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">diarrhea</span> <span class="op">==</span> <span class="st">"Control"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">country</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Bangladesh"</span>, <span class="st">"Kenya"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_corrplot.html">adiv_corrplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"age"</span>, test <span class="op">=</span> <span class="st">"emmeans"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_corrplot.html">adiv_corrplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"age"</span>, test <span class="op">=</span> <span class="st">"emmeans"</span>, stat.by <span class="op">=</span> <span class="st">"country"</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_corrplot.html">adiv_corrplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"age"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_corrplot.html">adiv_corrplot</a></span><span class="op">(</span><span class="va">biom</span>, x <span class="op">=</span> <span class="st">"age"</span>, test <span class="op">=</span> <span class="st">"emtrends"</span>, stat.by <span class="op">=</span> <span class="st">"country"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="va">p1</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'Is the mean non-zero?'</span><span class="op">)</span>, </span>
<span>  <span class="va">p2</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'Do means vary by group?'</span><span class="op">)</span>, </span>
<span>  <span class="va">p3</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'Is the slope non-horizontal?'</span><span class="op">)</span>, </span>
<span>  <span class="va">p4</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="st">'Do slopes vary by group?'</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">`+`</span>, <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, y <span class="op">=</span> <span class="cn">NULL</span>, caption <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">plots</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span></span></code></pre></div>
<p><img src="statistics_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<div class="section level3">
<h3 id="normality">Normality<a class="anchor" aria-label="anchor" href="#normality"></a>
</h3>
<p>A normal distribution is visualized as a “bell curve”, where values
further from the mean are observed less often. Microbial abundances do
not follow this pattern; it’s common to observe high or low abundances
more often than a “medium” abundance.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span></span>
<span>  widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1.5</span><span class="op">)</span>,</span>
<span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Normal Distribution"</span><span class="op">)</span>,</span>
<span>    </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="../reference/taxa_matrix.html">taxa_table</a></span><span class="op">(</span><span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span>, taxa <span class="op">=</span> <span class="fl">4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">.abundance</span><span class="op">)</span>, bins <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="st">".taxa"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Genera Abundance Distributions"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="statistics_files/figure-html/unnamed-chunk-10-1.png" width="768"></p>
<p>To compensate for this non-normality, rbiom uses the following
non-parametric tests for categorical variables that are based on ranking
or permutations.</p>
<table class="table">
<colgroup>
<col width="31%">
<col width="31%">
<col width="36%">
</colgroup>
<thead><tr class="header">
<th>Test</th>
<th>Function</th>
<th>Used For</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Wilcoxon Rank-Sum</td>
<td><code><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">stats::wilcox.test()</a></code></td>
<td>Pairwise boxplot</td>
</tr>
<tr class="even">
<td>Kruskal-Wallis Rank Sum</td>
<td><code><a href="https://rdrr.io/r/stats/kruskal.test.html" class="external-link">stats::kruskal.test()</a></code></td>
<td>Groupwise boxplot</td>
</tr>
<tr class="odd">
<td>Permutational MANOVA</td>
<td><code><a href="https://rdrr.io/pkg/vegan/man/adonis.html" class="external-link">vegan::adonis2()</a></code></td>
<td>
<code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code> clusters</td>
</tr>
</tbody>
</table>
<p>For correlation/regression analysis, rbiom provides diagnostic plots
to determine when residual distributions are cause for concern.</p>
<p>Further reading:</p>
<ul>
<li>
<a href="https://uw.pressbooks.pub/appliedmultivariatestatistics/" class="external-link">Applied
Multivariate Statistics in R</a>: <a href="https://uw.pressbooks.pub/appliedmultivariatestatistics/chapter/permanova/" class="external-link">PERMANOVA</a>,
<a href="https://uw.pressbooks.pub/appliedmultivariatestatistics/chapter/comparison-of-techniques/" class="external-link">Comparison
of Techniques</a>
</li>
<li>
<a href="https://medium.com/datasciencearth/r-applications-part-1-simple-linear-regression-ef5a0e19a05d" class="external-link">R
Applications — Part 1: Simple Linear Regression</a>, particularly the
section on <a href="https://medium.com/datasciencearth/r-applications-part-1-simple-linear-regression-ef5a0e19a05d#a6f6" class="external-link">assumption
control</a>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="compositionality">Compositionality<a class="anchor" aria-label="anchor" href="#compositionality"></a>
</h3>
<p>Compositional data arises when the counts don’t represent the entire
population. In microbiome studies, the number of microbes that get
sequenced is far less than the number of microbes from where the sample
was collected. Articles by <a href="https://doi.org/10.3389/fmicb.2017.02224" class="external-link">Gloor et al</a> and <a href="https://doi.org/10.1371/journal.pcbi.1003531" class="external-link">McMurdie and
Holmes</a> propose the use of their analysis tools ( <a href="https://doi.org/doi:10.18129/B9.bioc.ALDEx2" class="external-link">ALDEx2</a> and <a href="https://doi.org/doi:10.18129/B9.bioc.metagenomeSeq" class="external-link">metagenomeSeq</a>,
respectively) to apply the proper statistical methods for this
situation. Conversely, rbiom does not correct for compositionality. This
is because correcting for compositionality introduces extra noise into
the dataset and severely limits the selection of metrics and
visualizations, typically without any significant benefit to
analysis.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
