<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The biplot parameters (taxa, unc, p.top, and
p.adj) only only have an effect when rank is not NULL."><title>Calculate PCoA and other ordinations, including taxa biplots and statistics. — bdiv_ord_table • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Calculate PCoA and other ordinations, including taxa biplots and statistics. — bdiv_ord_table"><meta property="og:description" content="The biplot parameters (taxa, unc, p.top, and
p.adj) only only have an effect when rank is not NULL."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9126</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/rbiom.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6>
    <a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a>
    <a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a>
    <a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a>
    <a class="dropdown-item" href="../articles/diversity.html">Diversity Metrics</a>
    <a class="dropdown-item" href="../articles/statistics.html">Statistics</a>
    <a class="dropdown-item" href="../reference/speed.html">Speed Ups</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Plotting</h6>
    <a class="dropdown-item" href="../articles/plots.html">Plot Types</a>
    <a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a>
    <a class="dropdown-item" href="../articles/scales.html">Scales and Transformations</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Calculate PCoA and other ordinations, including taxa biplots and statistics.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/HEAD/R/bdiv_ord_table.r" class="external-link"><code>R/bdiv_ord_table.r</code></a></small>
      <div class="d-none name"><code>bdiv_ord_table.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The biplot parameters (<code>taxa</code>, <code>unc</code>, <code>p.top</code>, and
<code>p.adj</code>) only only have an effect when <code>rank</code> is not <code>NULL</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">bdiv_ord_table</span><span class="op">(</span></span>
<span>  <span class="va">biom</span>,</span>
<span>  bdiv <span class="op">=</span> <span class="st">"Bray-Curtis"</span>,</span>
<span>  ord <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  weighted <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  md <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  stat.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  split.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  test <span class="op">=</span> <span class="st">"adonis2"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  permutations <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  rank <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  taxa <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  p.top <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  p.adj <span class="op">=</span> <span class="st">"fdr"</span>,</span>
<span>  unc <span class="op">=</span> <span class="st">"singly"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>biom</dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt>bdiv</dt>
<dd><p>Beta diversity distance algorithm(s) to use. Options are:
<code>"Bray-Curtis"</code>, <code>"Manhattan"</code>, <code>"Euclidean"</code>,
<code>"Jaccard"</code>, and <code>"UniFrac"</code>. For <code>"UniFrac"</code>, a
phylogenetic tree must be present in <code>biom</code> or explicitly
provided via <code>tree=</code>. Default: <code>"Bray-Curtis"</code> <br><br>
Multiple/abbreviated values allowed.</p></dd>


<dt>ord</dt>
<dd><p>Method for reducing dimensionality. Options are:</p><ul><li><p><code>"UMAP"</code> -  Uniform manifold approximation and projection; <code><a href="https://jlmelville.github.io/uwot/reference/umap.html" class="external-link">uwot::umap()</a></code>.</p></li>
<li><p><code>"PCoA"</code> -  Principal coordinate analysis; <code><a href="https://rdrr.io/pkg/ape/man/pcoa.html" class="external-link">ape::pcoa()</a></code>.</p></li>
<li><p><code>"NMDS"</code> -  Nonmetric multidimensional scaling; <code><a href="https://rdrr.io/pkg/vegan/man/metaMDS.html" class="external-link">vegan::metaMDS()</a></code>.</p></li>
<li><p><code>"tSNE"</code> -  t-distributed stochastic neighbor embedding; <code><a href="https://rdrr.io/pkg/tsne/man/tsne.html" class="external-link">tsne::tsne()</a></code>.</p></li>
</ul><p>Default: <code>"UMAP"</code> <br><br>
Multiple/abbreviated values allowed.</p></dd>


<dt>weighted</dt>
<dd><p>Take relative abundances into account. When
<code>weighted=FALSE</code>, only presence/absence is considered.
Default: <code>TRUE</code> <br><br>
Multiple values allowed.</p></dd>


<dt>md</dt>
<dd><p>Dataset field(s) to include in the output data frame, or <code>'.all'</code>
to include all metadata fields. Default: <code>'.all'</code></p></dd>


<dt>k</dt>
<dd><p>Number of ordination dimensions to return. Either <code>2L</code> or
<code>3L</code>. Default: <code>2L</code></p></dd>


<dt>stat.by</dt>
<dd><p>The categorical or numeric metadata field over which
statistics should be calculated. Required.</p></dd>


<dt>split.by</dt>
<dd><p>Dataset field(s) that the data should be split by prior to
any calculations. Must be categorical. Default: <code>NULL</code></p></dd>


<dt>tree</dt>
<dd><p>A <code>phylo</code> object representing the phylogenetic
relationships of the taxa in <code>biom</code>. Only required when
computing UniFrac distances. Default: <code>biom$tree</code></p></dd>


<dt>test</dt>
<dd><p>Permutational test for accessing significance. Options are:</p><ul><li><p><code>"adonis2"</code> -  Permutational MANOVA; <code><a href="https://rdrr.io/pkg/vegan/man/vegan-deprecated.html" class="external-link">vegan::adonis2()</a></code>.</p></li>
<li><p><code>"mrpp"</code> -  Multiple response permutation procedure; <code><a href="https://rdrr.io/pkg/vegan/man/mrpp.html" class="external-link">vegan::mrpp()</a></code>.</p></li>
<li><p><code>"none"</code> -  Don't run any statistics.</p></li>
</ul><p>Default: <code>"adonis2"</code> <br><br>
Abbreviations are allowed.</p></dd>


<dt>seed</dt>
<dd><p>Random seed for permutations. Default: <code>0</code></p></dd>


<dt>permutations</dt>
<dd><p>Number of random permutations to use.
Default: <code>999</code></p></dd>


<dt>rank</dt>
<dd><p>What rank(s) of taxa to compute biplot coordinates and
statistics for, or <code>NULL</code> to disable. E.g. <code>"Phylum"</code>,
<code>"Genus"</code>, <code>".otu"</code>, etc. An integer vector can also be
given, where <code>1</code> is the highest rank, <code>2</code> is the second
highest, <code>-1</code> is the lowest rank, <code>-2</code> is the second
lowest, and <code>0</code> is the OTU "rank". Run <code>biom$ranks</code> to
see all options for a given rbiom object. Default: <code>NULL</code>.</p></dd>


<dt>taxa</dt>
<dd><p>Which taxa to display. An integer value will show the top n
most abundant taxa. A value 0 &lt;= n &lt; 1 will show any taxa with that
mean abundance or greater (e.g. <code>0.1</code> implies &gt;= 10%). A
character vector of taxa names will show only those named taxa.
Default: <code>6</code>.</p></dd>


<dt>p.top</dt>
<dd><p>Only display taxa with the most significant differences in
abundance. If <code>p.top</code> is &gt;= 1, then the <code>p.top</code> most
significant taxa are displayed. If <code>p.top</code> is less than one, all
taxa with an adjusted p-value &lt;= <code>p.top</code> are displayed.
Recommended to be used in combination with the <code>taxa</code> parameter
to set a lower bound on the mean abundance of considered taxa.
Default: <code>Inf</code></p></dd>


<dt>p.adj</dt>
<dd><p>Method to use for multiple comparisons adjustment of
p-values. Run <code>p.adjust.methods</code> for a list of available
options. Default: <code>"fdr"</code></p></dd>


<dt>unc</dt>
<dd><p>How to handle unclassified, uncultured, and similarly ambiguous
taxa names. Options are:</p><ul><li><p><code>"singly"</code> -  Replaces them with the OTU name.</p></li>
<li><p><code>"grouped"</code> -  Replaces them with a higher rank's name.</p></li>
<li><p><code>"drop"</code> -  Excludes them from the result.</p></li>
<li><p><code>"asis"</code> -  To not check/modify any taxa names.</p></li>
</ul><p>Default: <code>"singly"</code> <br><br>
Abbreviations are allowed.</p></dd>


<dt>...</dt>
<dd><p>Additional arguments to pass on to <code><a href="https://jlmelville.github.io/uwot/reference/umap.html" class="external-link">uwot::umap()</a></code>,
<code><a href="https://rdrr.io/pkg/ape/man/pcoa.html" class="external-link">ape::pcoa()</a></code>, <code><a href="https://rdrr.io/pkg/vegan/man/metaMDS.html" class="external-link">vegan::metaMDS()</a></code>, or <code><a href="https://rdrr.io/pkg/tsne/man/tsne.html" class="external-link">tsne::tsne()</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A data.frame with columns <code>.sample</code>, <code>.weighted</code>,
<code>.bdiv</code>, <code>.ord</code>, <code>.x</code>, <code>.y</code>, and (optionally)
<code>.z</code>. Any columns given by <code>md</code>, <code>split.by</code>, and
<code>stat.by</code> are included as well.<br></p>


<p>If <code>stat.by</code> is given, then <code>$stats</code> and
<code>$stats$code)</code> are set.<br></p>


<p>If <code>rank</code> is given, then <code>$taxa_coords</code>,
<code>$taxa_stats</code>, and <code>$taxa_stats$code</code> are set.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other beta_diversity: 
<code><a href="bdiv_boxplot.html">bdiv_boxplot</a>()</code>,
<code><a href="bdiv_corrplot.html">bdiv_corrplot</a>()</code>,
<code><a href="bdiv_heatmap.html">bdiv_heatmap</a>()</code>,
<code><a href="bdiv_ord_plot.html">bdiv_ord_plot</a>()</code>,
<code><a href="bdiv_stats.html">bdiv_stats</a>()</code>,
<code><a href="bdiv_table.html">bdiv_table</a>()</code>,
<code><a href="distmat_stats.html">distmat_stats</a>()</code></p>
<p>Other ordination: 
<code><a href="bdiv_ord_plot.html">bdiv_ord_plot</a>()</code>,
<code><a href="distmat_ord_table.html">distmat_ord_table</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">ord</span> <span class="op">&lt;-</span> <span class="fu">bdiv_ord_table</span><span class="op">(</span><span class="va">hmp50</span>, <span class="st">"bray"</span>, <span class="st">"pcoa"</span>, stat.by<span class="op">=</span><span class="st">"Body Site"</span>, rank<span class="op">=</span><span class="st">"g"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ord</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .weighted .bdiv       .ord  .sample     .x       .y `Body Site`  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> TRUE      Bray-Curtis PCoA  HMP01   -<span style="color: #BB0000;">0.387</span> -<span style="color: #BB0000;">0.008</span><span style="color: #BB0000; text-decoration: underline;">95</span> Buccal mucosa</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> TRUE      Bray-Curtis PCoA  HMP02   -<span style="color: #BB0000;">0.461</span> -<span style="color: #BB0000;">0.013</span><span style="color: #BB0000; text-decoration: underline;">5</span>  Buccal mucosa</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> TRUE      Bray-Curtis PCoA  HMP03   -<span style="color: #BB0000;">0.407</span> -<span style="color: #BB0000;">0.014</span><span style="color: #BB0000; text-decoration: underline;">3</span>  Saliva       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> TRUE      Bray-Curtis PCoA  HMP04   -<span style="color: #BB0000;">0.378</span> -<span style="color: #BB0000;">0.012</span><span style="color: #BB0000; text-decoration: underline;">3</span>  Saliva       </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> TRUE      Bray-Curtis PCoA  HMP05   -<span style="color: #BB0000;">0.452</span> -<span style="color: #BB0000;">0.011</span><span style="color: #BB0000; text-decoration: underline;">4</span>  Buccal mucosa</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> TRUE      Bray-Curtis PCoA  HMP06   -<span style="color: #BB0000;">0.411</span> -<span style="color: #BB0000;">0.012</span><span style="color: #BB0000; text-decoration: underline;">3</span>  Saliva       </span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">ord</span><span class="op">$</span><span class="va">stats</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Test:     adonis2 ~ `Body Site`. 999 permutations.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 1 × 7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .weighted .bdiv          .n .stat    .z .p.val .adj.p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> TRUE      Bray-Curtis    50  19.2  72.2  0.001  0.001</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">ord</span><span class="op">$</span><span class="va">taxa_stats</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Test:     adonis2 ~ taxa. 999 permutations.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 9</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .weighted .bdiv       .rank .taxa                .n .stat    .z .p.val .adj.p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> TRUE      Bray-Curtis Genus Lactobacillus        50  8.62  16.8  0.001  0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> TRUE      Bray-Curtis Genus Streptococcus        50 10.1   19.5  0.001  0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> TRUE      Bray-Curtis Genus Bacteroides          50  8.72  16.7  0.001  0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> TRUE      Bray-Curtis Genus Corynebacterium 1    50  8.53  16.0  0.001  0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> TRUE      Bray-Curtis Genus Haemophilus          50  6.56  12.3  0.001  0.001</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> TRUE      Bray-Curtis Genus Propionibacterium    50  5.84  11.3  0.001  0.001</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    </span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

