<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="documentation_default"><title>documentation_default — documentation_default • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="documentation_default — documentation_default"><meta property="og:description" content="documentation_default"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9129</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/rbiom.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6>
    <a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a>
    <a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a>
    <a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a>
    <a class="dropdown-item" href="../articles/statistics.html">Statistics</a>
    <a class="dropdown-item" href="../articles/regression.html">Regression</a>
    <a class="dropdown-item" href="../reference/speed.html">Speed Ups</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Plotting</h6>
    <a class="dropdown-item" href="../articles/plots.html">Plot Types</a>
    <a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>documentation_default</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/HEAD/R/documentation.r" class="external-link"><code>R/documentation.r</code></a></small>
      <div class="d-none name"><code>documentation_default.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>documentation_default</p>
    </div>


    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>biom</dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt>mtx</dt>
<dd><p>A matrix-like object.</p></dd>


<dt>tree</dt>
<dd><p>A <code>phylo</code> object representing the phylogenetic
relationships of the taxa in <code>biom</code>. Only required when
computing UniFrac distances. Default: <code>biom$tree</code></p></dd>


<dt>md</dt>
<dd><p>Dataset field(s) to include in the output data frame, or <code>'.all'</code>
to include all metadata fields. Default: <code>'.all'</code></p></dd>


<dt>adiv</dt>
<dd><p>Alpha diversity metric(s) to use. Options are: <code>"OTUs"</code>,
<code>"Shannon"</code>, <code>"Chao1"</code>, <code>"Simpson"</code>, and/or
<code>"InvSimpson"</code>. Set <code>adiv=".all"</code> to use all metrics.
Default: <code>"Shannon"</code> <br><br>
Multiple/abbreviated values allowed.</p></dd>


<dt>bdiv</dt>
<dd><p>Beta diversity distance algorithm(s) to use. Options are:
<code>"Bray-Curtis"</code>, <code>"Manhattan"</code>, <code>"Euclidean"</code>,
<code>"Jaccard"</code>, and <code>"UniFrac"</code>. For <code>"UniFrac"</code>, a
phylogenetic tree must be present in <code>biom</code> or explicitly
provided via <code>tree=</code>. Default: <code>"Bray-Curtis"</code> <br><br>
Multiple/abbreviated values allowed.</p></dd>


<dt>taxa</dt>
<dd><p>Which taxa to display. An integer value will show the top n
most abundant taxa. A value 0 &lt;= n &lt; 1 will show any taxa with that
mean abundance or greater (e.g. <code>0.1</code> implies &gt;= 10%). A
character vector of taxa names will show only those named taxa.
Default: <code>6</code>.</p></dd>


<dt>ord</dt>
<dd><p>Method for reducing dimensionality. Options are:</p><ul><li><p><code>"UMAP"</code> -  Uniform manifold approximation and projection; <code><a href="https://jlmelville.github.io/uwot/reference/umap.html" class="external-link">uwot::umap()</a></code>.</p></li>
<li><p><code>"PCoA"</code> -  Principal coordinate analysis; <code><a href="https://rdrr.io/pkg/ape/man/pcoa.html" class="external-link">ape::pcoa()</a></code>.</p></li>
<li><p><code>"NMDS"</code> -  Nonmetric multidimensional scaling; <code><a href="https://rdrr.io/pkg/vegan/man/metaMDS.html" class="external-link">vegan::metaMDS()</a></code>.</p></li>
<li><p><code>"tSNE"</code> -  t-distributed stochastic neighbor embedding; <code><a href="https://rdrr.io/pkg/tsne/man/tsne.html" class="external-link">tsne::tsne()</a></code>.</p></li>
</ul><p>Default: <code>"UMAP"</code> <br><br>
Multiple/abbreviated values allowed.</p></dd>


<dt>weighted</dt>
<dd><p>Take relative abundances into account. When
<code>weighted=FALSE</code>, only presence/absence is considered.
Default: <code>TRUE</code> <br><br>
Multiple values allowed.</p></dd>


<dt>delta</dt>
<dd><p>For numeric metadata, report the absolute difference in values
for the two samples, for instance <code>2</code> instead of <code>"10 vs 12"</code>.
Default: <code>TRUE</code></p></dd>


<dt>rank</dt>
<dd><p>What rank(s) of taxa to display. E.g. <code>"Phylum"</code>,
<code>"Genus"</code>, <code>".otu"</code>, etc. An integer vector can also be
given, where <code>1</code> is the highest rank, <code>2</code> is the second
highest, <code>-1</code> is the lowest rank, <code>-2</code> is the second
lowest, and <code>0</code> is the OTU "rank". Run <code>biom$ranks</code> to
see all options for a given rbiom object. Default: <code>-1</code>.</p></dd>


<dt>lineage</dt>
<dd><p>Include all ranks in the name of the taxa. For instance,
setting to <code>TRUE</code> will produce
<code>Bacteria; Actinobacteria; Coriobacteriia; Coriobacteriales</code>.
Otherwise the taxa name will simply be <code>Coriobacteriales</code>. You
want to set this to TRUE when <code>unc = "asis"</code> and you have taxa
names (such as <em>Incertae_Sedis</em>) that map to multiple higher
level ranks. Default: <code>FALSE</code></p></dd>


<dt>unc</dt>
<dd><p>How to handle unclassified, uncultured, and similarly ambiguous
taxa names. Options are:</p><ul><li><p><code>"singly"</code> -  Replaces them with the OTU name.</p></li>
<li><p><code>"grouped"</code> -  Replaces them with a higher rank's name.</p></li>
<li><p><code>"drop"</code> -  Excludes them from the result.</p></li>
<li><p><code>"asis"</code> -  To not check/modify any taxa names.</p></li>
</ul><p>Default: <code>"singly"</code> <br><br>
Abbreviations are allowed.</p></dd>


<dt>other</dt>
<dd><p>Sum all non-itemized taxa into an "Other" taxa. When
<code>FALSE</code>, only returns taxa matched by the <code>taxa</code>
argument. Specifying <code>TRUE</code> adds "Other" to the returned set.
A string can also be given to imply <code>TRUE</code>, but with that
value as the name to use instead of "Other".
Default: <code>FALSE</code></p></dd>


<dt>sparse</dt>
<dd><p>If true, returns a sparse matrix as described by
<code><a href="https://rdrr.io/pkg/slam/man/matrix.html" class="external-link">slam::simple_triplet_matrix()</a></code>, otherwise returns a normal R
matrix object. Default: <code>FALSE</code></p></dd>


<dt>p.top</dt>
<dd><p>Only display taxa with the most significant differences in
abundance. If <code>p.top</code> is &gt;= 1, then the <code>p.top</code> most
significant taxa are displayed. If <code>p.top</code> is less than one, all
taxa with an adjusted p-value &lt;= <code>p.top</code> are displayed.
Recommended to be used in combination with the <code>taxa</code> parameter
to set a lower bound on the mean abundance of considered taxa.
Default: <code>Inf</code></p></dd>


<dt>y.trans</dt>
<dd><p>The transformation to apply to the y-axis. Visualizing
differences of both high- and low-abundance taxa is best done with
a non-linear axis. Options are:</p><ul><li><p><code>"sqrt"</code> -  square-root transformation</p></li>
<li><p><code>"log1p"</code> -  log(y + 1) transformation</p></li>
<li><p><code>NULL</code> -  no transformation</p></li>
</ul><p>These methods allow visualization of both high- and low-abundance
taxa simultaneously, without complaint about 'zero' count
observations. Default: <code>"sqrt"</code></p></dd>


<dt>flip</dt>
<dd><p>Transpose the axes, so that taxa are present as rows instead
of columns. Default: <code>FALSE</code></p></dd>


<dt>stripe</dt>
<dd><p>Shade every other x position. Default: <em>same as flip</em></p></dd>


<dt>ci</dt>
<dd><p>How to calculate min/max of the <b>crossbar</b>,
<b>errorbar</b>, <b>linerange</b>, and <b>pointrange</b> layers.
Options are: <code>"ci"</code> (confidence interval), <code>"range"</code>,
<code>"sd"</code> (standard deviation), <code>"se"</code> (standard error), and
<code>"mad"</code> (median absolute deviation).
The center mark of <b>crossbar</b> and <b>pointrange</b> represents
the mean, except for <code>"mad"</code> in which case it represents the median.
Default: <code>"ci"</code></p></dd>


<dt>p.label</dt>
<dd><p>Minimum adjusted p-value to display on the plot with a
bracket.</p><ul><li><p><code>p.label = 0.05</code> -  Show p-values that are &lt;= 0.05.</p></li>
<li><p><code>p.label = 0</code> -  Don't show any p-values on the plot.</p></li>
<li><p><code>p.label = 1</code> -  Show all p-values on the plot.</p></li>
</ul><p>If a numeric vector with more than one value is
provided, they will be used as breaks for asterisk notation.
Default: <code>0.05</code></p></dd>


<dt>level</dt>
<dd><p>The confidence level for calculating a confidence interval.
Default: <code>0.95</code></p></dd>


<dt>caption</dt>
<dd><p>Add methodology caption beneath the plot.
Default: <code>TRUE</code></p></dd>


<dt>outliers</dt>
<dd><p>Show boxplot outliers? <code>TRUE</code> to always show.
<code>FALSE</code> to always hide. <code>NULL</code> to only hide them when
overlaying a dot or strip chart.  Default: <code>NULL</code></p></dd>


<dt>xlab.angle</dt>
<dd><p>Angle of the labels at the bottom of the plot.
Options are <code>"auto"</code>, <code>'0'</code>, <code>'30'</code>, and <code>'90'</code>.
Default: <code>"auto"</code>.</p></dd>


<dt>k</dt>
<dd><p>Number of ordination dimensions to return. Either <code>2L</code> or
<code>3L</code>. Default: <code>2L</code></p></dd>


<dt>split.by</dt>
<dd><p>Dataset field(s) that the data should be split by prior to
any calculations. Must be categorical. Default: <code>NULL</code></p></dd>


<dt>dm</dt>
<dd><p>A <code>dist</code>-class distance matrix, as returned from
<code><a href="bdiv_table.html">bdiv_distmat()</a></code> or <code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">stats::dist()</a></code>. Required.</p></dd>


<dt>groups</dt>
<dd><p>A named vector of grouping values. The names should
correspond to <code>attr(dm, 'Labels')</code>. Values can be either
categorical or numeric. Required.</p></dd>


<dt>df</dt>
<dd><p>The dataset (data.frame or tibble object). "Dataset fields"
mentioned below should match column names in <code>df</code>. Required.</p></dd>


<dt>regr</dt>
<dd><p>Dataset field with the x-axis (independent; predictive)
values. Must be numeric. Default: <code>NULL</code></p></dd>


<dt>resp</dt>
<dd><p>Dataset field with the y-axis (dependent; response) values,
such as taxa abundance or alpha diversity.
Default: <code>attr(df, 'response')</code></p></dd>


<dt>stat.by</dt>
<dd><p>Dataset field with the statistical groups. Must be
categorical. Default: <code>NULL</code></p></dd>


<dt>color.by</dt>
<dd><p>Dataset field with the group to color by. Must be
categorical. Default: <code>stat.by</code></p></dd>


<dt>shape.by</dt>
<dd><p>Dataset field with the group for shapes. Must be
categorical. Default: <code>stat.by</code></p></dd>


<dt>facet.by</dt>
<dd><p>Dataset field(s) to use for faceting. Must be categorical.
Default: <code>NULL</code></p></dd>


<dt>colors</dt>
<dd><p>How to color the groups. Options are:</p><ul><li><p><code>TRUE</code> -  Automatically select colorblind-friendly colors.</p></li>
<li><p><code>FALSE</code> or <code>NULL</code> -  Don't use colors.</p></li>
<li><p>a palette name -  Auto-select colors from this set. E.g. <code>"okabe"</code></p></li>
<li><p>character vector -  Custom colors to use. E.g. <code>c("red", "#00FF00")</code></p></li>
<li><p>named character vector -  Explicit mapping. E.g. <code>c(Male = "blue", Female = "red")</code></p></li>
</ul><p>See "Aesthetics" section below for additional information.
Default: <code>TRUE</code></p></dd>


<dt>shapes</dt>
<dd><p>Shapes for each group.
Options are similar to <code>colors</code>'s: <code>TRUE</code>, <code>FALSE</code>, <code>NULL</code>, shape
names (typically integers 0 - 17), or a named vector mapping
groups to specific shape names.
See "Aesthetics" section below for additional information.
Default: <code>TRUE</code></p></dd>


<dt>patterns</dt>
<dd><p>Patterns for each group.
Options are similar to <code>colors</code>'s: <code>TRUE</code>, <code>FALSE</code>, <code>NULL</code>, pattern
names (<code>"brick"</code>, <code>"chevron"</code>, <code>"fish"</code>, <code>"grid"</code>, etc), or a named
vector mapping groups to specific pattern names.
See "Aesthetics" section below for additional information.
Default: <code>FALSE</code></p></dd>


<dt>test</dt>
<dd><p>Method for computing p-values: <code>'wilcox'</code>, <code>'kruskal'</code>,
<code>'emmeans'</code>, or <code>'emtrends'</code>. Default: <code>'emmeans'</code></p></dd>


<dt>fit</dt>
<dd><p>How to fit the trendline. <code>'lm'</code>, <code>'log'</code>, or <code>'gam'</code>.
Default: <code>'gam'</code></p></dd>


<dt>at</dt>
<dd><p>Position(s) along the x-axis where the means or slopes should be
evaluated. Default: <code>NULL</code>, which samples 100 evenly spaced positions
and selects the position where the p-value is most significant.</p></dd>


<dt>alt</dt>
<dd><p>Alternative hypothesis direction. Options are <code>'!='</code>
(two-sided; not equal to <code>mu</code>), <code>'&lt;'</code> (less than <code>mu</code>), or <code>'&gt;'</code>
(greater than <code>mu</code>). Default: <code>'!='</code></p></dd>


<dt>mu</dt>
<dd><p>Reference value to test against. Default: <code>0</code></p></dd>


<dt>check</dt>
<dd><p>Generate additional plots to aid in assessing data normality.
Default: <code>FALSE</code></p></dd>


<dt>within, between</dt>
<dd><p>Dataset field(s) for intra- or inter- sample
comparisons. Alternatively, dataset field names given elsewhere can
be prefixed with <code>'=='</code> or <code>'!='</code> to assign them to <code>within</code> or
<code>between</code>, respectively. Default: <code>NULL</code></p></dd>


<dt>seed</dt>
<dd><p>Random seed for permutations. Default: <code>0</code></p></dd>


<dt>permutations</dt>
<dd><p>Number of random permutations to use.
Default: <code>999</code></p></dd>


<dt>p.adj</dt>
<dd><p>Method to use for multiple comparisons adjustment of
p-values. Run <code>p.adjust.methods</code> for a list of available
options. Default: <code>"fdr"</code></p></dd>


<dt>depths</dt>
<dd><p>Rarefaction depths to show in the plot, or <code>NULL</code> to
auto-select. Default: <code>NULL</code></p></dd>


<dt>rline</dt>
<dd><p>Where to draw a horizontal line on the plot, intended to show
a particular rarefaction depth. Set to <code>TRUE</code> to show an
auto-selected rarefaction depth or <code>FALSE</code> to not show a line.
Default: <code>NULL</code></p></dd>


<dt>clone</dt>
<dd><p>Create a copy of <code>biom</code> before modifying. If <code>FALSE</code>, <code>biom</code>
is modified in place as a side-effect. See <a href="speed.html">speed ups</a> for
use cases. Default: <code>TRUE</code></p></dd>


<dt>labels</dt>
<dd><p>Show sample names under each bar. Default: <code>FALSE</code></p></dd>


<dt>trans</dt>
<dd><p>Transformation to apply. Options are:
<code>c("none", "rank", "log", "log1p", "sqrt")</code>. <code>"rank"</code> is useful for
correcting for non-normally distributions before applying regression
statistics. Default: <code>"none"</code></p></dd>

</dl></div>

  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

