<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Function reference â€¢ rbiom</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Function reference" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-index">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rbiom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.2.9030</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cmmr/rbiom/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index">

    <colgroup>
      
      <col class="alias" />
      <col class="title" />
    </colgroup>

    <tbody>
      <tr>
        <th colspan="2">
          <h2 id="section-all-functions" class="hasAnchor"><a href="#section-all-functions" class="anchor"></a>All functions</h2>
          <p class="section-desc"></p>
        </th>
      </tr>
      
      
    </tbody><tbody>
      
      
      <tr>
        
        <td>
          <p><code><a href="alpha.div.html">alpha.div()</a></code> </p>
        </td>
        <td><p>Estimate the diversity of each sample.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="apcoa.html">apcoa()</a></code> </p>
        </td>
        <td><p>PCoA adjusted for corvariates.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="beta.div.html">beta.div()</a></code> </p>
        </td>
        <td><p>Make a distance matrix of samples vs samples.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="counts.html">counts()</a></code> </p>
        </td>
        <td><p>Get the abundance counts.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="depth.html">depth()</a></code> </p>
        </td>
        <td><p>The rarefaction depth of a BIOM object.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="distill.html">distill()</a></code> </p>
        </td>
        <td><p>Convert to a data.frame with metadata and diversity/abundance values.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="has.phylogeny.html">has.phylogeny()</a></code> </p>
        </td>
        <td><p>Checks if a phylogenetic tree is present.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="has.sequences.html">has.sequences()</a></code> </p>
        </td>
        <td><p>Checks if a DNA sequences are present.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="info.html">info()</a></code> </p>
        </td>
        <td><p>Get biom's misc information.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="is.rarefied.html">is.rarefied()</a></code> </p>
        </td>
        <td><p>Checks if a BIOM object is rarefied.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="metadata.html">metadata()</a></code> </p>
        </td>
        <td><p>Get the sample metadata.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="nsamples.html">nsamples()</a></code> </p>
        </td>
        <td><p>Number of samples in a BIOM.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ntaxa.html">ntaxa()</a></code> </p>
        </td>
        <td><p>Number of taxa in a BIOM.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="ordinate.html">ordinate()</a></code> </p>
        </td>
        <td><p>Reduce a distance matrix to two or three dimensions</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="phylogeny.html">phylogeny()</a></code> </p>
        </td>
        <td><p>Get the phylogenetic tree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="plot.html">plot(<i>&lt;BIOM&gt;</i>)</a></code> </p>
        </td>
        <td><p>Visualize diversity or abundance as a boxplot, dotplot, etc.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="print.html">print(<i>&lt;BIOM&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summarize the contents of a BIOM object</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rarefy.html">rarefy()</a></code> </p>
        </td>
        <td><p>Subset counts so that all samples have the same number of observations.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="rbiom.html">rbiom</a></code> </p>
        </td>
        <td><p>rbiom: Read/Write, Transform, and Summarize BIOM Data</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read.biom.html">read.biom()</a></code> </p>
        </td>
        <td><p>Extracts counts, metadata, taxonomy, and phylogeny from a biom file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read.fasta.html">read.fasta()</a></code> </p>
        </td>
        <td><p>Parse a fasta file into a named character vector.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="read.tree.html">read.tree()</a></code> </p>
        </td>
        <td><p>Read a newick formatted phylogenetic tree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="repair.html">repair()</a></code> </p>
        </td>
        <td><p>Run after manually editing a BIOM object's content.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample.names.html">sample.names()</a></code> </p>
        </td>
        <td><p>Get the sample names.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sample.sums.html">sample.sums()</a></code> </p>
        </td>
        <td><p>Sum the observations in each sample.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="select.html">select()</a></code> </p>
        </td>
        <td><p>Reduce samples to a specific list</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="sequences.html">sequences()</a></code> </p>
        </td>
        <td><p>DNA sequence associated with each taxonomic identifier.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="stats.table.html">stats.table()</a></code> </p>
        </td>
        <td><p>Compute p-values for data vs metadata.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subset.html">subset(<i>&lt;BIOM&gt;</i>)</a></code> </p>
        </td>
        <td><p>Subset samples using the BIOM object's metadata</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="subtree.html">subtree()</a></code> </p>
        </td>
        <td><p>Create a subtree by specifying tips to keep.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="taxa.names.html">taxa.names()</a></code> </p>
        </td>
        <td><p>Get the taxa names.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="taxa.ranks.html">taxa.ranks()</a></code> </p>
        </td>
        <td><p>Get the taxa ranks.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="taxa.rollup.html">taxa.rollup()</a></code> </p>
        </td>
        <td><p>Generate a matrix of samples by taxa, at the specified taxonomic rank.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="taxonomy.html">taxonomy()</a></code> </p>
        </td>
        <td><p>Get the taxonomy table.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="tips.html">tips()</a></code> </p>
        </td>
        <td><p>Names of a phylogenetic tree's tips/leafs.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="unifrac.html">unifrac()</a></code> </p>
        </td>
        <td><p>Compute Weighted and Unweighted UniFrac distance matrices.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write.biom.html">write.biom()</a></code> </p>
        </td>
        <td><p>Write counts, metadata, taxonomy, and phylogeny to a biom file.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write.fasta.html">write.fasta()</a></code> </p>
        </td>
        <td><p>Write sequences from a BIOM object to a file in fasta format.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write.tree.html">write.tree()</a></code> </p>
        </td>
        <td><p>Write a newick formatted phylogenetic tree.</p></td>
      </tr><tr>
        
        <td>
          <p><code><a href="write.xlsx.html">write.xlsx()</a></code> </p>
        </td>
        <td><p>Write data and summary information to a Microsoft Excel-compatible workbook.</p></td>
      </tr>
    </tbody>
    </table>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Daniel P. Smith.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


