<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9003</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="adiv_boxplot.html">adiv_boxplot()</a></code> 
        </dt>
        <dd>Visualize alpha diversity with boxplots.</dd>
      </dl><dl><dt>
          
          <code><a href="adiv_corrplot.html">adiv_corrplot()</a></code> 
        </dt>
        <dd>Visualize alpha diversity with scatterplots and trendlines.</dd>
      </dl><dl><dt>
          
          <code><a href="adiv_table.html">adiv_table()</a></code> 
        </dt>
        <dd>Calculate the alpha diversity of each sample.</dd>
      </dl><dl><dt>
          
          <code><a href="as_percent.html">as_percent()</a></code> 
        </dt>
        <dd>Convert absolute counts to relative abundances.</dd>
      </dl><dl><dt>
          
          <code><a href="bdiv_boxplot.html">bdiv_boxplot()</a></code> 
        </dt>
        <dd>Visualize BIOM data with boxplots.</dd>
      </dl><dl><dt>
          
          <code><a href="bdiv_distmat.html">bdiv_distmat()</a></code> 
        </dt>
        <dd>Make a distance matrix of samples vs samples.</dd>
      </dl><dl><dt>
          
          <code><a href="bdiv_heatmap.html">bdiv_heatmap()</a></code> 
        </dt>
        <dd>Display beta diversities in an all vs all grid.</dd>
      </dl><dl><dt>
          
          <code><a href="bdiv_ord_plot.html">bdiv_ord_plot()</a></code> 
        </dt>
        <dd>Ordinate samples and taxa on a 2D plane based on beta diversity distances.</dd>
      </dl><dl><dt>
          
          <code><a href="bdiv_ord_table.html">bdiv_ord_table()</a></code> 
        </dt>
        <dd>Calculate PCoA and other ordinations, including taxa biplots and statistics.</dd>
      </dl><dl><dt>
          
          <code><a href="bdiv_table.html">bdiv_table()</a></code> 
        </dt>
        <dd>Make a data.frame of distances between samples.</dd>
      </dl><dl><dt>
          
          <code><a href="bdply.html">bdply()</a></code> 
        </dt>
        <dd>Split BIOM by metadata, apply function, and return results in a data frame.</dd>
      </dl><dl><dt>
          
          <code><a href="biom_build.html">biom_build()</a></code> 
        </dt>
        <dd>Create a BIOM object.</dd>
      </dl><dl><dt>
          
          <code><a href="biom_id.html">biom_id()</a></code> <code><a href="biom_id.html">`biom_id&lt;-`()</a></code> <code><a href="biom_id.html">biom_comment()</a></code> <code><a href="biom_id.html">`biom_comment&lt;-`()</a></code> 
        </dt>
        <dd>Get or set a <code>BIOM</code> object's id or comment.</dd>
      </dl><dl><dt>
          
          <code><a href="biom_info.html">biom_info()</a></code> 
        </dt>
        <dd>Get <code>BIOM</code> object's miscellaneous information.</dd>
      </dl><dl><dt>
          
          <code><a href="biom_merge.html">biom_merge()</a></code> 
        </dt>
        <dd>Combine several BIOM datasets into one.</dd>
      </dl><dl><dt>
          
          <code><a href="biom_repair.html">biom_repair()</a></code> 
        </dt>
        <dd>Run after manually editing a BIOM object's content.</dd>
      </dl><dl><dt>
          
          <code><a href="blply.html">blply()</a></code> 
        </dt>
        <dd>Split BIOM by metadata, apply function, and return results in a list.</dd>
      </dl><dl><dt>
          
          <code><a href="convert_from_phyloseq.html">convert_from_phyloseq()</a></code> 
        </dt>
        <dd>Convert a phyloseq object to an rbiom BIOM object.</dd>
      </dl><dl><dt>
          
          <code><a href="convert_to_SummarizedExperiment.html">convert_to_SummarizedExperiment()</a></code> 
        </dt>
        <dd>Convert a BIOM object to a SummarizedExperiment object.</dd>
      </dl><dl><dt>
          
          <code><a href="convert_to_TreeSummarizedExperiment.html">convert_to_TreeSummarizedExperiment()</a></code> 
        </dt>
        <dd>Convert a BIOM object to a TreeSummarizedExperiment object.</dd>
      </dl><dl><dt>
          
          <code><a href="distmat_stats.html">distmat_stats()</a></code> 
        </dt>
        <dd>Run statistics on a distance matrix vs a categorical or numeric variable.</dd>
      </dl><dl><dt>
          
          <code><a href="hmp50.html">hmp50</a></code> 
        </dt>
        <dd>Human Microbiome Project - 50 sample demo dataset.</dd>
      </dl><dl><dt>
          
          <code><a href="n_samples.html">n_samples()</a></code> <code><a href="n_samples.html">n_otus()</a></code> <code><a href="n_samples.html">n_ranks()</a></code> 
        </dt>
        <dd>Report the number of samples, otus, or ranks in a BIOM.</dd>
      </dl><dl><dt>
          
          <code><a href="otu_matrix.html">otu_matrix()</a></code> <code><a href="otu_matrix.html">`otu_matrix&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the abundance counts.</dd>
      </dl><dl><dt>
          
          <code><a href="otu_names.html">otu_names()</a></code> <code><a href="otu_names.html">`otu_names&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the OTU names.</dd>
      </dl><dl><dt>
          
          <code><a href="otu_sequences.html">otu_sequences()</a></code> <code><a href="otu_sequences.html">`otu_sequences&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the nucleotide sequences associated with each OTU.</dd>
      </dl><dl><dt>
          
          <code><a href="otu_taxonomy.html">otu_taxonomy()</a></code> <code><a href="otu_taxonomy.html">`otu_taxonomy&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the taxonomy table.</dd>
      </dl><dl><dt>
          
          <code><a href="otu_tree.html">otu_tree()</a></code> <code><a href="otu_tree.html">`otu_tree&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the phylogenetic tree.</dd>
      </dl><dl><dt>
          
          <code><a href="plot_heatmap.html">plot_heatmap()</a></code> 
        </dt>
        <dd>Create a heatmap with tracks and dendrograms from any matrix.</dd>
      </dl><dl><dt>
          
          <code><a href="print.html">print(<i>&lt;BIOM&gt;</i>)</a></code> 
        </dt>
        <dd>Summarize the contents of a BIOM object</dd>
      </dl><dl><dt>
          
          <code><a href="properties.html">has_tree()</a></code> <code><a href="properties.html">has_sequences()</a></code> <code><a href="properties.html">is_rarefied()</a></code> 
        </dt>
        <dd>Check specific properties of a BIOM object.</dd>
      </dl><dl><dt>
          
          <code><a href="rare_barplot.html">rare_barplot()</a></code> 
        </dt>
        <dd>Visualize the number of observations per sample.</dd>
      </dl><dl><dt>
          
          <code><a href="rare_corrplot.html">rare_corrplot()</a></code> 
        </dt>
        <dd>Visualize rarefaction curves with scatterplots and trendlines.</dd>
      </dl><dl><dt>
          
          <code><a href="rare_depth.html">rare_depth()</a></code> 
        </dt>
        <dd>If the samples are rarefied, report the level (depth).</dd>
      </dl><dl><dt>
          
          <code><a href="rare_multiplot.html">rare_multiplot()</a></code> 
        </dt>
        <dd>Combines rare_corrplot and rare_barplot into a single figure.</dd>
      </dl><dl><dt>
          
          <code><a href="rare_suggest.html">rare_suggest()</a></code> 
        </dt>
        <dd>Suggest a 'good' rarefaction depth.</dd>
      </dl><dl><dt>
          
          <code><a href="rbiom.html">rbiom</a></code> 
        </dt>
        <dd>rbiom: Read/Write, Transform, and Summarize BIOM Data</dd>
      </dl><dl><dt>
          
          <code><a href="read_biom.html">read_biom()</a></code> 
        </dt>
        <dd>Extracts counts, metadata, taxonomy, and phylogeny from a biom file.</dd>
      </dl><dl><dt>
          
          <code><a href="read_fasta.html">read_fasta()</a></code> 
        </dt>
        <dd>Parse a fasta file into a named character vector.</dd>
      </dl><dl><dt>
          
          <code><a href="read_tree.html">read_tree()</a></code> 
        </dt>
        <dd>Read a newick formatted phylogenetic tree.</dd>
      </dl><dl><dt>
          
          <code><a href="sample_metadata.html">sample_metadata()</a></code> <code><a href="sample_metadata.html">`sample_metadata&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the sample metadata.</dd>
      </dl><dl><dt>
          
          <code><a href="sample_names.html">sample_names()</a></code> <code><a href="sample_names.html">`sample_names&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the sample names.</dd>
      </dl><dl><dt>
          
          <code><a href="sample_rarefy.html">sample_rarefy()</a></code> 
        </dt>
        <dd>Subset counts so that all samples have the same number of observations.</dd>
      </dl><dl><dt>
          
          <code><a href="sample_select.html">sample_select()</a></code> 
        </dt>
        <dd>Reduce samples to a specific list</dd>
      </dl><dl><dt>
          
          <code><a href="sample_subset.html">sample_subset()</a></code> 
        </dt>
        <dd>Subset samples using the BIOM's metadata or taxonomy.</dd>
      </dl><dl><dt>
          
          <code><a href="sample_sums.html">sample_sums()</a></code> 
        </dt>
        <dd>Sum the observations in each sample.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_barplot.html">taxa_barplot()</a></code> 
        </dt>
        <dd>Display taxa abundances as a stacked bar graph.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_boxplot.html">taxa_boxplot()</a></code> 
        </dt>
        <dd>Visualize BIOM data with boxplots.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_corrplot.html">taxa_corrplot()</a></code> 
        </dt>
        <dd>Visualize taxa abundance with scatterplots and trendlines.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_heatmap.html">taxa_heatmap()</a></code> 
        </dt>
        <dd>Display taxa abundances as a heatmap.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_matrix.html">taxa_matrix()</a></code> 
        </dt>
        <dd>Taxa abundances per sample, at the specified taxonomic rank.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_means.html">taxa_means()</a></code> 
        </dt>
        <dd>Get the taxa abundances.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_ranks.html">taxa_ranks()</a></code> <code><a href="taxa_ranks.html">`taxa_ranks&lt;-`()</a></code> 
        </dt>
        <dd>Get or set the names of the taxonomic ranks.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_sums.html">taxa_sums()</a></code> 
        </dt>
        <dd>Get the taxa abundances.</dd>
      </dl><dl><dt>
          
          <code><a href="taxa_table.html">taxa_table()</a></code> 
        </dt>
        <dd>Generate a data.frame of taxa abundance for each sample at the specified taxonomic rank.</dd>
      </dl><dl><dt>
          
          <code><a href="top_taxa.html">top_taxa()</a></code> 
        </dt>
        <dd>Get the names of the most abundant taxa.</dd>
      </dl><dl><dt>
          
          <code><a href="tree_data.html">tree_data()</a></code> 
        </dt>
        <dd>Provides a 'treedata' S4 object for use in ggtree functions.</dd>
      </dl><dl><dt>
          
          <code><a href="tree_plot.html">tree_plot()</a></code> 
        </dt>
        <dd>Display a dendrogram of the phylogenetic tree.</dd>
      </dl><dl><dt>
          
          <code><a href="tree_subset.html">tree_subset()</a></code> 
        </dt>
        <dd>Create a subtree by specifying tips to keep.</dd>
      </dl><dl><dt>
          
          <code><a href="tree_tips.html">tree_tips()</a></code> 
        </dt>
        <dd>Names of a phylogenetic tree's tips/leafs.</dd>
      </dl><dl><dt>
          
          <code><a href="unifrac.html">unifrac()</a></code> 
        </dt>
        <dd>Compute Weighted and Unweighted UniFrac distance matrices.</dd>
      </dl><dl><dt>
          
          <code><a href="write_biom.html">write_biom()</a></code> 
        </dt>
        <dd>Write counts, metadata, taxonomy, and phylogeny to a biom file.</dd>
      </dl><dl><dt>
          
          <code><a href="write_counts.html">write_counts()</a></code> 
        </dt>
        <dd>Write BIOM counts to a tab-separated value (tsv) file.</dd>
      </dl><dl><dt>
          
          <code><a href="write_fasta.html">write_fasta()</a></code> 
        </dt>
        <dd>Write DNA sequences to a file in fasta format.</dd>
      </dl><dl><dt>
          
          <code><a href="write_metadata.html">write_metadata()</a></code> 
        </dt>
        <dd>Write BIOM metadata to a tab-separated value (tsv) file.</dd>
      </dl><dl><dt>
          
          <code><a href="write_taxonomy.html">write_taxonomy()</a></code> 
        </dt>
        <dd>Write BIOM taxonomy map to a tab-separated value (tsv) file.</dd>
      </dl><dl><dt>
          
          <code><a href="write_tree.html">write_tree()</a></code> 
        </dt>
        <dd>Write a newick formatted phylogenetic tree.</dd>
      </dl><dl><dt>
          
          <code><a href="write_xlsx.html">write_xlsx()</a></code> 
        </dt>
        <dd>Write data and summary information to a Microsoft Excel-compatible workbook.</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Daniel P. Smith.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

