<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ rbiom</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rbiom</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.3.9063</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/cmmr/rbiom/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="adiv_boxplot.html">adiv_boxplot()</a></code> </p>
        </td>
        <td><p>Visualize alpha diversity with boxplots.</p></td>
      </tr><tr><td>
          <p><code><a href="adiv_corrplot.html">adiv_corrplot()</a></code> </p>
        </td>
        <td><p>Visualize alpha diversity with scatterplots and trendlines.</p></td>
      </tr><tr><td>
          <p><code><a href="adiv_table.html">adiv_table()</a></code> </p>
        </td>
        <td><p>Estimate the diversity of each sample.</p></td>
      </tr><tr><td>
          <p><code><a href="apcoa.html">apcoa()</a></code> </p>
        </td>
        <td><p>PCoA adjusted for corvariates.</p></td>
      </tr><tr><td>
          <p><code><a href="as_percent.html">as_percent()</a></code> </p>
        </td>
        <td><p>Convert absolute counts to relative abundances.</p></td>
      </tr><tr><td>
          <p><code><a href="bdiv_biplot.html">bdiv_biplot()</a></code> </p>
        </td>
        <td><p>Display distances between samples as a 2-D scatter plot.</p></td>
      </tr><tr><td>
          <p><code><a href="bdiv_boxplot.html">bdiv_boxplot()</a></code> </p>
        </td>
        <td><p>Visualize BIOM data with boxplots.</p></td>
      </tr><tr><td>
          <p><code><a href="bdiv_distmat.html">bdiv_distmat()</a></code> </p>
        </td>
        <td><p>Make a distance matrix of samples vs samples.</p></td>
      </tr><tr><td>
          <p><code><a href="bdiv_heatmap.html">bdiv_heatmap()</a></code> </p>
        </td>
        <td><p>Display beta diversities in an all vs all grid.</p></td>
      </tr><tr><td>
          <p><code><a href="bdiv_ord_table.html">bdiv_ord_table()</a></code> </p>
        </td>
        <td><p>Reduce a distance matrix to two or three dimensions</p></td>
      </tr><tr><td>
          <p><code><a href="bdiv_table.html">bdiv_table()</a></code> </p>
        </td>
        <td><p>Make a data.frame of distances between samples.</p></td>
      </tr><tr><td>
          <p><code><a href="bdply.html">bdply()</a></code> </p>
        </td>
        <td><p>Split BIOM by metadata, apply function, and return results in a data frame.</p></td>
      </tr><tr><td>
          <p><code><a href="blply.html">blply()</a></code> </p>
        </td>
        <td><p>Split BIOM by metadata, apply function, and return results in a list.</p></td>
      </tr><tr><td>
          <p><code><a href="counts-set.html">`counts&lt;-`()</a></code> </p>
        </td>
        <td><p>Set the abundance counts.</p></td>
      </tr><tr><td>
          <p><code><a href="counts.html">counts()</a></code> </p>
        </td>
        <td><p>Get the abundance counts.</p></td>
      </tr><tr><td>
          <p><code><a href="depth.html">depth()</a></code> </p>
        </td>
        <td><p>The rarefaction depth of a BIOM object.</p></td>
      </tr><tr><td>
          <p><code><a href="distill.html">distill()</a></code> </p>
        </td>
        <td><p>Convert to a data.frame with metadata and diversity/abundance values.</p></td>
      </tr><tr><td>
          <p><code><a href="embed_code.html">embed_code()</a></code> </p>
        </td>
        <td><p>Wraps R code with Markdown syntax highlighting.</p></td>
      </tr><tr><td>
          <p><code><a href="embed_csv.html">embed_csv()</a></code> </p>
        </td>
        <td><p>Converts a data.frame to a download link.</p></td>
      </tr><tr><td>
          <p><code><a href="has_phylogeny.html">has_phylogeny()</a></code> </p>
        </td>
        <td><p>Checks if a phylogenetic tree is present.</p></td>
      </tr><tr><td>
          <p><code><a href="has_sequences.html">has_sequences()</a></code> </p>
        </td>
        <td><p>Checks if DNA sequences are present.</p></td>
      </tr><tr><td>
          <p><code><a href="hmp50.html">hmp50</a></code> </p>
        </td>
        <td><p>Human Microbiome Project - 50 sample demo dataset.</p></td>
      </tr><tr><td>
          <p><code><a href="id-set.html">`id&lt;-`()</a></code> </p>
        </td>
        <td><p>Set a <code>BIOM</code> object's id or comment.</p></td>
      </tr><tr><td>
          <p><code><a href="id.html">id()</a></code> </p>
        </td>
        <td><p>Get <code>BIOM</code> object's identifier / title.</p></td>
      </tr><tr><td>
          <p><code><a href="info.html">info()</a></code> </p>
        </td>
        <td><p>Get <code>BIOM</code> object's miscellaneous information.</p></td>
      </tr><tr><td>
          <p><code><a href="init_cache.html">init_cache()</a></code> </p>
        </td>
        <td><p>Initialize caching for expensive biom operations.</p></td>
      </tr><tr><td>
          <p><code><a href="is_rarefied.html">is_rarefied()</a></code> </p>
        </td>
        <td><p>Checks if a BIOM object is rarefied.</p></td>
      </tr><tr><td>
          <p><code><a href="metadata-set.html">`metadata&lt;-`()</a></code> </p>
        </td>
        <td><p>Set the <code>BIOM</code> object's metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="metadata.html">metadata()</a></code> </p>
        </td>
        <td><p>Get the sample metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="metrics.html">metrics()</a></code> </p>
        </td>
        <td><p>List all the options for each type of metric.</p></td>
      </tr><tr><td>
          <p><code><a href="nsamples.html">nsamples()</a></code> </p>
        </td>
        <td><p>Number of samples in a BIOM.</p></td>
      </tr><tr><td>
          <p><code><a href="ntaxa.html">ntaxa()</a></code> </p>
        </td>
        <td><p>Number of taxa in a BIOM.</p></td>
      </tr><tr><td>
          <p><code><a href="phylogeny-set.html">`phylogeny&lt;-`()</a></code> </p>
        </td>
        <td><p>Set the phylogenetic tree.</p></td>
      </tr><tr><td>
          <p><code><a href="phylogeny.html">phylogeny()</a></code> </p>
        </td>
        <td><p>Get the phylogenetic tree.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_heatmap.html">plot_heatmap()</a></code> </p>
        </td>
        <td><p>Create a heatmap with tracks and dendrograms from any matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="print.html">print(<i>&lt;BIOM&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summarize the contents of a BIOM object</p></td>
      </tr><tr><td>
          <p><code><a href="rarefy.html">rarefy()</a></code> </p>
        </td>
        <td><p>Subset counts so that all samples have the same number of observations.</p></td>
      </tr><tr><td>
          <p><code><a href="rare_corrplot.html">rare_corrplot()</a></code> </p>
        </td>
        <td><p>Visualize rarefaction curves with scatterplots and trendlines.</p></td>
      </tr><tr><td>
          <p><code><a href="rbiom.html">rbiom</a></code> </p>
        </td>
        <td><p>rbiom: Read/Write, Transform, and Summarize BIOM Data</p></td>
      </tr><tr><td>
          <p><code><a href="read_biom.html">read_biom()</a></code> </p>
        </td>
        <td><p>Extracts counts, metadata, taxonomy, and phylogeny from a biom file.</p></td>
      </tr><tr><td>
          <p><code><a href="read_fasta.html">read_fasta()</a></code> </p>
        </td>
        <td><p>Parse a fasta file into a named character vector.</p></td>
      </tr><tr><td>
          <p><code><a href="read_tree.html">read_tree()</a></code> </p>
        </td>
        <td><p>Read a newick formatted phylogenetic tree.</p></td>
      </tr><tr><td>
          <p><code><a href="repair.html">repair()</a></code> </p>
        </td>
        <td><p>Run after manually editing a BIOM object's content.</p></td>
      </tr><tr><td>
          <p><code><a href="sample_names-set.html">`sample_names&lt;-`()</a></code> </p>
        </td>
        <td><p>Set the sample names.</p></td>
      </tr><tr><td>
          <p><code><a href="sample_names.html">sample_names()</a></code> </p>
        </td>
        <td><p>Get the sample names.</p></td>
      </tr><tr><td>
          <p><code><a href="sample_sums.html">sample_sums()</a></code> </p>
        </td>
        <td><p>Sum the observations in each sample.</p></td>
      </tr><tr><td>
          <p><code><a href="select.html">select()</a></code> </p>
        </td>
        <td><p>Reduce samples to a specific list</p></td>
      </tr><tr><td>
          <p><code><a href="sequences-set.html">`sequences&lt;-`()</a></code> </p>
        </td>
        <td><p>Set nucleotide sequences associated with each taxonomic identifier.</p></td>
      </tr><tr><td>
          <p><code><a href="sequences.html">sequences()</a></code> </p>
        </td>
        <td><p>Nucleotide sequences associated with each taxonomic identifier.</p></td>
      </tr><tr><td>
          <p><code><a href="stats_table.html">stats_table()</a></code> </p>
        </td>
        <td><p>Compute p-values for data vs metadata.</p></td>
      </tr><tr><td>
          <p><code><a href="subset.html">subset(<i>&lt;BIOM&gt;</i>)</a></code> </p>
        </td>
        <td><p>Subset samples using the BIOM's metadata or taxonomy.</p></td>
      </tr><tr><td>
          <p><code><a href="subtree.html">subtree()</a></code> </p>
        </td>
        <td><p>Create a subtree by specifying tips to keep.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_boxplot.html">taxa_boxplot()</a></code> </p>
        </td>
        <td><p>Visualize BIOM data with boxplots.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_corrplot.html">taxa_corrplot()</a></code> </p>
        </td>
        <td><p>Visualize taxa abundance with scatterplots and trendlines.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_heatmap.html">taxa_heatmap()</a></code> </p>
        </td>
        <td><p>Display taxa abundances as a heatmap.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_matrix.html">taxa_matrix()</a></code> </p>
        </td>
        <td><p>Generate a matrix of samples by taxa, at the specified taxonomic rank.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_means.html">taxa_means()</a></code> </p>
        </td>
        <td><p>Get the taxa abundances.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_names-set.html">`taxa_names&lt;-`()</a></code> </p>
        </td>
        <td><p>Set the taxa names.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_names.html">taxa_names()</a></code> </p>
        </td>
        <td><p>Get the taxa names.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_ranks-set.html">`taxa_ranks&lt;-`()</a></code> </p>
        </td>
        <td><p>Set the taxa rank names.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_ranks.html">taxa_ranks()</a></code> </p>
        </td>
        <td><p>Get the taxa ranks.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_stacked.html">taxa_stacked()</a></code> </p>
        </td>
        <td><p>Display taxa abundances as a stacked bar graph.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_sums.html">taxa_sums()</a></code> </p>
        </td>
        <td><p>Get the taxa abundances.</p></td>
      </tr><tr><td>
          <p><code><a href="taxa_table.html">taxa_table()</a></code> </p>
        </td>
        <td><p>Generate a data.frame of taxa abundance for each sample at the specified taxonomic rank.</p></td>
      </tr><tr><td>
          <p><code><a href="taxonomy-set.html">`taxonomy&lt;-`()</a></code> </p>
        </td>
        <td><p>Set the taxonomy table.</p></td>
      </tr><tr><td>
          <p><code><a href="taxonomy.html">taxonomy()</a></code> </p>
        </td>
        <td><p>Get the taxonomy table.</p></td>
      </tr><tr><td>
          <p><code><a href="tips.html">tips()</a></code> </p>
        </td>
        <td><p>Names of a phylogenetic tree's tips/leafs.</p></td>
      </tr><tr><td>
          <p><code><a href="top_taxa.html">top_taxa()</a></code> </p>
        </td>
        <td><p>Get the names of the most abundant taxa.</p></td>
      </tr><tr><td>
          <p><code><a href="unifrac.html">unifrac()</a></code> </p>
        </td>
        <td><p>Compute Weighted and Unweighted UniFrac distance matrices.</p></td>
      </tr><tr><td>
          <p><code><a href="write_biom.html">write_biom()</a></code> </p>
        </td>
        <td><p>Write counts, metadata, taxonomy, and phylogeny to a biom file.</p></td>
      </tr><tr><td>
          <p><code><a href="write_fasta.html">write_fasta()</a></code> </p>
        </td>
        <td><p>Write sequences from a BIOM object to a file in fasta format.</p></td>
      </tr><tr><td>
          <p><code><a href="write_tree.html">write_tree()</a></code> </p>
        </td>
        <td><p>Write a newick formatted phylogenetic tree.</p></td>
      </tr><tr><td>
          <p><code><a href="write_xlsx.html">write_xlsx()</a></code> </p>
        </td>
        <td><p>Write data and summary information to a Microsoft Excel-compatible workbook.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Daniel P. Smith.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

