<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Taxa abundances per sample. — taxa_table • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Taxa abundances per sample. — taxa_table"><meta name="description" content="

taxa_matrix() -
Accepts a single rank and returns a matrix.

taxa_table() -
Can accept more than one rank and returns a tibble data.frame.



"><meta property="og:description" content="

taxa_matrix() -
Accepts a single rank and returns a matrix.

taxa_table() -
Can accept more than one rank and returns a tibble data.frame.



"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9135</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Taxa abundances per sample.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/HEAD/R/taxa_table.r" class="external-link"><code>R/taxa_table.r</code></a></small>
      <div class="d-none name"><code>taxa_matrix.Rd</code></div>
    </div>

    <div class="ref-description section level2">

<dl><dt><code>taxa_matrix()</code> - </dt>
<dd><p>Accepts a single <code>rank</code> and returns a matrix.</p></dd>

<dt><code>taxa_table()</code> - </dt>
<dd><p>Can accept more than one <code>rank</code> and returns a tibble data.frame.</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">taxa_table</span><span class="op">(</span></span>
<span>  <span class="va">biom</span>,</span>
<span>  rank <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  taxa <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  lineage <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  md <span class="op">=</span> <span class="st">".all"</span>,</span>
<span>  unc <span class="op">=</span> <span class="st">"singly"</span>,</span>
<span>  other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trans <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  ties <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">taxa_matrix</span><span class="op">(</span></span>
<span>  <span class="va">biom</span>,</span>
<span>  rank <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  taxa <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  lineage <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sparse <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  unc <span class="op">=</span> <span class="st">"singly"</span>,</span>
<span>  other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  trans <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  ties <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-biom">biom<a class="anchor" aria-label="anchor" href="#arg-biom"></a></dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt id="arg-rank">rank<a class="anchor" aria-label="anchor" href="#arg-rank"></a></dt>
<dd><p>What rank(s) of taxa to display. E.g. <code>"Phylum"</code>,
<code>"Genus"</code>, <code>".otu"</code>, etc. An integer vector can also be
given, where <code>1</code> is the highest rank, <code>2</code> is the second
highest, <code>-1</code> is the lowest rank, <code>-2</code> is the second
lowest, and <code>0</code> is the OTU "rank". Run <code>biom$ranks</code> to
see all options for a given rbiom object. Default: <code>-1</code>.</p></dd>


<dt id="arg-taxa">taxa<a class="anchor" aria-label="anchor" href="#arg-taxa"></a></dt>
<dd><p>Which taxa to display. An integer value will show the top n
most abundant taxa. A value 0 &lt;= n &lt; 1 will show any taxa with that
mean abundance or greater (e.g. <code>0.1</code> implies &gt;= 10%). A
character vector of taxa names will show only those named taxa.
Default: <code>6</code>.</p></dd>


<dt id="arg-lineage">lineage<a class="anchor" aria-label="anchor" href="#arg-lineage"></a></dt>
<dd><p>Include all ranks in the name of the taxa. For instance,
setting to <code>TRUE</code> will produce
<code>Bacteria; Actinobacteria; Coriobacteriia; Coriobacteriales</code>.
Otherwise the taxa name will simply be <code>Coriobacteriales</code>. You
want to set this to TRUE when <code>unc = "asis"</code> and you have taxa
names (such as <em>Incertae_Sedis</em>) that map to multiple higher
level ranks. Default: <code>FALSE</code></p></dd>


<dt id="arg-md">md<a class="anchor" aria-label="anchor" href="#arg-md"></a></dt>
<dd><p>Dataset field(s) to include in the output data frame, or <code>'.all'</code>
to include all metadata fields. Default: <code>'.all'</code></p></dd>


<dt id="arg-unc">unc<a class="anchor" aria-label="anchor" href="#arg-unc"></a></dt>
<dd><p>How to handle unclassified, uncultured, and similarly ambiguous
taxa names. Options are:</p><dl><dt><code>"singly"</code> - </dt>
<dd><p>Replaces them with the OTU name.</p></dd>

<dt><code>"grouped"</code> - </dt>
<dd><p>Replaces them with a higher rank's name.</p></dd>

<dt><code>"drop"</code> - </dt>
<dd><p>Excludes them from the result.</p></dd>

<dt><code>"asis"</code> - </dt>
<dd><p>To not check/modify any taxa names.</p></dd>


</dl><p>Default: <code>"singly"</code> <br><br>
Abbreviations are allowed.</p></dd>


<dt id="arg-other">other<a class="anchor" aria-label="anchor" href="#arg-other"></a></dt>
<dd><p>Sum all non-itemized taxa into an "Other" taxa. When
<code>FALSE</code>, only returns taxa matched by the <code>taxa</code>
argument. Specifying <code>TRUE</code> adds "Other" to the returned set.
A string can also be given to imply <code>TRUE</code>, but with that
value as the name to use instead of "Other".
Default: <code>FALSE</code></p></dd>


<dt id="arg-trans">trans<a class="anchor" aria-label="anchor" href="#arg-trans"></a></dt>
<dd><p>Transformation to apply. Options are:
<code>c("none", "rank", "log", "log1p", "sqrt")</code>. <code>"rank"</code> is useful for
correcting for non-normally distributions before applying regression
statistics. Default: <code>"none"</code></p></dd>


<dt id="arg-ties">ties<a class="anchor" aria-label="anchor" href="#arg-ties"></a></dt>
<dd><p>When <code>trans="rank"</code>, how to rank identical values.
Options are: <code>c("average", "first", "last", "random", "max", "min")</code>.
See <code><a href="https://rdrr.io/r/base/rank.html" class="external-link">rank()</a></code> for details. Default: <code>"random"</code></p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed for permutations. Default: <code>0</code></p></dd>


<dt id="arg-sparse">sparse<a class="anchor" aria-label="anchor" href="#arg-sparse"></a></dt>
<dd><p>If true, returns a sparse matrix as described by
<code>simple_triplet_matrix()</code>, otherwise returns a normal R
matrix object. Default: <code>FALSE</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>

<dl><dt><code>taxa_matrix()</code> - </dt>
<dd><p>A numeric matrix with taxa as rows, and samples as columns.</p></dd>

<dt><code>taxa_table()</code> - </dt>
<dd><p>A tibble data frame with column names .sample, .taxa, .abundance, and any requested by <code>md</code>.</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other taxa_abundance:
<code><a href="taxa_boxplot.html">taxa_boxplot</a>()</code>,
<code><a href="taxa_clusters.html">taxa_clusters</a>()</code>,
<code><a href="taxa_corrplot.html">taxa_corrplot</a>()</code>,
<code><a href="taxa_heatmap.html">taxa_heatmap</a>()</code>,
<code><a href="taxa_stacked.html">taxa_stacked</a>()</code>,
<code><a href="taxa_stats.html">taxa_stats</a>()</code>,
<code><a href="taxa_sums.html">taxa_sums</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">biom</span><span class="op">$</span><span class="va">ranks</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'biom' not found</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">taxa_matrix</span><span class="op">(</span><span class="va">hmp50</span>, <span class="st">'Phylum'</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                     HMP01 HMP02 HMP03 HMP04 HMP05 HMP06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Actinobacteria         18    60   126   120    30    71</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Bacteroidetes         276   221   313   218   144   880</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cyanobacteria           0     0     0     0     0     0</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Deinococcus Thermus     0     0     0     0     0     0</span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">taxa_table</span><span class="op">(</span><span class="va">hmp50</span>, <span class="st">'Phylum'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 300 × 8</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .rank  .sample .taxa          .abundance   Age   BMI `Body Site`   Sex   </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Phylum HMP01   Firmicutes           <span style="text-decoration: underline;">1</span>208    22    20 Buccal mucosa Female</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> Phylum HMP01   Bacteroidetes         276    22    20 Buccal mucosa Female</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Phylum HMP01   Actinobacteria         18    22    20 Buccal mucosa Female</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Phylum HMP01   Proteobacteria        103    22    20 Buccal mucosa Female</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Phylum HMP01   Fusobacteria           41    22    20 Buccal mucosa Female</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Phylum HMP01   Tenericutes             0    22    20 Buccal mucosa Female</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Phylum HMP02   Firmicutes            931    24    23 Buccal mucosa Male  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> Phylum HMP02   Bacteroidetes         221    24    23 Buccal mucosa Male  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Phylum HMP02   Actinobacteria         60    24    23 Buccal mucosa Male  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Phylum HMP02   Proteobacteria        112    24    23 Buccal mucosa Male  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 290 more rows</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

