<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="A convenience wrapper for taxa_table() + stats_table()."><title>Test taxa abundances for associations with metadata. — taxa_stats • rbiom</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Test taxa abundances for associations with metadata. — taxa_stats"><meta property="og:description" content="A convenience wrapper for taxa_table() + stats_table()."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0.9129</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/rbiom.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6>
    <a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a>
    <a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a>
    <a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a>
    <a class="dropdown-item" href="../articles/statistics.html">Statistics</a>
    <a class="dropdown-item" href="../articles/regression.html">Regression</a>
    <a class="dropdown-item" href="../reference/speed.html">Speed Ups</a>
    <div class="dropdown-divider"></div>
    <h6 class="dropdown-header" data-toc-skip>Plotting</h6>
    <a class="dropdown-item" href="../articles/plots.html">Plot Types</a>
    <a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Test taxa abundances for associations with metadata.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/HEAD/R/stats_table.r" class="external-link"><code>R/stats_table.r</code></a></small>
      <div class="d-none name"><code>taxa_stats.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A convenience wrapper for <code><a href="taxa_matrix.html">taxa_table()</a></code> + <code><a href="stats_table.html">stats_table()</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">taxa_stats</span><span class="op">(</span></span>
<span>  <span class="va">biom</span>,</span>
<span>  regr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stat.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rank <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>,</span>
<span>  taxa <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  lineage <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  unc <span class="op">=</span> <span class="st">"singly"</span>,</span>
<span>  other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  split.by <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  trans <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  test <span class="op">=</span> <span class="st">"emmeans"</span>,</span>
<span>  fit <span class="op">=</span> <span class="st">"gam"</span>,</span>
<span>  at <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  level <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  alt <span class="op">=</span> <span class="st">"!="</span>,</span>
<span>  mu <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  p.adj <span class="op">=</span> <span class="st">"fdr"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>biom</dt>
<dd><p>An <a href="rbiom_objects.html">rbiom object</a>, such as from <code><a href="as_rbiom.html">as_rbiom()</a></code>.
Any value accepted by <code><a href="as_rbiom.html">as_rbiom()</a></code> can also be given here.</p></dd>


<dt>regr</dt>
<dd><p>Dataset field with the x-axis (independent; predictive)
values. Must be numeric. Default: <code>NULL</code></p></dd>


<dt>stat.by</dt>
<dd><p>Dataset field with the statistical groups. Must be
categorical. Default: <code>NULL</code></p></dd>


<dt>rank</dt>
<dd><p>What rank(s) of taxa to display. E.g. <code>"Phylum"</code>,
<code>"Genus"</code>, <code>".otu"</code>, etc. An integer vector can also be
given, where <code>1</code> is the highest rank, <code>2</code> is the second
highest, <code>-1</code> is the lowest rank, <code>-2</code> is the second
lowest, and <code>0</code> is the OTU "rank". Run <code>biom$ranks</code> to
see all options for a given rbiom object. Default: <code>-1</code>.</p></dd>


<dt>taxa</dt>
<dd><p>Which taxa to display. An integer value will show the top n
most abundant taxa. A value 0 &lt;= n &lt; 1 will show any taxa with that
mean abundance or greater (e.g. <code>0.1</code> implies &gt;= 10%). A
character vector of taxa names will show only those named taxa.
Default: <code>6</code>.</p></dd>


<dt>lineage</dt>
<dd><p>Include all ranks in the name of the taxa. For instance,
setting to <code>TRUE</code> will produce
<code>Bacteria; Actinobacteria; Coriobacteriia; Coriobacteriales</code>.
Otherwise the taxa name will simply be <code>Coriobacteriales</code>. You
want to set this to TRUE when <code>unc = "asis"</code> and you have taxa
names (such as <em>Incertae_Sedis</em>) that map to multiple higher
level ranks. Default: <code>FALSE</code></p></dd>


<dt>unc</dt>
<dd><p>How to handle unclassified, uncultured, and similarly ambiguous
taxa names. Options are:</p><ul><li><p><code>"singly"</code> -  Replaces them with the OTU name.</p></li>
<li><p><code>"grouped"</code> -  Replaces them with a higher rank's name.</p></li>
<li><p><code>"drop"</code> -  Excludes them from the result.</p></li>
<li><p><code>"asis"</code> -  To not check/modify any taxa names.</p></li>
</ul><p>Default: <code>"singly"</code> <br><br>
Abbreviations are allowed.</p></dd>


<dt>other</dt>
<dd><p>Sum all non-itemized taxa into an "Other" taxa. When
<code>FALSE</code>, only returns taxa matched by the <code>taxa</code>
argument. Specifying <code>TRUE</code> adds "Other" to the returned set.
A string can also be given to imply <code>TRUE</code>, but with that
value as the name to use instead of "Other".
Default: <code>FALSE</code></p></dd>


<dt>split.by</dt>
<dd><p>Dataset field(s) that the data should be split by prior to
any calculations. Must be categorical. Default: <code>NULL</code></p></dd>


<dt>trans</dt>
<dd><p>Transformation to apply. Options are:
<code>c("none", "rank", "log", "log1p", "sqrt")</code>. <code>"rank"</code> is useful for
correcting for non-normally distributions before applying regression
statistics. Default: <code>"none"</code></p></dd>


<dt>test</dt>
<dd><p>Method for computing p-values: <code>'wilcox'</code>, <code>'kruskal'</code>,
<code>'emmeans'</code>, or <code>'emtrends'</code>. Default: <code>'emmeans'</code></p></dd>


<dt>fit</dt>
<dd><p>How to fit the trendline. <code>'lm'</code>, <code>'log'</code>, or <code>'gam'</code>.
Default: <code>'gam'</code></p></dd>


<dt>at</dt>
<dd><p>Position(s) along the x-axis where the means or slopes should be
evaluated. Default: <code>NULL</code>, which samples 100 evenly spaced positions
and selects the position where the p-value is most significant.</p></dd>


<dt>level</dt>
<dd><p>The confidence level for calculating a confidence interval.
Default: <code>0.95</code></p></dd>


<dt>alt</dt>
<dd><p>Alternative hypothesis direction. Options are <code>'!='</code>
(two-sided; not equal to <code>mu</code>), <code>'&lt;'</code> (less than <code>mu</code>), or <code>'&gt;'</code>
(greater than <code>mu</code>). Default: <code>'!='</code></p></dd>


<dt>mu</dt>
<dd><p>Reference value to test against. Default: <code>0</code></p></dd>


<dt>p.adj</dt>
<dd><p>Method to use for multiple comparisons adjustment of
p-values. Run <code>p.adjust.methods</code> for a list of available
options. Default: <code>"fdr"</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A tibble data.frame with fields from the table below. This tibble
object provides the <code>$code</code> operator to print the R code used to generate
the statistics.</p><table class="table table"><tr><td><strong>Field</strong></td><td><strong>Description</strong></td></tr><tr><td>.mean</td><td>Estimated marginal mean. See <code><a href="https://rvlenth.github.io/emmeans/reference/emmeans.html" class="external-link">emmeans::emmeans()</a></code>.</td></tr><tr><td>.mean.diff</td><td>Difference in means.</td></tr><tr><td>.slope</td><td>Trendline slope. See <code><a href="https://rvlenth.github.io/emmeans/reference/emtrends.html" class="external-link">emmeans::emtrends()</a></code>.</td></tr><tr><td>.slope.diff</td><td>Difference in slopes.</td></tr><tr><td>.h1</td><td>Alternate hypothesis.</td></tr><tr><td>.p.val</td><td>Probability that null hypothesis is correct.</td></tr><tr><td>.adj.p</td><td><code>.p.val</code> after adjusting for multiple comparisons.</td></tr><tr><td>.effect.size</td><td>Effect size. See <code><a href="https://rvlenth.github.io/emmeans/reference/eff_size.html" class="external-link">emmeans::eff_size()</a></code>.</td></tr><tr><td>.lower</td><td>Confidence interval lower bound.</td></tr><tr><td>.upper</td><td>Confidence interval upper bound.</td></tr><tr><td>.se</td><td>Standard error.</td></tr><tr><td>.n</td><td>Number of samples.</td></tr><tr><td>.df</td><td>Degrees of freedom.</td></tr><tr><td>.stat</td><td>Wilcoxon or Kruskal-Wallis rank sum statistic.</td></tr><tr><td>.t.ratio</td><td><code>.mean</code> / <code>.se</code></td></tr><tr><td>.r.sqr</td><td>Percent of variation explained by the model.</td></tr><tr><td>.adj.r</td><td><code>.r.sqr</code>, taking degrees of freedom into account.</td></tr><tr><td>.aic</td><td>Akaike Information Criterion (predictive models).</td></tr><tr><td>.bic</td><td>Bayesian Information Criterion (descriptive models).</td></tr><tr><td>.loglik</td><td>Log-likelihood goodness-of-fit score.</td></tr><tr><td>.fit.p</td><td>P-value for observing this fit by chance.</td></tr></table></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Other taxa_abundance: 
<code><a href="taxa_boxplot.html">taxa_boxplot</a>()</code>,
<code><a href="taxa_corrplot.html">taxa_corrplot</a>()</code>,
<code><a href="taxa_heatmap.html">taxa_heatmap</a>()</code>,
<code><a href="taxa_stacked.html">taxa_stacked</a>()</code>,
<code><a href="taxa_sums.html">taxa_sums</a>()</code>,
<code><a href="taxa_matrix.html">taxa_table</a>()</code></p>
<p>Other stats_tables: 
<code><a href="adiv_stats.html">adiv_stats</a>()</code>,
<code><a href="bdiv_stats.html">bdiv_stats</a>()</code>,
<code><a href="distmat_stats.html">distmat_stats</a>()</code>,
<code><a href="stats_table.html">stats_table</a>()</code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span>    <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">hmp50</span><span class="op">)</span></span></span>
<span class="r-in"><span>    </span></span>
<span class="r-in"><span>    <span class="fu">taxa_stats</span><span class="op">(</span><span class="va">biom</span>, stat.by <span class="op">=</span> <span class="st">"Body Site"</span>, rank <span class="op">=</span> <span class="st">"Family"</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># Model:    gam(.abundance ~ `Body Site`, method = "REML")</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 60 × 6</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    .taxa            `Body Site`                .mean.diff .h1     .p.val  .adj.p</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 1</span> Bacteroidaceae   Mid vagina - Stool              -<span style="color: #BB0000;">603.</span> != 0  4.96<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span> 7.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 2</span> Bacteroidaceae   Saliva - Stool                  -<span style="color: #BB0000;">602.</span> != 0  5.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span> 7.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 3</span> Bacteroidaceae   Anterior nares - Stool          -<span style="color: #BB0000;">602.</span> != 0  5.13<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span> 7.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 4</span> Bacteroidaceae   Buccal mucosa - Stool           -<span style="color: #BB0000;">602.</span> != 0  5.14<span style="color: #949494;">e</span><span style="color: #BB0000;">-10</span> 7.72<span style="color: #949494;">e</span><span style="color: #BB0000;">-9</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 5</span> Streptococcaceae Buccal mucosa - Mid vagina       784. != 0  1.35<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 1.62<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 6</span> Streptococcaceae Anterior nares - Buccal m…      -<span style="color: #BB0000;">765.</span> != 0  2.10<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 2.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 7</span> Lactobacillaceae Buccal mucosa - Mid vagina     -<span style="color: #BB0000; text-decoration: underline;">1</span><span style="color: #BB0000;">048.</span> != 0  3.12<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 2.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 8</span> Lactobacillaceae Mid vagina - Saliva             <span style="text-decoration: underline;">1</span>048. != 0  3.12<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 2.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;"> 9</span> Lactobacillaceae Anterior nares - Mid vagi…     -<span style="color: #BB0000; text-decoration: underline;">1</span><span style="color: #BB0000;">047.</span> != 0  3.13<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 2.08<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">10</span> Streptococcaceae Buccal mucosa - Stool            785. != 0  5.05<span style="color: #949494;">e</span><span style="color: #BB0000;">- 8</span> 3.03<span style="color: #949494;">e</span><span style="color: #BB0000;">-7</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 50 more rows</span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

