<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>QIIME 2, mothur, and BioConductor • rbiom</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="QIIME 2, mothur, and BioConductor">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>QIIME 2, mothur, and BioConductor</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/vignettes/articles/convert.Rmd" class="external-link"><code>vignettes/articles/convert.Rmd</code></a></small>
      <div class="d-none name"><code>convert.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="qiime-2">QIIME 2<a class="anchor" aria-label="anchor" href="#qiime-2"></a>
</h2>
<div class="section level3">
<h3 id="rbiom-qiime-2">rbiom ➡ QIIME 2<a class="anchor" aria-label="anchor" href="#rbiom-qiime-2"></a>
</h3>
<p>In R, export QIIME 2-compatible files.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># where to save the files</span></span>
<span><span class="va">project_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/export.html">write_qiime2</a></span><span class="op">(</span>biom <span class="op">=</span> <span class="va">hmp50</span>, dir <span class="op">=</span> <span class="va">project_dir</span>, prefix <span class="op">=</span> <span class="st">'hmp50_'</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>This command creates files named <code>'hmp50_counts.tsv'</code>,
<code>'hmp50_metadata.tsv'</code>,
<code>'hmp50_taxonomy.tsv', 'hmp50_tree.nwk'</code>, and
<code>'hmp50_seqs.fna'</code>.</p>
</blockquote>
<p><br> On the command line, convert files to QIIME 2 format (.qza).</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Convert classic BIOM table to HDF5</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">biom</span> convert <span class="at">-i</span> hmp50_counts.tsv <span class="at">-o</span> hmp50_counts.hdf5 <span class="at">--to-hdf5</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># Import counts</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="ex">qiime</span> tools import <span class="dt">\</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="at">--input-path</span> hmp50_counts.hdf5 <span class="dt">\</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  <span class="at">--type</span> <span class="st">'FeatureTable[Frequency]'</span> <span class="dt">\</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>  <span class="at">--input-format</span> BIOMV210Format <span class="dt">\</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  <span class="at">--output-path</span> hmp50-counts.qza</span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a><span class="co"># Import taxonomy</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a><span class="ex">qiime</span> tools import <span class="dt">\</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a> <span class="at">--input-path</span> hmp50_taxonomy.tsv <span class="dt">\</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a> <span class="at">--type</span> FeatureData<span class="pp">[</span><span class="ss">Taxonomy</span><span class="pp">]</span> <span class="dt">\</span></span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a> <span class="at">--output-path</span> hmp50-taxonomy.qza</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co"># Import phylogenetic tree</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a><span class="ex">qiime</span> tools import <span class="dt">\</span></span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>  <span class="at">--input-path</span> hmp50_tree.nwk <span class="dt">\</span></span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>  <span class="at">--type</span> <span class="st">'Phylogeny[Rooted]'</span> <span class="dt">\</span></span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>  <span class="at">--output-path</span> hmp50-tree.qza</span></code></pre></div>
<p>Examples: running unifrac and browsing metadata</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">qiime</span> diversity-lib weighted-unifrac <span class="dt">\</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  <span class="at">--i-table</span> hmp50-counts.qza <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">--i-phylogeny</span> hmp50-tree.qza <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">--o-distance-matrix</span> weighted-unifrac-dm.qza</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  </span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="ex">qiime</span> metadata tabulate <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>  <span class="at">--m-input-file</span> hmp50_metadata.tsv <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>  <span class="at">--o-visualization</span> hmp50-metadata-browser.qzv</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a><span class="co"># View .qzv files with https://view.qiime2.org</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="qiime-2-rbiom">QIIME 2 ➡ rbiom<a class="anchor" aria-label="anchor" href="#qiime-2-rbiom"></a>
</h3>
<p>On the command line, export data files from QIIME 2.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Export a feature table, taxonomy, and tree</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">qiime</span> tools export <span class="at">--input-path</span> hmp50-counts.qza   <span class="at">--output-path</span> .</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="ex">qiime</span> tools export <span class="at">--input-path</span> hmp50-taxonomy.qza <span class="at">--output-path</span> .</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="ex">qiime</span> tools export <span class="at">--input-path</span> hmp50-tree.qza     <span class="at">--output-path</span> .</span></code></pre></div>
<blockquote>
<p>These commands create files named <code>'feature-table.biom'</code>,
<code>'taxonomy.tsv'</code>, and <code>'tree.nwk'</code>.</p>
</blockquote>
<p><br> In R, import the data files into rbiom.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># project_dir = directory with relevant files</span></span>
<span></span>
<span><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_dir.html" class="external-link">with_dir</a></span><span class="op">(</span><span class="va">project_dir</span>, <span class="op">{</span></span>
<span></span>
<span>  <span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span></span>
<span>    biom     <span class="op">=</span> <span class="st">'feature-table.biom'</span>, </span>
<span>    metadata <span class="op">=</span> <span class="st">'hmp50_metadata.tsv'</span>, </span>
<span>    taxonomy <span class="op">=</span> <span class="st">'taxonomy.tsv'</span>, </span>
<span>    tree     <span class="op">=</span> <span class="st">'tree.nwk'</span> <span class="op">)</span></span>
<span>  </span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="mothur">mothur<a class="anchor" aria-label="anchor" href="#mothur"></a>
</h2>
<div class="section level3">
<h3 id="rbiom-mothur">rbiom ➡ mothur<a class="anchor" aria-label="anchor" href="#rbiom-mothur"></a>
</h3>
<p>In R, export mothur-compatible files.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># where to save the files</span></span>
<span><span class="va">project_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/export.html">write_mothur</a></span><span class="op">(</span>biom <span class="op">=</span> <span class="va">hmp50</span>, dir <span class="op">=</span> <span class="va">project_dir</span>, prefix <span class="op">=</span> <span class="st">'hmp50_'</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>This command creates files named <code>'hmp50_counts.tsv'</code>,
<code>'hmp50_metadata.tsv'</code>,
<code>'hmp50_taxonomy.tsv', 'hmp50_tree.nwk'</code>, and
<code>'hmp50_seqs.fna'</code>.</p>
</blockquote>
<p><br> At the mothur command prompt, import the BIOM data.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">mothur</span> <span class="op">&gt;</span> make.shared<span class="er">(</span><span class="va">count</span><span class="op">=</span>hmp50_counts.tsv, <span class="va">label</span><span class="op">=</span>asv<span class="kw">)</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="ex">mothur</span> <span class="op">&gt;</span> unifrac.unweighted<span class="er">(</span><span class="va">tree</span><span class="op">=</span>hmp50_tree.nwk, <span class="va">count</span><span class="op">=</span>hmp50_counts.tsv<span class="kw">)</span></span></code></pre></div>
<blockquote>
<p>The <code>make.shared()</code> command creates files named
<code>'hmp50_counts.asv.list'</code> and
<code>'hmp50_counts.asv.shared'</code>.</p>
</blockquote>
</div>
<div class="section level3">
<h3 id="mothur-rbiom">mothur ➡ rbiom<a class="anchor" aria-label="anchor" href="#mothur-rbiom"></a>
</h3>
<p>At the mothur command prompt, export a biom file.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">mothur</span> <span class="op">&gt;</span> make.biom<span class="er">(</span> <span class="dt">\</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>  <span class="va">shared</span><span class="op">=</span>hmp50_counts.asv.shared, <span class="dt">\</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="va">constaxonomy</span><span class="op">=</span>hmp50_taxonomy.tsv, <span class="dt">\</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="va">metadata</span><span class="op">=</span>hmp50_metadata.tsv, <span class="dt">\</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  <span class="va">output</span><span class="op">=</span>simple <span class="kw">)</span></span></code></pre></div>
<blockquote>
<p>The <code>make.biom()</code> command creates a file named
<code>'hmp50_counts.asv.asv.biom'</code>.</p>
</blockquote>
<p>In R, import the biom file with rbiom.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># project_dir = directory with relevant files</span></span>
<span></span>
<span><span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_dir.html" class="external-link">with_dir</a></span><span class="op">(</span><span class="va">project_dir</span>, <span class="op">{</span></span>
<span></span>
<span>  <span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span><span class="st">'hmp50_counts.asv.asv.biom'</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># Optionally add a tree and/or sequences</span></span>
<span>  <span class="va">biom</span><span class="op">$</span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="st">'hmp50_tree.nwk'</span></span>
<span>  <span class="va">biom</span><span class="op">$</span><span class="va">seqs</span> <span class="op">&lt;-</span> <span class="st">'hmp50_seqs.fna'</span></span>
<span>  </span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="bioconductor-r-packages">BioConductor R Packages<a class="anchor" aria-label="anchor" href="#bioconductor-r-packages"></a>
</h2>
<div class="section level3">
<h3 id="rbiom-phyloseq">rbiom ➡ phyloseq<a class="anchor" aria-label="anchor" href="#rbiom-phyloseq"></a>
</h3>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># An rbiom object</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="va"><a href="../reference/hmp50.html">hmp50</a></span></span>
<span></span>
<span><span class="co"># A phyloseq object</span></span>
<span><span class="va">physeq</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="fu"><a href="../reference/convert_to.html">convert_to_phyloseq</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="phyloseq-rbiom">phyloseq ➡ rbiom<a class="anchor" aria-label="anchor" href="#phyloseq-rbiom"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># A phyloseq object</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">enterotype</span>, package <span class="op">=</span> <span class="st">'phyloseq'</span><span class="op">)</span></span>
<span><span class="va">physeq</span> <span class="op">&lt;-</span> <span class="va">enterotype</span></span>
<span></span>
<span><span class="co"># An rbiom object</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span><span class="va">physeq</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rbiom-summarizedexperiment">rbiom ➡ SummarizedExperiment<a class="anchor" aria-label="anchor" href="#rbiom-summarizedexperiment"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># An rbiom object</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="va"><a href="../reference/hmp50.html">hmp50</a></span></span>
<span></span>
<span><span class="co"># A SummarizedExperiment object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="fu"><a href="../reference/convert_to.html">convert_to_SE</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="summarizedexperiment-rbiom">SummarizedExperiment ➡ rbiom<a class="anchor" aria-label="anchor" href="#summarizedexperiment-rbiom"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># `se` is a SummarizedExperiment object</span></span>
<span></span>
<span><span class="co"># Convert to rbiom object</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rbiom-treesummarizedexperiment">rbiom ➡ TreeSummarizedExperiment<a class="anchor" aria-label="anchor" href="#rbiom-treesummarizedexperiment"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># An rbiom object</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="va"><a href="../reference/hmp50.html">hmp50</a></span></span>
<span></span>
<span><span class="co"># A TreeSummarizedExperiment object</span></span>
<span><span class="va">tse</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="fu"><a href="../reference/convert_to.html">convert_to_TSE</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="treesummarizedexperiment-rbiom">TreeSummarizedExperiment ➡ rbiom<a class="anchor" aria-label="anchor" href="#treesummarizedexperiment-rbiom"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># `tse` is a TreeSummarizedExperiment object</span></span>
<span></span>
<span><span class="co"># Convert to rbiom object</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu">rbiom</span><span class="fu">::</span><span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span><span class="va">tse</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
