<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>rbiom • rbiom</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.10/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="rbiom">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>rbiom</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/vignettes/articles/rbiom.Rmd" class="external-link"><code>vignettes/articles/rbiom.Rmd</code></a></small>
      <div class="d-none name"><code>rbiom.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="create">Create<a class="anchor" aria-label="anchor" href="#create"></a>
</h2>
<p>The general purpose <code><a href="../reference/as_rbiom.html">as_rbiom()</a></code> function can convert most
data types into an <a href="../reference/rbiom_objects.html">rbiom
object</a> - see <a href="importing.html">importing</a> for details.
Here we’ll import a dataset from a BIOM file.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cmmr.github.io/rbiom/">rbiom</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>package <span class="op">=</span> <span class="st">"rbiom"</span>, <span class="st">"extdata"</span>, <span class="st">"hmp50.bz2"</span><span class="op">)</span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span></span>
<span></span>
<span><span class="va">biom</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #808080;">══ </span><span style="color: #BBBBBB;">Human Microbiome Project - 50 Sample Demo</span><span style="color: #808080;"> ═══════════════</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Oral, nasal, vaginal, and fecal samples from a diverse set</span></span>
<span><span class="co">#&gt; of healthy volunteers. Source: Human Microbiome Project</span></span>
<span><span class="co">#&gt; (<span style="color: #808080; font-style: italic;">&lt;https://hmpdacc.org&gt;</span>).</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      50 Samples: HMP01, HMP02, HMP03, ..., and HMP50</span></span>
<span><span class="co">#&gt;     490 OTUs:    Unc01yki, Unc53100, LtbAci52, ...</span></span>
<span><span class="co">#&gt;       7 Ranks:   .otu, Kingdom, Phylum, ..., and Genus</span></span>
<span><span class="co">#&gt;       5 Fields:  .sample, Age, BMI, Body Site, and Sex</span></span>
<span><span class="co">#&gt;         Tree:    <span style="font-style: italic;">&lt;present&gt;</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #808080;">── 182 - 22k reads/sample ──────────────────── 2023-09-22 ──</span></span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="inspect">Inspect<a class="anchor" aria-label="anchor" href="#inspect"></a>
</h2>
<p>The rbiom object has many helpful accessors.</p>
<table class="table">
<colgroup>
<col width="31%">
<col width="68%">
</colgroup>
<thead><tr class="header">
<th>Accessor</th>
<th>Content</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>$counts</code></td>
<td>Abundance of each OTU in each sample.</td>
</tr>
<tr class="even">
<td><code>$metadata</code></td>
<td>Sample mappings to metadata (treatment, patient, etc).</td>
</tr>
<tr class="odd">
<td><code>$taxonomy</code></td>
<td>OTU mappings to taxonomic ranks (genus, phylum, etc).</td>
</tr>
<tr class="even">
<td>
<code>$otus</code>, <code>$n_otus</code>
</td>
<td>OTU names.</td>
</tr>
<tr class="odd">
<td>
<code>$samples</code>, <code>$n_samples</code>
</td>
<td>Sample names.</td>
</tr>
<tr class="even">
<td>
<code>$fields</code>, <code>$n_fields</code>
</td>
<td>Metadata field names.</td>
</tr>
<tr class="odd">
<td>
<code>$ranks</code>, <code>$n_ranks</code>
</td>
<td>Taxonomic rank names.</td>
</tr>
<tr class="even">
<td>
<code>$tree</code>, <code>$sequences</code>
</td>
<td>Phylogenetic tree / sequences for the OTUs, or
<code>NULL</code>.</td>
</tr>
<tr class="odd">
<td>
<code>$id</code>, <code>$comment</code>
</td>
<td>Arbitrary strings for describing the dataset.</td>
</tr>
<tr class="even">
<td><code>$depth</code></td>
<td>Rarefaction depth, or <code>NULL</code> if unrarefied.</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biom</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;          HMP01 HMP02 HMP03 HMP04 HMP05 HMP06 HMP07 HMP08</span></span>
<span><span class="co">#&gt; Unc01yki     0     0     0     0     0     0     0     0</span></span>
<span><span class="co">#&gt; Unc53100  1083   543   301   223  2672   748  2436   321</span></span>
<span><span class="co">#&gt; LtbAci52     0     0     0     0     0     0     0     0</span></span>
<span><span class="co">#&gt; CnbTube3     0     0     0     0     0     0     0     0</span></span>
<span></span>
<span><span class="va">biom</span><span class="op">$</span><span class="va">fields</span></span>
<span><span class="co">#&gt; [1] ".sample"   "Age"       "BMI"       "Body Site" "Sex"</span></span>
<span></span>
<span><span class="co"># Use pull() to automatically setNames().</span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">biom</span>, <span class="st">'Age'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; HMP01 HMP02 HMP03 HMP04 HMP05 HMP06 </span></span>
<span><span class="co">#&gt;    22    24    28    25    27    32</span></span></code></pre></div>
<p><code><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull()</a></code>, <code><a href="../reference/sample_sums.html">sample_sums()</a></code>,
<code><a href="../reference/taxa_matrix.html">taxa_matrix()</a></code>, <code><a href="../reference/taxa_sums.html">taxa_means()</a></code>,
<code><a href="../reference/taxa_sums.html">taxa_sums()</a></code></p>
</div>
<div class="section level2">
<h2 id="clone">Clone<a class="anchor" aria-label="anchor" href="#clone"></a>
</h2>
<p>Rbiom objects are <u><b>passed by reference</b></u>. The common
<code>&lt;-</code> assignment operator creates a second reference to the
same object - it does not create a second object. To create a copy of an
rbiom object, use the object’s <code>$clone()</code> method.</p>
<div class="container">
<div class="row">
<div class="col">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="va">a</span><span class="op">$</span><span class="fu">clone</span><span class="op">(</span><span class="op">)</span>  <span class="co"># Correct</span></span>
<span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">n_samples</span></span>
<span><span class="co">#&gt; [1] 50</span></span>
<span><span class="va">b</span><span class="op">$</span><span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: <span style="color: #00BBBB;">ℹ</span> Dropping 44 samples from biom object since they are not in the new metadata:</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">"HMP07"</span>, <span style="color: #0000BB;">"HMP08"</span>, <span style="color: #0000BB;">"HMP09"</span>, <span style="color: #0000BB;">"HMP10"</span>, <span style="color: #0000BB;">"HMP11"</span>, <span style="color: #0000BB;">"HMP12"</span>, <span style="color: #0000BB;">"HMP13"</span>, <span style="color: #0000BB;">"HMP14"</span>,</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">"HMP15"</span>, <span style="color: #0000BB;">"HMP16"</span>, <span style="color: #0000BB;">"HMP17"</span>, <span style="color: #0000BB;">"HMP18"</span>, <span style="color: #0000BB;">"HMP19"</span>, <span style="color: #0000BB;">"HMP20"</span>, <span style="color: #0000BB;">"HMP21"</span>, <span style="color: #0000BB;">"HMP22"</span>,</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">"HMP23"</span>, <span style="color: #0000BB;">"HMP24"</span>, …, <span style="color: #0000BB;">"HMP49"</span>, and <span style="color: #0000BB;">"HMP50"</span>.</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">n_samples</span></span>
<span><span class="co">#&gt; [1] 50</span></span></code></pre></div>
</div>
<div class="col">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_rbiom.html">as_rbiom</a></span><span class="op">(</span><span class="va">file</span><span class="op">)</span></span>
<span><span class="va">b</span> <span class="op">&lt;-</span> <span class="va">a</span>  <span class="co"># Wrong</span></span>
<span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">n_samples</span></span>
<span><span class="co">#&gt; [1] 50</span></span>
<span><span class="va">b</span><span class="op">$</span><span class="va">metadata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: <span style="color: #00BBBB;">ℹ</span> Dropping 44 samples from biom object since they are not in the new metadata:</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">"HMP07"</span>, <span style="color: #0000BB;">"HMP08"</span>, <span style="color: #0000BB;">"HMP09"</span>, <span style="color: #0000BB;">"HMP10"</span>, <span style="color: #0000BB;">"HMP11"</span>, <span style="color: #0000BB;">"HMP12"</span>, <span style="color: #0000BB;">"HMP13"</span>, <span style="color: #0000BB;">"HMP14"</span>,</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">"HMP15"</span>, <span style="color: #0000BB;">"HMP16"</span>, <span style="color: #0000BB;">"HMP17"</span>, <span style="color: #0000BB;">"HMP18"</span>, <span style="color: #0000BB;">"HMP19"</span>, <span style="color: #0000BB;">"HMP20"</span>, <span style="color: #0000BB;">"HMP21"</span>, <span style="color: #0000BB;">"HMP22"</span>,</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">"HMP23"</span>, <span style="color: #0000BB;">"HMP24"</span>, …, <span style="color: #0000BB;">"HMP49"</span>, and <span style="color: #0000BB;">"HMP50"</span>.</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">n_samples</span></span>
<span><span class="co">#&gt; [1] 6</span></span></code></pre></div>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="modify">Modify<a class="anchor" aria-label="anchor" href="#modify"></a>
</h2>
<p>There are seven components of an rbiom object which you can modify
directly. Assigning new values to these components will trigger
validation checks and inter-component OTU/sample synchronization. See <a href="../reference/rbiom_objects.html">Working with rbiom Objects</a>
for additional details.</p>
<table class="table">
<colgroup>
<col width="23%">
<col width="76%">
</colgroup>
<thead><tr class="header">
<th>Component</th>
<th>What can be assigned.</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>$counts</code></td>
<td>matrix of abundances; OTUs (rows) by samples (columns)</td>
</tr>
<tr class="even">
<td><code>$metadata</code></td>
<td>data.frame with <code>'.sample'</code> as the first column</td>
</tr>
<tr class="odd">
<td><code>$taxonomy</code></td>
<td>data.frame with <code>'.otu'</code> as the first column</td>
</tr>
<tr class="even">
<td><code>$tree</code></td>
<td>phylo object with the phylogenetic tree for OTUs</td>
</tr>
<tr class="odd">
<td><code>$sequences</code></td>
<td>character vector of reference sequences for OTUs</td>
</tr>
<tr class="even">
<td><code>$id</code></td>
<td>string with a title for the dataset</td>
</tr>
<tr class="odd">
<td><code>$comment</code></td>
<td>string with additional dataset information</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="rarefy-counts">Rarefy Counts<a class="anchor" aria-label="anchor" href="#rarefy-counts"></a>
</h3>
<p>A common way to normalize microbiome count data is to rarefy it. This
process drops samples with too few observations, and randomly removes
observations from the remaining samples, so that all samples have the
same “rarefaction depth”.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sample_sums.html">sample_sums</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; HMP01 HMP02 HMP03 HMP04 HMP05 HMP06 </span></span>
<span><span class="co">#&gt;  1660  1371  1353  1895  3939  4150</span></span>
<span></span>
<span><span class="va">biom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/sample_sums.html">sample_sums</a></span><span class="op">(</span><span class="va">biom</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; HMP01 HMP02 HMP03 HMP04 HMP05 HMP06 </span></span>
<span><span class="co">#&gt;  1183  1183  1183  1183  1183  1183</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-metadata">Add Metadata<a class="anchor" aria-label="anchor" href="#add-metadata"></a>
</h3>
<p>Additional sample metadata columns can be added to
<code>biom$metadata</code> (a tibble data.frame). The first column,
<code>'.sample'</code>, is used by rbiom to link sample metadata to
samples in the abundance table.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biom</span><span class="op">$</span><span class="va">metadata</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'A'</span>, <span class="st">'B'</span><span class="op">)</span>, <span class="va">biom</span><span class="op">$</span><span class="va">n_samples</span>, <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">biom</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Obese <span class="op">=</span> <span class="va">BMI</span> <span class="op">&gt;=</span> <span class="fl">30</span>, Sex <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">biom</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="st">'Years Old'</span> <span class="op">=</span> <span class="st">"Age"</span><span class="op">)</span></span>
<span><span class="va">biom</span><span class="op">$</span><span class="va">metadata</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 49 × 6</span></span></span>
<span><span class="co">#&gt;   .sample `Years Old`   BMI `Body Site`   group Obese</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> HMP01            22    20 Buccal mucosa A     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> HMP02            24    23 Buccal mucosa A     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> HMP03            28    26 Saliva        B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> HMP04            25    23 Saliva        A     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> HMP05            27    24 Buccal mucosa A     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 44 more rows</span></span></span></code></pre></div>
<p><code><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename()</a></code></p>
</div>
<div class="section level3">
<h3 id="subset-samples">Subset Samples<a class="anchor" aria-label="anchor" href="#subset-samples"></a>
</h3>
<p>Removing samples from the metadata will remove those samples from the
entire rbiom object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">biom</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">`Body Site`</span> <span class="op">==</span> <span class="st">"Anterior nares"</span><span class="op">)</span></span>
<span><span class="va">biom</span><span class="op">$</span><span class="va">metadata</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 10 × 6</span></span></span>
<span><span class="co">#&gt;   .sample `Years Old`   BMI `Body Site`    group Obese</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">*</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> HMP10            22    20 Anterior nares B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> HMP15            25    21 Anterior nares A     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> HMP16            24    19 Anterior nares B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> HMP25            33    32 Anterior nares A     TRUE </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> HMP31            31    20 Anterior nares A     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 5 more rows</span></span></span>
<span><span class="va">biom</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #808080;">══ </span><span style="color: #BBBBBB;">Human Microbiome Project - 50 Sample Demo</span><span style="color: #808080;"> ═══════════════</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Oral, nasal, vaginal, and fecal samples from a diverse set</span></span>
<span><span class="co">#&gt; of healthy volunteers. Source: Human Microbiome Project</span></span>
<span><span class="co">#&gt; (<span style="color: #808080; font-style: italic;">&lt;https://hmpdacc.org&gt;</span>).</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;      10 Samples: HMP10, HMP15, HMP16, ..., and HMP48</span></span>
<span><span class="co">#&gt;     135 OTUs:    Unc01yki, Unc53100, LtbAci52, ...</span></span>
<span><span class="co">#&gt;       7 Ranks:   .otu, Kingdom, Phylum, ..., and Genus</span></span>
<span><span class="co">#&gt;       6 Fields:  .sample, Years Old, BMI, ..., and Obese</span></span>
<span><span class="co">#&gt;         Tree:    <span style="font-style: italic;">&lt;present&gt;</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #808080;">── 1183 reads/sample ───────────────────────── 2023-09-22 ──</span></span></span>
<span><span class="co">#&gt; </span></span></code></pre></div>
<p><code><a href="../reference/subset.html">subset()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice()</a></code></p>
</div>
</div>
<div class="section level2">
<h2 id="compute">Compute<a class="anchor" aria-label="anchor" href="#compute"></a>
</h2>
<p>Functions that end in <code>_table</code> or <code>_matrix</code>
return calculation results for use outside of rbiom. The
<code>_table</code> suffix indicates the returned object will be a
tibble data.frame with one computed value per row. Alternatively,
<code>_matrix</code> will return a base R matrix.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/taxa_matrix.html">taxa_table</a></span><span class="op">(</span><span class="va">biom</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 60 × 9</span></span></span>
<span><span class="co">#&gt;   .rank  .sample .taxa      .abundance `Years Old`   BMI `Body Site` group Obese</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Phylum HMP10   Actinobac…        724          22    20 Anterior n… B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Phylum HMP10   Firmicutes        446          22    20 Anterior n… B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Phylum HMP10   Proteobac…          3          22    20 Anterior n… B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Phylum HMP10   Bacteroid…         10          22    20 Anterior n… B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Phylum HMP10   Cyanobact…          0          22    20 Anterior n… B     FALSE</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 55 more rows</span></span></span>
<span></span>
<span><span class="fu"><a href="../reference/taxa_matrix.html">taxa_matrix</a></span><span class="op">(</span><span class="va">biom</span>, rank <span class="op">=</span> <span class="st">"Phylum"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></span>
<span><span class="co">#&gt;                     HMP10 HMP15 HMP16 HMP25 HMP31 HMP32 HMP34 HMP40</span></span>
<span><span class="co">#&gt; Actinobacteria        724   847   654   448   668   622  1023   876</span></span>
<span><span class="co">#&gt; Bacteroidetes          10     4     0     1    79    12    12     2</span></span>
<span><span class="co">#&gt; Cyanobacteria           0     0     3     0     0    25     0     0</span></span>
<span><span class="co">#&gt; Deinococcus Thermus     0     0     0     0     0     0     0     3</span></span></code></pre></div>
<p>Functions for taxa abundance (like the above
<code><a href="../reference/taxa_matrix.html">taxa_table()</a></code>) are prefixed by <code>taxa_</code>.
Similarly, <code>adiv_</code> is used for alpha diversity and
<code>bdiv_</code> for beta diversity.</p>
<p><code><a href="../reference/adiv_table.html">adiv_matrix()</a></code>, <code><a href="../reference/adiv_table.html">adiv_table()</a></code>,
<code><a href="../reference/bdiv_table.html">bdiv_table()</a></code>, <code><a href="../reference/bdiv_ord_table.html">bdiv_ord_table()</a></code>,
<code><a href="../reference/taxa_matrix.html">taxa_table()</a></code>, <code><a href="../reference/taxa_matrix.html">taxa_matrix()</a></code></p>
</div>
<div class="section level2">
<h2 id="visualize">Visualize<a class="anchor" aria-label="anchor" href="#visualize"></a>
</h2>
<p>The plotting functions in rbiom make it easy to produce informative
visualizations about alpha diversity, beta diversity, and taxa
abundance, and explore associations between those metrics and sample
metadata.</p>
<p>See the <a href="plots.html">Plot Types</a> article for an overview
of all the different plot options, and <a href="aes.html">Mapping
Metadata to Aesthetics</a> for guidance on using colors, shapes, and
patterns to represent metadata values.</p>
<p>For example, to display an ordination colored by body site:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot</a></span><span class="op">(</span>biom <span class="op">=</span> <span class="va">hmp50</span>, stat.by <span class="op">=</span> <span class="st">"Body Site"</span><span class="op">)</span></span></code></pre></div>
<p><img src="rbiom_files/figure-html/unnamed-chunk-10-1.png" class="r-plt" alt="" width="768"></p>
<p><code><a href="../reference/adiv_boxplot.html">adiv_boxplot()</a></code>, <code><a href="../reference/adiv_corrplot.html">adiv_corrplot()</a></code>,
<code><a href="../reference/bdiv_boxplot.html">bdiv_boxplot()</a></code>, <code><a href="../reference/bdiv_corrplot.html">bdiv_corrplot()</a></code>,
<code><a href="../reference/bdiv_heatmap.html">bdiv_heatmap()</a></code>, <code><a href="../reference/bdiv_ord_plot.html">bdiv_ord_plot()</a></code>,
<code><a href="../reference/rare_stacked.html">rare_stacked()</a></code>, <code><a href="../reference/rare_corrplot.html">rare_corrplot()</a></code>,
<code><a href="../reference/rare_multiplot.html">rare_multiplot()</a></code>, <code><a href="../reference/taxa_stacked.html">taxa_stacked()</a></code>,
<code><a href="../reference/taxa_boxplot.html">taxa_boxplot()</a></code>, <code><a href="../reference/taxa_corrplot.html">taxa_corrplot()</a></code>,
<code><a href="../reference/taxa_heatmap.html">taxa_heatmap()</a></code></p>
</div>
<div class="section level2">
<h2 id="analyze">Analyze<a class="anchor" aria-label="anchor" href="#analyze"></a>
</h2>
<p>Visualizations are an excellent way to observe differences between
sample groups. When <code>stat.by</code> is set, boxplots, corrplots,
and ord_plots will include the following:</p>
<ul>
<li>p-values and brief methodology on the plot itself.</li>
<li>
<code>$stats</code> attribute with a detailed statistics table.</li>
<li>
<code>$stats$code</code> attribute with the R code used to generate
the table.</li>
</ul>
<p>The <a href="statistics.html">rbiom statistics article</a> reviews
this topic in greater detail.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/adiv_boxplot.html">adiv_boxplot</a></span><span class="op">(</span><span class="va">hmp50</span>, stat.by <span class="op">=</span> <span class="st">"Body Site"</span>, facet.by <span class="op">=</span> <span class="st">"Sex"</span><span class="op">)</span></span>
<span><span class="va">p</span></span></code></pre></div>
<p><img src="rbiom_files/figure-html/unnamed-chunk-11-1.png" class="r-plt" alt="" width="768"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    kruskal.test(.diversity ~ `Body Site`)</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 2 × 7</span></span></span>
<span><span class="co">#&gt;   Sex    .stat .h1      .p.val   .adj.p    .n   .df</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Female  24.1 &gt; 0   0.000<span style="text-decoration: underline;">076</span>1 0.000<span style="text-decoration: underline;">152</span>    30     4</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Male    13.9 &gt; 0   0.003<span style="text-decoration: underline;">02</span>   0.003<span style="text-decoration: underline;">02</span>     20     3</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">stats</span><span class="op">$</span><span class="va">code</span></span>
<span><span class="co">#&gt; data <span style="color: #00BB00;">&lt;-</span> <span style="color: #00BBBB;">adiv_table</span><span style="color: #BBBB00;">(</span>biom, <span style="color: #BBBB00;">"Shannon"</span>, <span style="color: #00BBBB;">c</span><span style="color: #BB00BB;">(</span><span style="color: #BBBB00;">"Body Site"</span>, <span style="color: #BBBB00;">"Sex"</span><span style="color: #BB00BB;">)</span><span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; data <span style="color: #00BB00;">%&lt;&gt;%</span> dplyr::<span style="color: #00BBBB;">rename</span><span style="color: #BBBB00;">(</span></span></span>
<span><span class="co">#&gt;   .resp    = <span style="color: #BBBB00;">".diversity"</span>, </span></span>
<span><span class="co">#&gt;   .stat.by = <span style="color: #BBBB00;">"Body Site"</span> <span style="color: #BBBB00;">)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; stats <span style="color: #00BB00;">&lt;-</span> plyr::<span style="color: #00BBBB;">ddply</span><span style="color: #BBBB00;">(</span>data, <span style="color: #00BBBB;">.</span><span style="color: #BB00BB;">(</span>Sex<span style="color: #BB00BB;">)</span>, <span style="color: #BB0000;">function</span> <span style="color: #BB00BB;">(</span>data<span style="color: #BB00BB;">)</span> <span style="color: #BB00BB;">{</span></span></span>
<span><span class="co">#&gt;   <span style="color: #00BBBB;">tryCatch(</span>error = <span style="color: #BB0000;">function</span> <span style="color: #BBBB00;">(</span>e<span style="color: #BBBB00;">)</span> <span style="color: #00BBBB;">data.frame</span><span style="color: #BBBB00;">()[</span><span style="color: #BB00BB;">1</span>,<span style="color: #BBBB00;">]</span>, <span style="color: #00BBBB;">suppressWarnings</span><span style="color: #BBBB00;">(</span><span style="color: #BB00BB;">{</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     data <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;       stats::<span style="color: #00BBBB;">kruskal.test(</span>.resp <span style="color: #00BB00;">~</span> .stat.by, .<span style="color: #00BBBB;">)</span> <span style="color: #00BB00;">%&gt;%</span></span></span>
<span><span class="co">#&gt;       <span style="color: #00BBBB;">with(tibble</span><span style="color: #BBBB00;">(</span></span></span>
<span><span class="co">#&gt;         .stat  = statistic, </span></span>
<span><span class="co">#&gt;         .h1    = <span style="color: #00BBBB;">factor</span><span style="color: #BB00BB;">(</span><span style="color: #BBBB00;">'&gt; 0'</span><span style="color: #BB00BB;">)</span>, </span></span>
<span><span class="co">#&gt;         .p.val = p.value, </span></span>
<span><span class="co">#&gt;         .n     = <span style="color: #00BBBB;">nrow</span><span style="color: #BB00BB;">(</span>data<span style="color: #BB00BB;">)</span>, </span></span>
<span><span class="co">#&gt;         .df    = parameter <span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   <span style="color: #BB00BB;">}</span><span style="color: #BBBB00;">)</span><span style="color: #00BBBB;">)</span></span></span>
<span><span class="co">#&gt; <span style="color: #BB00BB;">}</span><span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   tibble::<span style="color: #00BBBB;">as_tibble</span><span style="color: #BBBB00;">()</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   dplyr::<span style="color: #00BBBB;">mutate</span><span style="color: #BBBB00;">(</span>.adj.p = <span style="color: #00BBBB;">p.adjust</span><span style="color: #BB00BB;">(</span>.p.val, <span style="color: #BBBB00;">'fdr'</span><span style="color: #BB00BB;">)</span>, .after = .p.val<span style="color: #BBBB00;">)</span> <span style="color: #00BB00;">%&gt;%</span> </span></span>
<span><span class="co">#&gt;   dplyr::<span style="color: #00BBBB;">arrange</span><span style="color: #BBBB00;">(</span>.p.val<span style="color: #BBBB00;">)</span></span></span></code></pre></div>
<p><code><a href="../reference/adiv_stats.html">adiv_stats()</a></code>, <code><a href="../reference/bdiv_stats.html">bdiv_stats()</a></code>,
<code><a href="../reference/distmat_stats.html">distmat_stats()</a></code>, <code><a href="../reference/taxa_stats.html">taxa_stats()</a></code></p>
</div>
<div class="section level2">
<h2 id="share">Share<a class="anchor" aria-label="anchor" href="#share"></a>
</h2>
<p>Plots are essentially ggplot objects, and
<code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggplot2::ggsave()</a></code> can save them as PNG, PDF, SVG, and many
other image file types.</p>
<p>To save your rbiom object to a BIOM file, use
<code><a href="../reference/write_biom.html">write_biom()</a></code>. When sharing a dataset with someone who is
unfamiliar with accessing BIOM files, <code><a href="../reference/write_biom.html">write_xlsx()</a></code> is also
an option.</p>
<p>A few additional resources to know about:</p>
<ul>
<li>
<a href="https://rmarkdown.rstudio.com/" class="external-link">R Markdown</a> -
reproducible analyses/reports.</li>
<li>
<a href="https://pkgs.rstudio.com/flexdashboard/" class="external-link">flexdashboard</a>
- interactive dashboards.</li>
<li>
<a href="http://atima.research.bcm.edu" class="external-link">atima</a> - rbiom in website
form.</li>
</ul>
<p><code><a href="https://ggplot2.tidyverse.org/reference/ggsave.html" class="external-link">ggplot2::ggsave()</a></code>, <code><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">patchwork::wrap_plots()</a></code>,
<code><a href="../reference/write_biom.html">write_biom()</a></code>, <code><a href="../reference/write_biom.html">write_xlsx()</a></code></p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
