<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Regression • rbiom</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/Noto_Sans-0.4.9/font.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Regression">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rbiom</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.13.9005</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rbiom.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><h6 class="dropdown-header" data-toc-skip>Analysis Basics</h6></li>
    <li><a class="dropdown-item" href="../articles/rbiom.html">Getting Started</a></li>
    <li><a class="dropdown-item" href="../articles/importing.html">Importing Your Data</a></li>
    <li><a class="dropdown-item" href="../reference/rbiom_objects.html">Working with rbiom Objects</a></li>
    <li><a class="dropdown-item" href="../articles/statistics.html">Statistics</a></li>
    <li><a class="dropdown-item" href="../articles/regression.html">Regression</a></li>
    <li><a class="dropdown-item" href="../reference/speed.html">Speed Ups</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Plotting</h6></li>
    <li><a class="dropdown-item" href="../articles/plots.html">Plot Types</a></li>
    <li><a class="dropdown-item" href="../articles/aes.html">Mapping Metadata to Aesthetics</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/convert.html">QIIME 2, mothur, and BioConductor</a></li>
    <li><a class="dropdown-item" href="../articles/unifrac.html">UniFrac Equations</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/cmmr/rbiom/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Regression</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cmmr/rbiom/blob/master/vignettes/articles/regression.Rmd" class="external-link"><code>vignettes/articles/regression.Rmd</code></a></small>
      <div class="d-none name"><code>regression.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Regression functions are provided for examining the association
between two continuous variables.</p>
<p>Examples of continuous variables are alpha diversity metrics, taxa
abundances, and sample metadata such as age or BMI. These all have a
numeric range in which a data point can take on any value in that
range.</p>
</div>
<div class="section level2">
<h2 id="children-heights">Children Heights<a class="anchor" aria-label="anchor" href="#children-heights"></a>
</h2>
<p>To start with a simple example, consider the <code>children</code>
dataset from the R package <code>npregfast</code>. It records age, sex,
and height for 2500 children aged 5 through 19 years.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span><span class="op">)</span></span>
<span><span class="co">#&gt;      sex height   age</span></span>
<span><span class="co">#&gt; 1   male 150.77 13.25</span></span>
<span><span class="co">#&gt; 2 female 170.59 14.17</span></span>
<span><span class="co">#&gt; 3 female 167.31 15.17</span></span>
<span><span class="co">#&gt; 4 female 165.72 16.58</span></span>
<span><span class="co">#&gt; 5 female 171.67 15.17</span></span>
<span><span class="co">#&gt; 6 female 143.74 12.67</span></span></code></pre></div>
<div class="section level3">
<h3 id="height-vs-age">Height vs Age<a class="anchor" aria-label="anchor" href="#height-vs-age"></a>
</h3>
<p>First, let’s plot ‘age’ vs ‘height’, and overlay a best-fit linear
trendline.</p>
<p>With this plot, we’re testing the hypothesis “as you get older, your
height changes”.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tp'</span>,                 <span class="co"># show trendline and points</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emtrends'</span>,           <span class="co"># run stats on the slope</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'lm'</span>,                 <span class="co"># straight trendline</span></span>
<span>  pt.size  <span class="op">=</span> <span class="fl">0.2</span>,                  <span class="co"># make points smaller</span></span>
<span>  pt.alpha <span class="op">=</span> <span class="fl">0.2</span> <span class="op">)</span>                 <span class="co"># and semi-transparent</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-3-1.png" width="768"></p>
<p>As you’d expect, the trendline’s slope is positive (p-value = 0).
Therefore it’s safe to say that height changes with age.</p>
</div>
<div class="section level3">
<h3 id="grouped-by-sex">Grouped by Sex<a class="anchor" aria-label="anchor" href="#grouped-by-sex"></a>
</h3>
<p>Is the rate of growth influenced by sex? We can add
<code>stat.by = "sex"</code> to draw separate lines for males and
females.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups   &lt;====</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'lm'</span>,                 <span class="co"># straight trendline</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tp'</span>,                 <span class="co"># show trendline and points</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emtrends'</span>,           <span class="co"># run stats on the slopes</span></span>
<span>  pt.size  <span class="op">=</span> <span class="fl">0.2</span> <span class="op">)</span>                 <span class="co"># make points smaller</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-4-1.png" width="768"></p>
<p>Here, a p-value of 3.9e-51 indicates that males and females do have
different growth rates.</p>
<p>If you look at the captions of the last two plots, you’ll notice that
the p-value test automatically changes.</p>
<ul>
<li>When <code>stat.by = NULL</code>, <code>test = "emtrends"</code>
will test if the <strong>slope is zero</strong>.</li>
<li>When <code>stat.by</code> is not <code>NULL</code>,
<code>test = "emtrends"</code> will test if the trendline <strong>slopes
are equal</strong>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="smoothed-fit">Smoothed Fit<a class="anchor" aria-label="anchor" href="#smoothed-fit"></a>
</h3>
<p>Linear trendlines probably aren’t the best way to model growth rates.
Setting <code>fit = "gam"</code> will use a generalized additive model
which fits several sub-ranges of age with independent splines.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'gam'</span>,                <span class="co"># smoothed trendline   &lt;====</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tp'</span>,                 <span class="co"># show trendline and points</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emtrends'</span>,           <span class="co"># run stats on the slopes</span></span>
<span>  pt.size  <span class="op">=</span> <span class="fl">0.2</span> <span class="op">)</span>                 <span class="co"># make points smaller</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-5-1.png" width="768"></p>
<p>This gives us a much better idea of the moving average over time.</p>
<p>We still have <code>test = "emtrends"</code>, so the p-value is
reported where the difference in slope is most significant - in this
case, at age = 5.9 years.</p>
</div>
<div class="section level3">
<h3 id="difference-in-means">Difference in Means<a class="anchor" aria-label="anchor" href="#difference-in-means"></a>
</h3>
<p>Let’s set <code>test = "emmeans"</code> instead, and show the 95%
confidence interval instead of all the data points.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'gam'</span>,                <span class="co"># smoothed trendline</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tc'</span>,                 <span class="co"># show trendline with conf. int.</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emmeans'</span> <span class="op">)</span>           <span class="co"># run stats on the means   &lt;====</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-6-1.png" width="768"></p>
<p>Differences in height are most significant at age = 16.8 (where
p-value = 8.9e-40).</p>
</div>
</div>
<div class="section level2">
<h2 id="statistics-table">Statistics Table<a class="anchor" aria-label="anchor" href="#statistics-table"></a>
</h2>
<p>The complete statistical output is attached to the returned plot as
<code>$stats</code>.</p>
<p>You can also directly generate this table with the
<code><a href="../reference/stats_table.html">stats_table()</a></code> function.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">st</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  regr     <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  resp     <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'gam'</span>,                <span class="co"># generalized additive model</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emmeans'</span> <span class="op">)</span>           <span class="co"># run stats on the means</span></span>
<span></span>
<span><span class="co"># Print the table transposed, since it's one row by 14 columns.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">st</span><span class="op">)</span></span>
<span><span class="co">#&gt;              [,1]           </span></span>
<span><span class="co">#&gt; age          "16.8"         </span></span>
<span><span class="co">#&gt; sex          "female - male"</span></span>
<span><span class="co">#&gt; .mean.diff   "-7.730025"    </span></span>
<span><span class="co">#&gt; .h1          "!= 0"         </span></span>
<span><span class="co">#&gt; .p.val       "8.875599e-40" </span></span>
<span><span class="co">#&gt; .adj.p       "8.875599e-40" </span></span>
<span><span class="co">#&gt; .effect.size "-1.205401"    </span></span>
<span><span class="co">#&gt; .se          "0.5753545"    </span></span>
<span><span class="co">#&gt; .n           "2500"         </span></span>
<span><span class="co">#&gt; .df          "2485"         </span></span>
<span><span class="co">#&gt; .t.ratio     "-13.43524"    </span></span>
<span><span class="co">#&gt; .adj.r       "0.8313946"    </span></span>
<span><span class="co">#&gt; .aic         "16403.78"     </span></span>
<span><span class="co">#&gt; .bic         "16497.75"     </span></span>
<span><span class="co">#&gt; .loglik      "-8185.756"</span></span></code></pre></div>
<p>Depending on the arguments given, the <em>estimate</em> term will be
one of:</p>
<table style="width:100%;" class="table">
<colgroup>
<col width="16%">
<col width="36%">
<col width="31%">
<col width="15%">
</colgroup>
<thead><tr class="header">
<th><strong>Field</strong></th>
<th><strong>Description</strong></th>
<th><code>test</code></th>
<th><code>stat.by</code></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>.mean</td>
<td>Estimated marginal mean.</td>
<td><code><a href="https://rvlenth.github.io/emmeans/reference/emmeans.html" class="external-link">emmeans::emmeans()</a></code></td>
<td><code>NULL</code></td>
</tr>
<tr class="even">
<td>.mean.diff</td>
<td>Difference in means.</td>
<td><code><a href="https://rvlenth.github.io/emmeans/reference/emmeans.html" class="external-link">emmeans::emmeans()</a></code></td>
<td>not <code>NULL</code>
</td>
</tr>
<tr class="odd">
<td>.slope</td>
<td>Trendline slope.</td>
<td><code><a href="https://rvlenth.github.io/emmeans/reference/emtrends.html" class="external-link">emmeans::emtrends()</a></code></td>
<td><code>NULL</code></td>
</tr>
<tr class="even">
<td>.slope.diff</td>
<td>Difference in slopes.</td>
<td><code><a href="https://rvlenth.github.io/emmeans/reference/emtrends.html" class="external-link">emmeans::emtrends()</a></code></td>
<td>not <code>NULL</code>
</td>
</tr>
</tbody>
</table>
<p>Other fields in this table include:</p>
<table class="table">
<thead><tr class="header">
<th><strong>Field</strong></th>
<th><strong>Description</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>.h1</td>
<td>Alternate hypothesis.</td>
</tr>
<tr class="even">
<td>.p.val</td>
<td>Probability that null hypothesis is correct.</td>
</tr>
<tr class="odd">
<td>.adj.p</td>
<td>
<code>.p.val</code> after adjusting for multiple comparisons.</td>
</tr>
<tr class="even">
<td>.effect.size</td>
<td>Effect size. See <code><a href="https://rvlenth.github.io/emmeans/reference/eff_size.html" class="external-link">emmeans::eff_size()</a></code>.</td>
</tr>
<tr class="odd">
<td>.lower</td>
<td>Confidence interval lower bound.</td>
</tr>
<tr class="even">
<td>.upper</td>
<td>Confidence interval upper bound.</td>
</tr>
<tr class="odd">
<td>.se</td>
<td>Standard error.</td>
</tr>
<tr class="even">
<td>.n</td>
<td>Number of samples.</td>
</tr>
<tr class="odd">
<td>.df</td>
<td>Degrees of freedom.</td>
</tr>
<tr class="even">
<td>.t.ratio</td>
<td>
<em>(.mean, .mean.diff, .slope, or .slope.diff)</em> /
<code>.se</code>
</td>
</tr>
<tr class="odd">
<td>.r.sqr</td>
<td>Percent of variation explained by the model.</td>
</tr>
<tr class="even">
<td>.adj.r</td>
<td>
<code>.r.sqr</code>, taking degrees of freedom into account.</td>
</tr>
<tr class="odd">
<td>.aic</td>
<td>Akaike Information Criterion (predictive models).</td>
</tr>
<tr class="even">
<td>.bic</td>
<td>Bayesian Information Criterion (descriptive models).</td>
</tr>
<tr class="odd">
<td>.loglik</td>
<td>Log-likelihood goodness-of-fit score.</td>
</tr>
<tr class="even">
<td>.fit.p</td>
<td>P-value for observing this fit by chance.</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="marginal-means">Marginal Means<a class="anchor" aria-label="anchor" href="#marginal-means"></a>
</h3>
<p>Estimated marginal means (EMMs; also known as least-squares means)
are means extracted from a statistical model. This allows EMMs to take
into account more complex associations and produce confidence intervals
in addition to estimates of the mean. This approach is employed by
rbiom’s <code>'emmeans'</code> and <code>'emtrends'</code> tests, using
the <code><a href="https://rvlenth.github.io/emmeans/reference/emmeans.html" class="external-link">emmeans::emmeans</a></code> package.</p>
<p>Additional information on EMMs is available at:</p>
<ul>
<li><a href="https://www.andrewheiss.com/blog/2022/05/20/marginalia/#marginal-effects-at-the-mean-the-default-in-emmeans" class="external-link">Marginalia:
A Guide to Figuring Out What the Heck Marginal Effects Are</a></li>
<li><a href="https://stats.oarc.ucla.edu/r/seminars/interactions-r" class="external-link">Decomposing,
Probing, and Plotting Interactions in R</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="goodness-of-fit">Goodness of Fit<a class="anchor" aria-label="anchor" href="#goodness-of-fit"></a>
</h3>
<p>The statistics table has several fields for assessing how well the
model fits the data.</p>
<ul>
<li>
<code>.r.sqr</code> - Coefficient of Determination
(<i>R</i><sup>2</sup>); range: 0-1, lower is better. Percent of
variation explained by the model.</li>
<li>
<code>.adj.r</code> - <i>R</i><sup>2</sup>, taking degrees of
freedom into account.</li>
<li>
<code>.aic</code> - Akaike Information Criterion; lower is better.
Preferred when using model for prediction.</li>
<li>
<code>.bic</code> - Bayesian Information Criterion; lower is better.
Preferred when using model for interpretation.</li>
<li>
<code>.loglik</code> - Log-Likelihood (negative values); higher
values (closer to zero) are better.</li>
<li>
<code>.fit.p</code> - P-value for observing this fit by chance;
range: 0-1, lower is better.</li>
</ul>
<p>Not all of these values can be computed for every model.</p>
<p>As an example, lets pull the AIC value from the statistics table
generated in the last section.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">st</span><span class="op">$</span><span class="va">.aic</span></span>
<span><span class="co">#&gt; [1] 16403.78</span></span></code></pre></div>
<p>These values can be used to decide which <code>fit</code> argument to
use (<code>"lm"</code>, <code>"log"</code>, or <code>"gam"</code>).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/ldply.html" class="external-link">ldply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>lm<span class="op">=</span><span class="st">"lm"</span>, log<span class="op">=</span><span class="st">"log"</span>, gam<span class="op">=</span><span class="st">"gam"</span><span class="op">)</span>, .id <span class="op">=</span> <span class="st">"fit"</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">fit</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>, <span class="st">"age"</span>, <span class="st">"height"</span>, <span class="st">"sex"</span>, fit <span class="op">=</span> <span class="va">fit</span><span class="op">)</span> <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">fit</span>, <span class="va">.r.sqr</span><span class="op">:</span><span class="va">.fit.p</span><span class="op">)</span></span>
<span><span class="co">#&gt;   fit    .r.sqr    .adj.r     .aic     .bic   .loglik .fit.p</span></span>
<span><span class="co">#&gt; 1  lm 0.7896923 0.7894395 16947.72 16976.84 -8468.858      0</span></span>
<span><span class="co">#&gt; 2 log 0.8031003 0.8028636 16783.02 16812.14 -8386.512      0</span></span>
<span><span class="co">#&gt; 3 gam        NA 0.8313946 16403.78 16497.75 -8185.756     NA</span></span></code></pre></div>
<p>Based on the above table, <code>fit = "gam"</code> has the lowest
AIC, lowest BIC, and highest Log-Likelihood, and is therefore the best
model of the three to use for this set of data and variables.</p>
</div>
<div class="section level3">
<h3 id="alternative-hypothesis">Alternative Hypothesis<a class="anchor" aria-label="anchor" href="#alternative-hypothesis"></a>
</h3>
<p>The three columns <code>.mean.diff</code>, <code>.h1</code>, and
<code>.p.val</code> together show the hypothesis and its outcome. Above,
we’re asking if <code>.mean.diff</code> is non-zero. Since
<code>.p.val</code> is less than 0.05 we can say that it is.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">st</span>, <span class="va">.mean.diff</span><span class="op">:</span><span class="va">.p.val</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    gam(height ~ s(age, by = sex, bs = "cs") + sex, method = "REML")</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 3</span></span></span>
<span><span class="co">#&gt;   .mean.diff .h1     .p.val</span></span>
<span><span class="co">#&gt;        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>      -<span style="color: #BB0000;">7.73</span> != 0  8.88<span style="color: #949494;">e</span><span style="color: #BB0000;">-40</span></span></span></code></pre></div>
<p>For a more complete review of alternative hypotheses, see the
<a href="statistics.html#hypothesis">rbiom statistics article</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="checking-residuals">Checking Residuals<a class="anchor" aria-label="anchor" href="#checking-residuals"></a>
</h2>
<p>One of the best ways to assess the quality of your model is to
interrogate the residuals. A “residual” is the vertical distance from a
data point to the trendline; the difference between observed and
predicted (fitted) values.</p>
<p>High quality models will have randomly distributed residuals. Three
common plots for visualizing residuals are normal Q-Q plots, residual vs
fitted value plots, and scale-location plots. You can add these plots to
any rbiom corrplot by setting <code>check = TRUE</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span></span>
<span>  df       <span class="op">=</span> <span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>,  <span class="co"># dataset</span></span>
<span>  x        <span class="op">=</span> <span class="st">'age'</span>,                <span class="co"># x-axis variable</span></span>
<span>  y        <span class="op">=</span> <span class="st">'height'</span>,             <span class="co"># y-axis variable</span></span>
<span>  stat.by  <span class="op">=</span> <span class="st">'sex'</span>,                <span class="co"># statistical groups</span></span>
<span>  fit      <span class="op">=</span> <span class="st">'gam'</span>,                <span class="co"># smoothed trendline</span></span>
<span>  layers   <span class="op">=</span> <span class="st">'tc'</span>,                 <span class="co"># show trendline with conf. int.</span></span>
<span>  test     <span class="op">=</span> <span class="st">'emmeans'</span>,            <span class="co"># run stats on the means</span></span>
<span>  check    <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span>                <span class="co"># display diagnostic plots   &lt;====</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-11-1.png" width="768"></p>
<p>The subtitle for each plot tells you what a high quality model will
look like. For example, the “Normal Q-Q” plot says that “Points should
fall on the line.” In this case they do, so that particular check
indicates high-quality.</p>
<p>These diagnostic plots are largely subjective/qualitative. That is,
you must use your best judgement to decide if the residuals appear to
pass or fail the checks.</p>
<p>A complete explanation of these diagnostic plots is beyond the scope
of this article. Readers are encouraged to explore the topic further at
sites such as:</p>
<ul>
<li><a href="https://www.statology.org/linear-regression-assumptions/" class="external-link">The Four
Assumptions of Linear Regression</a></li>
<li><a href="http://www.sthda.com/english/articles/39-regression-model-diagnostics/161-linear-regression-assumptions-and-diagnostics-in-r-essentials/" class="external-link">Linear
Regression Assumptions and Diagnostics in R: Essentials</a></li>
<li>
<a href="https://library.virginia.edu/data/articles/diagnostic-plots" class="external-link">Understanding
Diagnostic Plots for Linear Regression Analysis</a>.</li>
<li><a href="https://rpubs.com/mbh038/725314" class="external-link">Quantile-Quantile
plots</a></li>
<li><a href="https://www.andrew.cmu.edu/user/achoulde/94842/homework/regression_diagnostics.html" class="external-link">Regression
diagnostic plots</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="gems-dataset">GEMS Dataset<a class="anchor" aria-label="anchor" href="#gems-dataset"></a>
</h2>
<p>Going forward we’ll use the <code>gems</code> dataset included with
rbiom. GEMS is a case-control study comprising 1,006 stool samples from
young children with or without diarrhea in four low-income countries.
The original Nature article by Eric J. de Muinck and Pal Trosvik for
this data set is available online for free at <a href="https://www.nature.com/articles/s41467-018-04641-7" class="external-link">Individuality
and Convergence of the Infant Gut Microbiota During the First Year of
Life</a>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gems</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #808080;">══ </span><span style="color: #BBBBBB;">Global Enteric Multicenter Study</span><span style="color: #808080;"> ════════════════════════</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Stool samples collected from 1006 participants under the</span></span>
<span><span class="co">#&gt; age of 5 from Bangladesh, Gambia, Kenya, and Mali. Includes</span></span>
<span><span class="co">#&gt; 492 healthy controls and 514 moderate-to-severe diarrhea</span></span>
<span><span class="co">#&gt; (MSD) cases. Original study by Pop et al, 2014</span></span>
<span><span class="co">#&gt; (<span style="color: #808080; font-style: italic;">&lt;https://doi.org/10.1186/gb-2014-15-6-r76&gt;</span>); dataset</span></span>
<span><span class="co">#&gt; retrieved from MicrobiomeDB</span></span>
<span><span class="co">#&gt; (<span style="color: #808080; font-style: italic;">&lt;https://doi.org/10.1093/nar/gkx1027&gt;</span>).</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    1006 Samples: SRS608640, SRS608279, SRS607780, ...</span></span>
<span><span class="co">#&gt;     767 OTUs:    OTU001, OTU002, OTU003, ..., and OTU767</span></span>
<span><span class="co">#&gt;       8 Ranks:   .otu, Kingdom, Phylum, ..., and Species</span></span>
<span><span class="co">#&gt;       4 Fields:  .sample, diarrhea, age, and country</span></span>
<span><span class="co">#&gt;         Tree:    <span style="font-style: italic;">&lt;absent&gt;</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #808080;">── 116 - 11k reads/sample ──────────────────── 2021-03-15 ──</span></span></span>
<span><span class="co">#&gt; </span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html" class="external-link">glimpse</a></span><span class="op">(</span><span class="va">gems</span><span class="op">)</span></span>
<span><span class="co">#&gt; Rows: 1,006</span></span>
<span><span class="co">#&gt; Columns: 4</span></span>
<span><span class="co">#&gt; $ .sample  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> "SRS608640", "SRS608279", "SRS607780", "SRS607987", "SRS60842…</span></span>
<span><span class="co">#&gt; $ diarrhea <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Control, Case, Case, Case, Control, Case, Case, Control, Cont…</span></span>
<span><span class="co">#&gt; $ age      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> 0.2, 1.4, 0.8, 1.8, 0.6, 4.0, 0.5, 2.2, 0.4, 1.4, 2.1, 2.0, 0…</span></span>
<span><span class="co">#&gt; $ country  <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> Bangladesh, Kenya, Gambia, Gambia, Kenya, Kenya, Mali, Gambia…</span></span></code></pre></div>
<p>The <code>gems</code> metadata includes two categorical fields and
one continuous field:</p>
<table class="table">
<thead><tr class="header">
<th>Field</th>
<th>Type</th>
<th>Values</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>diarrhea</td>
<td>categorical</td>
<td>Case or Control</td>
</tr>
<tr class="even">
<td>age</td>
<td>continuous</td>
<td>0 - 4.8 (years old)</td>
</tr>
<tr class="odd">
<td>country</td>
<td>categorical</td>
<td>Bangladesh, Gambia, Kenya, or Mali</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="bangladesh-vs-kenya-controls">Bangladesh vs Kenya Controls<a class="anchor" aria-label="anchor" href="#bangladesh-vs-kenya-controls"></a>
</h3>
<p>To start, let’s select just the healthy controls from two countries,
then rarefy those samples to an even sequencing depth.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hbk</span> <span class="op">&lt;-</span> <span class="va">gems</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">diarrhea</span> <span class="op">==</span> <span class="st">"Control"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/subset.html">subset</a></span><span class="op">(</span><span class="va">country</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Bangladesh"</span>, <span class="st">"Kenya"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="../reference/rarefy.html">rarefy</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hbk</span><span class="op">$</span><span class="va">n_samples</span></span>
<span><span class="co">#&gt; [1] 256</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="alpha-diversity">Alpha Diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h2>
<p>Since we’re working with an rbiom object - <code>hbk</code> - we can
use the convenience function <code><a href="../reference/adiv_corrplot.html">adiv_corrplot()</a></code>, which
combines the steps of generating a table of alpha diversity values
(<code><a href="../reference/adiv_table.html">adiv_table()</a></code>) and generating a correlation plot
(<code><a href="../reference/stats_corrplot.html">stats_corrplot()</a></code>).</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/adiv_corrplot.html">adiv_corrplot</a></span><span class="op">(</span></span>
<span>  biom    <span class="op">=</span> <span class="va">hbk</span>,         <span class="co"># healthy controls from Bangladesh and Kenya</span></span>
<span>  x       <span class="op">=</span> <span class="st">"age"</span>,       <span class="co"># x-axis variable</span></span>
<span>  stat.by <span class="op">=</span> <span class="st">"country"</span>,   <span class="co"># statistical groups</span></span>
<span>  fit     <span class="op">=</span> <span class="st">"gam"</span>,       <span class="co"># smoothed trendline</span></span>
<span>  check   <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span>       <span class="co"># display diagnostic plots</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-14-1.png" width="768"></p>
<p>Here, the diagnostic plots look pretty good, so we can confidently
say that children have different Shannon diversity index values,
depending on whether they live in Bangladesh vs Kenya, and that this
difference is most significant at around 1.35 years old.</p>
</div>
<div class="section level2">
<h2 id="taxa-abundance">Taxa Abundance<a class="anchor" aria-label="anchor" href="#taxa-abundance"></a>
</h2>
<p>Similar to <code><a href="../reference/adiv_corrplot.html">adiv_corrplot()</a></code>, <code><a href="../reference/taxa_corrplot.html">taxa_corrplot()</a></code>
combines <code><a href="../reference/taxa_matrix.html">taxa_table()</a></code> and <code><a href="../reference/stats_corrplot.html">stats_corrplot()</a></code>
into a single streamlined function.</p>
<p>We’ll set <code>layers = "tcrp"</code> to show the residuals on this
main plot. This will help in explaining an artifact on the “Residuals vs
Fitted” diagnostic plot.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/taxa_corrplot.html">taxa_corrplot</a></span><span class="op">(</span></span>
<span>  biom    <span class="op">=</span> <span class="va">hbk</span>,         <span class="co"># healthy controls from Bangladesh and Kenya</span></span>
<span>  layers  <span class="op">=</span> <span class="st">"tcrp"</span>,      <span class="co"># show residuals on the main plot</span></span>
<span>  x       <span class="op">=</span> <span class="st">"age"</span>,       <span class="co"># x-axis variable</span></span>
<span>  taxa    <span class="op">=</span> <span class="fl">1</span>,           <span class="co"># select the most abundant taxa</span></span>
<span>  rank    <span class="op">=</span> <span class="st">"Genus"</span>,     <span class="co"># at the genus level</span></span>
<span>  stat.by <span class="op">=</span> <span class="st">"country"</span>,   <span class="co"># statistical groups</span></span>
<span>  fit     <span class="op">=</span> <span class="st">"gam"</span>,       <span class="co"># smoothed trendline</span></span>
<span>  t.size  <span class="op">=</span> <span class="fl">2</span>,           <span class="co"># make the trendline thicker</span></span>
<span>  check   <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span>       <span class="co"># display diagnostic plots</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<p>Above, all the orange and blue points falling on y=0 indicate there
are zero Prevotella reads for that sample. This can confound taxa
abundance analyses, as indicated by the diagnostic plots.</p>
<ul>
<li>Normal Q-Q: This curved shape indicates residuals are more often
below the trendline.</li>
<li>Residuals vs Fitted: The points forming a straight black diagonal
‘line’ are the result of a overabundance of a single value in the data
(zero).</li>
<li>Scale-Location: The first half of the curve is due to lots of zero
counts. The second half of the curve is due to having few samples with
greater than 350 rarefied Prevotella counts, thereby allowing the gam
curve to intercept those sample points with better precision.</li>
</ul>
<div class="section level3">
<h3 id="rank-transformation">Rank Transformation<a class="anchor" aria-label="anchor" href="#rank-transformation"></a>
</h3>
<p>One method for addessing zero abundances is to rank tranform all the
abundances. In rbiom, setting <code>trans = "rank"</code> will rank
transform all the abundances, without consideration of facets. The
default parameters <code>ties = "random", seed = 0</code> will assign
unique integers (see <code><a href="https://rdrr.io/r/base/rank.html" class="external-link">base::rank()</a></code> for other
<code>ties</code> options).</p>
<p>This will improve the diagnostic plots, but adds noise to the
data.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/taxa_corrplot.html">taxa_corrplot</a></span><span class="op">(</span></span>
<span>  biom    <span class="op">=</span> <span class="va">hbk</span>,         <span class="co"># healthy controls from Bangladesh and Kenya</span></span>
<span>  layers  <span class="op">=</span> <span class="st">"tcrp"</span>,      <span class="co"># show residuals on the main plot</span></span>
<span>  x       <span class="op">=</span> <span class="st">"age"</span>,       <span class="co"># x-axis variable</span></span>
<span>  taxa    <span class="op">=</span> <span class="fl">1</span>,           <span class="co"># select the most abundant taxa</span></span>
<span>  rank    <span class="op">=</span> <span class="st">"Genus"</span>,     <span class="co"># at the genus level</span></span>
<span>  stat.by <span class="op">=</span> <span class="st">"country"</span>,   <span class="co"># statistical groups</span></span>
<span>  fit     <span class="op">=</span> <span class="st">"gam"</span>,       <span class="co"># smoothed trendline</span></span>
<span>  trans   <span class="op">=</span> <span class="st">"rank"</span>,      <span class="co"># rank-transform the abundances</span></span>
<span>  t.size  <span class="op">=</span> <span class="fl">2</span>,           <span class="co"># make the trendline thicker</span></span>
<span>  check   <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span>       <span class="co"># display diagnostic plots</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
</div>
</div>
<div class="section level2">
<h2 id="beta-diversity">Beta Diversity<a class="anchor" aria-label="anchor" href="#beta-diversity"></a>
</h2>
<p>Like the last two corrplot functions, <code><a href="../reference/bdiv_corrplot.html">bdiv_corrplot()</a></code>
combines <code><a href="../reference/bdiv_table.html">bdiv_table()</a></code> and <code><a href="../reference/stats_corrplot.html">stats_corrplot()</a></code> to
generate a beta diversity correlation plot from an rbiom object.</p>
<p>Beta diversity correlation plots are more conceptually complex, so
let’s quickly reveal what we’ll be plotting.</p>
<ul>
<li>x-axis: the difference in a numeric metadata field between two
samples.</li>
<li>y-axis: the difference in overall microbiome composition.</li>
</ul>
<p>As with most <code>bdiv_*</code> functions, we can supply a
<code>within</code> and/or <code>between</code> parameters to limit
which pairs of samples are included in the plot. Setting
<code>within = "country"</code> as below will only plot pairs of samples
that came from the same country, and prevent <code>stat.by</code> from
producing a “Bangladesh vs Kenya” trendline.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bdiv_corrplot.html">bdiv_corrplot</a></span><span class="op">(</span></span>
<span>  biom    <span class="op">=</span> <span class="va">hbk</span>,         <span class="co"># healthy controls from Bangladesh and Kenya</span></span>
<span>  x       <span class="op">=</span> <span class="st">"age"</span>,       <span class="co"># x-axis variable</span></span>
<span>  stat.by <span class="op">=</span> <span class="st">"country"</span>,   <span class="co"># statistical groups </span></span>
<span>  within  <span class="op">=</span> <span class="st">"country"</span>,   <span class="co"># don't compare samples from different countries</span></span>
<span>  fit     <span class="op">=</span> <span class="st">"gam"</span>,       <span class="co"># smoothed trendline</span></span>
<span>  check   <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span>       <span class="co"># display diagnostic plots</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>Here, the main plot tells us:</p>
<ul>
<li>In Bangladesh, the closer two children are in age to each other, the
more similar their microbiomes will be. Therefore, you could infer that
age is an important factor driving microbiome structure for this
population.</li>
<li>Kenyan children’s microbiomes are more consistently structured,
compared to children in Bangladesh. Additionally, the age effect is much
less pronounced.</li>
<li>Children about 1 year apart in age have much more similar microbiome
if they live in Kenya, compared to Bangalesh (p = 7e-143).</li>
</ul>
<p>The diagnostic plots here indicate less than optimal distributions of
residuals, with the “Normal Q-Q” plot showing that many more points lie
above the trendline than below it. The main plot’s trendlines don’t
spend much time between y = 0.75 and 0.83, hence the sparsity of points
between x = 0.75 and 0.83 on the “Residuals vs Fitted” and
“Scale-Location” plots.</p>
<div class="section level3">
<h3 id="rank-transformation-1">Rank Transformation<a class="anchor" aria-label="anchor" href="#rank-transformation-1"></a>
</h3>
<p>Running <code><a href="../reference/bdiv_corrplot.html">bdiv_corrplot()</a></code> with <code>trans = "rank"</code>
could help make the residuals more normally distributed.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/bdiv_corrplot.html">bdiv_corrplot</a></span><span class="op">(</span></span>
<span>  biom    <span class="op">=</span> <span class="va">hbk</span>,         <span class="co"># healthy controls from Bangladesh and Kenya</span></span>
<span>  x       <span class="op">=</span> <span class="st">"age"</span>,       <span class="co"># x-axis variable</span></span>
<span>  stat.by <span class="op">=</span> <span class="st">"country"</span>,   <span class="co"># statistical groups </span></span>
<span>  within  <span class="op">=</span> <span class="st">"country"</span>,   <span class="co"># don't compare samples from different countries</span></span>
<span>  fit     <span class="op">=</span> <span class="st">"gam"</span>,       <span class="co"># smoothed trendline</span></span>
<span>  trans   <span class="op">=</span> <span class="st">"rank"</span>,      <span class="co"># rank-transform the abundances</span></span>
<span>  check   <span class="op">=</span> <span class="cn">TRUE</span> <span class="op">)</span>       <span class="co"># display diagnostic plots</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-18-1.png" width="768"></p>
<p>After rank-transforming:</p>
<ul>
<li>The main plot tells a slightly different story.</li>
<li>The “Normal Q-Q” plot indicates that the residuals now have a
light-tailed distribution.</li>
</ul>
<p>You could argue either for or against preferring the transformed
figure over the original.</p>
</div>
<div class="section level3">
<h3 id="goodness-of-fit-1">Goodness of Fit<a class="anchor" aria-label="anchor" href="#goodness-of-fit-1"></a>
</h3>
<p>Comparing the AIC, BIC, and Log-Likelihood values for the
untransformed and rank-transformed data can tell you which model is
better, right? Wrong! These values scale with the range of the data, so
the untransformed beta diversity data in the range of 0 - 1 will result
in drastically different AIC, BIC, and Log-Likelihood values than the
rank-transformed data in the range of 1 - 16400.</p>
<p>Generally speaking, you cannot compare goodness-of-fit values derived
from different datasets. However, it is arguable that these two datasets
are the same, just scaled differently. Therefore if you don’t mind
potentially incurring a reviewer’s wrath, you can rescale both datasets
to the same range and compare goodness-of-fit values on those.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/ldply.html" class="external-link">ldply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>none<span class="op">=</span><span class="st">"none"</span>, rank<span class="op">=</span><span class="st">"rank"</span><span class="op">)</span>, .id <span class="op">=</span> <span class="st">"trans"</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">trans</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/bdiv_table.html">bdiv_table</a></span><span class="op">(</span>biom <span class="op">=</span> <span class="va">hbk</span>, within <span class="op">=</span> <span class="st">"country"</span>, trans <span class="op">=</span> <span class="va">trans</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>.distance <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">.distance</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span>regr <span class="op">=</span> <span class="st">"age"</span>, resp <span class="op">=</span> <span class="st">".distance"</span>, stat.by <span class="op">=</span> <span class="st">"country"</span>, fit <span class="op">=</span> <span class="st">"gam"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.aic</span>, <span class="va">.bic</span>, <span class="va">.loglik</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="co">#&gt;   trans      .aic      .bic   .loglik</span></span>
<span><span class="co">#&gt; 1  none -3301.860 -3217.703  1661.852</span></span>
<span><span class="co">#&gt; 2  rank  4550.565  4666.182 -2260.277</span></span></code></pre></div>
<p>These goodness-of-fit metrics strongly favor the untransformed
data.</p>
</div>
</div>
<div class="section level2">
<h2 id="specific-timepoints">Specific Timepoints<a class="anchor" aria-label="anchor" href="#specific-timepoints"></a>
</h2>
<p>By default, the <code>*_corrplot()</code> functions will
automatically find the x-axis location with the most significant term of
interest. However, you can override this by providing the
<code>at</code> parameter.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_corrplot.html">stats_corrplot</a></span><span class="op">(</span><span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>, <span class="st">'age'</span>, <span class="st">'height'</span>, stat.by <span class="op">=</span> <span class="st">'sex'</span>, at <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span></code></pre></div>
<p><img src="regression_files/figure-html/unnamed-chunk-20-1.png" width="768"></p>
<p>Although you’re limited to one <code>at</code> value for
<code>*_corrplot()</code> functions, you can provide multiple
<code>at</code> locations to <code><a href="../reference/stats_table.html">stats_table()</a></code> or
<code>*_stats()</code> functions.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/stats_table.html">stats_table</a></span><span class="op">(</span><span class="fu">npregfast</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/npregfast/man/children.html" class="external-link">children</a></span>, <span class="st">'age'</span>, <span class="st">'height'</span>, stat.by <span class="op">=</span> <span class="st">'sex'</span>, at <span class="op">=</span> <span class="fl">11</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Model:    gam(height ~ s(age, by = sex, bs = "cs") + sex, method = "REML")</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 15</span></span></span>
<span><span class="co">#&gt;     age sex    .mean.diff .h1     .p.val   .adj.p .effect.size   .se    .n   .df</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span>    12 femal…       7.23 != 0  2.46<span style="color: #949494;">e</span><span style="color: #BB0000;">-29</span> 1.23<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span>        1.13  0.635  <span style="text-decoration: underline;">2</span>500  <span style="text-decoration: underline;">2</span>485</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span>    11 femal…       6.02 != 0  2.75<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span> 6.88<span style="color: #949494;">e</span><span style="color: #BB0000;">-22</span>        0.939 0.614  <span style="text-decoration: underline;">2</span>500  <span style="text-decoration: underline;">2</span>485</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span>    13 femal…       6.04 != 0  6.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-21</span> 1.01<span style="color: #949494;">e</span><span style="color: #BB0000;">-20</span>        0.943 0.638  <span style="text-decoration: underline;">2</span>500  <span style="text-decoration: underline;">2</span>485</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span>    15 femal…      -<span style="color: #BB0000;">2.58</span> != 0  7.21<span style="color: #949494;">e</span><span style="color: #BB0000;">- 7</span> 9.02<span style="color: #949494;">e</span><span style="color: #BB0000;">- 7</span>       -<span style="color: #BB0000;">0.402</span> 0.519  <span style="text-decoration: underline;">2</span>500  <span style="text-decoration: underline;">2</span>485</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span>    14 femal…       2.00 != 0  3.14<span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span> 3.14<span style="color: #949494;">e</span><span style="color: #BB0000;">- 4</span>        0.311 0.553  <span style="text-decoration: underline;">2</span>500  <span style="text-decoration: underline;">2</span>485</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 5 more variables: .t.ratio &lt;dbl&gt;, .adj.r &lt;dbl&gt;, .aic &lt;dbl&gt;, .bic &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   .loglik &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Daniel P. Smith, Alkek Center for Metagenomics and Microbiome Research.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
